<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-07-16T22:27:38-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-07-16T22:27:38-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-07-16T22:27:38-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
     <div class="excerpt">
        <ul><li>cboettig pushed to master at cboettig/reml: <em>Merge branch 'master' of github.com:ropensci/reml documentation updated from vignette</em> <a href="https://github.com/cboettig/reml/compare/606e9b537b...3ed47007f9">07:27 2013/07/25</a></li><li>cboettig pushed to master at ropensci/reml: <em>add comment about coercion methods</em> <a href="https://github.com/ropensci/reml/compare/dc275adc0b...3ed47007f9">07:27 2013/07/25</a></li><li>cboettig pushed to master at cboettig/labnotebook: <em>layout: post</em> <a href="https://github.com/cboettig/labnotebook/compare/bc95dd9c43...c97b66b55b">07:20 2013/07/25</a></li><li>cboettig pushed to master at cboettig/labnotebook: <em>fix code block rendering define post layout</em> <a href="https://github.com/cboettig/labnotebook/compare/daf4633c2a...bc95dd9c43">07:06 2013/07/25</a></li><li>cboettig pushed to master at cboettig/cboettig.github.com: <em>update site</em> <a href="https://github.com/cboettig/cboettig.github.com/compare/25e6ae25d6...0d4e5574e3">07:01 2013/07/25</a></li></ul>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
       <ul><li><p>Revisiting my thoughts on software review: <a href="http://t.co/KStq4RyrLK">http://t.co/KStq4RyrLK</a> clearer, better criteria? @omearabrian @dwbapst @ethanwhite @phylorich ..</p>
 <a href="http://twitter.com/cboettig/statuses/360477303538585600">12:11 2013/07/25</a> </li><li><p>@ucfagls @noamross @recology_ Yup. Ideally I would just maintain my @ORCID_Org id and everything else would update automatically from that</p>
 <a href="http://twitter.com/cboettig/statuses/360399014757269504">06:59 2013/07/25</a> </li><li><p>RT @_inundata: Heading to #ESA2013?
There are a ton of great workshops + sessions organized by @nceas @DataONEorg  and @ropensci 
<a href="http://t.%E2%80%A6">http://t.…</a></p>
 <a href="http://twitter.com/cboettig/statuses/360398203285278720">06:56 2013/07/25</a> </li><li><p>&quot;On writing, sharing, collaborating, and hosting code for science&quot; <a href="http://t.co/Mqnnov9CIV">http://t.co/Mqnnov9CIV</a> from @recology_</p>
 <a href="http://twitter.com/cboettig/statuses/360249754124435457">09:06 2013/07/24</a> </li><li><p>@noamross True, but If there&#39;s anything they should be signing up for, it&#39;s probably @ORCID_Org (vs needing wordpress account, etc)</p>
 <a href="http://twitter.com/cboettig/statuses/360242446455668740">08:37 2013/07/24</a> </li></ul>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <ul>
<li>Stephan B Munch, A Kottas. 2009. A Bayesian modeling approach for determining productivity regimes and their characteristics <em>Ecological Applications</em> <strong>19</strong> 2 10.1890/07-2116.1</li>
<li>Mark John Costello, Marta Coll, Roberto Danovaro et al. 2010. A Census of Marine Biodiversity Knowledge, Resources, and Future Challenges <em>PLoS ONE</em> <strong>5</strong> 8 10.1371/journal.pone.0012110</li>
<li>R. F. Hechinger, K. D. Lafferty, a. P. Dobson et al. 2011. A Common Scaling Rule for Abundance, Energetics, and Production of Parasitic and Free-Living Species <em>Science</em> <strong>333</strong> 6041 10.1126/science.1204337</li>
<li>Jonathan B. Shurin, Elizabeth T. Borer, Eric W. Seabloom et al. 2002. A cross-ecosystem comparison of the strength of trophic cascades <em>Ecology Letters</em> <strong>5</strong> 6 10.1046/j.1461-0248.2002.00381.x</li>
<li>Wayne M Getz, David Saltz. 2008. A framework for generating and analyzing movement paths on ecological landscapes. <em>Proceedings of the National Academy of Sciences of the United States of America</em> <strong>105</strong> 49 10.1073/pnas.0801732105</li>

    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/07/22/XML-parsing-strategies.html">Xml Parsing Strategies</a></h4></header>
<p><span>parsing NeXML</span></p>
<p style="font-style:italic"> 22 Jul 2013</p>

<p style="font-style:italic"> pageviews: 1 </p>

<article>
<div class="excerpt">
<p>Writing out some advice for our GSOC student on XML parsing. (Now filed as <a href="https://github.com/shumelchyk/RNeXML/issues/11">RNeXML/#11</a></p>
<hr />
<p>Here is some quick background on different ways we might go about extracting NeXML into an R object we want to work with. We can use S4 classes, R’s native <code>data.frame</code> and <code>list</code> types, or extract specific terms of interest with xpath. I illustrate each of these below using the example “trees.xml” in your repository.</p>
<p>It would be good to decide on which strategy you want. As you seem to be leaning to the S4 already, note how I have made the slot labels correspond exactly to the attribute and node labels of the NeXML. This lets us use the <code>xmlToS4</code> method:</p>
<h2 id="using-s4-classes">Using S4 classes</h2>
<p>We can define S4 class structures to correspond to the elements we want to extract. Elements omitted from our definition will be ignored.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;node&quot;</span>,
    <span class="kw">representation</span>(<span class="dt">id    =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">label =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">otu   =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">about =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">meta =</span> <span class="st">&quot;meta&quot;</span>))</code></pre>
<p>Note that class “meta” is not yet defined, we will define it below. Note also that we could omit any slots we don’t care about, and they would simply be ignored. (Try deleting the <code>meta</code> line, or the <code>about</code> line.)</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;edge&quot;</span>,
    <span class="kw">representation</span>(<span class="dt">source =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">target =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">id     =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">length =</span> <span class="st">&quot;numeric&quot;</span>))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;meta&quot;</span>,
         <span class="kw">representation</span>(<span class="dt">id =</span><span class="st">&quot;character&quot;</span>,
                        <span class="dt">property=</span><span class="st">&quot;character&quot;</span>,
                        <span class="dt">content=</span><span class="st">&quot;logical&quot;</span>,
                        <span class="st">&#39;xsi:type&#39;</span>=<span class="st">&quot;character&quot;</span>,
                        <span class="dt">datatype=</span><span class="st">&quot;character&quot;</span>))

<span class="kw">setAs</span>(<span class="st">&quot;XMLInternalElementNode&quot;</span>, <span class="st">&quot;meta&quot;</span>, function(from) <span class="kw">xmlToS4</span>(from))</code></pre>
<p>In order for <code>xmlToS4</code> to work on nodes, I need the Coercion method for <code>meta</code> too, defined above. We are now ready to rock and roll:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(XML)
doc &lt;- <span class="kw">xmlParse</span>(<span class="st">&quot;tests/examples/trees.xml&quot;</span>)

edges &lt;- <span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>)
edges &lt;- <span class="kw">lapply</span>(edges, xmlToS4)

nodes &lt;- <span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>)
nodes &lt;- <span class="kw">lapply</span>(nodes, xmlToS4)</code></pre>
<p>My opinion is that we should not bother writing validator methods etc for these classes we have just defined at this time. We can already validate the XML produced against the schema, and as the class definitions follow those from the schema, we should be reasonably robust. Happy to hear counter-arguments.</p>
<h2 id="extracting-attributes-or-values-directly-by-xpath-query">Extracting Attributes or values directly by xpath query</h2>
<p>A different route does not define class types at all, but just extracts the attributes we want. This is a bit more fast and loose.</p>
<pre class="sourceCode r"><code class="sourceCode r">edges = <span class="kw">xpathSApply</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces =</span> <span class="st">&quot;x&quot;</span>, xmlAttrs)
nodes = <span class="kw">xpathSApply</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces =</span> <span class="st">&quot;x&quot;</span>, 
  function(x) 
    <span class="kw">c</span>(<span class="kw">xmlGetAttr</span>(x, <span class="st">&quot;id&quot;</span>, <span class="ot">NA</span>, as.character),  
      <span class="kw">xmlGetAttr</span>(x, <span class="st">&quot;otu&quot;</span>, <span class="ot">NA</span>, as.character)))</code></pre>
<h2 id="coercing-into-standard-r-forms">Coercing into standard R forms</h2>
<p>This approach can be very quick and powerful if the data conforms to this structure.</p>
<pre class="sourceCode r"><code class="sourceCode r">XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>))
XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>))</code></pre>
<h2 id="without-xpath">without XPath</h2>
<p>We can skip over XPath based expressions by using xmlToList immediately:</p>
<pre class="sourceCode r"><code class="sourceCode r">nex &lt;- <span class="kw">xmlToList</span>(doc)</code></pre>
<p>Unfortunately, this will do some possibly unexpected things: for instance, attributes are converted to list elements (<code>node$id</code>, <code>node$otu</code>, etc) on simple nodes (e.g. <code>&lt;nodes&gt;</code> without a <code>&lt;meta&gt;</code> node), while if they contain another node, attributes are one sub-list and the containing nodes another (e.g. <code>node$meta</code> and <code>node$.attr</code>, etc)</p>
<p>Also note that instead of xpath queries as in the <code>xpathSApply</code> and <code>getNodeSet</code> examples above, we can index nodes the way would lists:</p>
<p>e.g. <code>xmlRoot(doc)[[&quot;trees&quot;]][[&quot;tree&quot;]]</code> instead of <code>getNodeSet(doc, &quot;//trees/tree&quot;)[[1]]</code> (note that the former returns the first <code>&lt;tree&gt;</code> node in the <code>&lt;trees&gt;</code> node, while the latter returns all <code>&lt;tree&gt;</code> nodes unless we just ask for the <code>[[1]]</code></p>
<p>So this gives us a table of all node and edge elements:</p>
<pre class="sourceCode r"><code class="sourceCode r">XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">xmlRoot</span>(doc)[[<span class="st">&quot;trees&quot;</span>]][[<span class="st">&quot;tree&quot;</span>]])</code></pre>
<p>Which gives us:</p>
<pre class="sourceCode r"><code class="sourceCode r">   id label root  otu about source target  length
<span class="dv">1</span>  n1    n1 true &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">2</span>  n2    n2 &lt;<span class="ot">NA</span>&gt;   t1  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">3</span>  n3    n3 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">4</span>  n4    n4 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;   <span class="co">#n4   &lt;NA&gt;   &lt;NA&gt;    &lt;NA&gt;</span>
<span class="dv">5</span>  n5    n5 &lt;<span class="ot">NA</span>&gt;   t3  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">6</span>  n6    n6 &lt;<span class="ot">NA</span>&gt;   t2  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">7</span>  n7    n7 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">8</span>  n8    n8 &lt;<span class="ot">NA</span>&gt;   t5  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">9</span>  n9    n9 &lt;<span class="ot">NA</span>&gt;   t4  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">10</span> e1  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n1     n3 <span class="fl">0.34534</span>
<span class="dv">11</span> e2  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n1     n2  <span class="fl">0.4353</span>
<span class="dv">12</span> e3  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n3     n4   <span class="fl">0.324</span>
<span class="dv">13</span> e4  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n3     n7  <span class="fl">0.3247</span>
<span class="dv">14</span> e5  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n4     n5   <span class="fl">0.234</span>
<span class="dv">15</span> e6  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n4     n6  <span class="fl">0.3243</span>
<span class="dv">16</span> e7  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n7     n8 <span class="fl">0.32443</span>
<span class="dv">17</span> e8  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n7     n9  <span class="fl">0.2342</span></code></pre>
<p>Sometimes this notation is cleaner than the xpath, sort of up to you.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/22/XML-parsing-strategies.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/19/notes.html">Notes from chat with Duncan</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 19 Jul 2013</p>

<p style="font-style:italic"> pageviews: 1 </p>

<article>
<div class="excerpt">
<ul>
<li>Keep metadata as part of the object</li>
<li>Recursion problem coming up.</li>
</ul>
<h2 id="validating">Validating:</h2>
<ul>
<li>https://github.com/ropensci/reml/issues/7</li>
<li>Duncan has code using Matt’s java API</li>
<li>If the EML isn’t valid, so what? Important for developers, users don’t care.</li>
</ul>
<h2 id="api-design">API design</h2>
<ul>
<li>what’s with the <code>do.call</code>? Can lead to unexpected behavior.</li>
<li>Schema to create S4 objects</li>
<li>Create R objects first, and then toXML</li>
</ul>
<h2 id="xmlschema">XMLSchema</h2>
<ul>
<li>will improve as we go</li>
</ul>
<h2 id="paper-outline">Paper outline</h2>
<ul>
<li>EML exists – metadata vs vertical</li>
<li>metadata manual in, manual out</li>
<li>data synthesis?</li>
</ul>
<h3 id="r-object-to-xml">R object to XML</h3>
<ul>
<li>Accept / identify losely defined fields (lists)</li>
<li>Manually create a coercision method between R’s objects, e.g. “person” (attempt match of named elements to list slots automatically).<br /></li>
<li>Separate coersion methods for read and write.</li>
</ul>
<h3 id="xml-to-r-object">XML to R object</h3>
<ul>
<li>ParseEML - because it saves us a lot to know this is EML.<br /></li>
<li>Coerce into generic S4 R classes (class definitions from schema)</li>
<li>Coerce automatically if element names are the same. Otherwise, manually coerce.</li>
</ul>
<h3 id="further-possibilities">Further possibilities</h3>
<ul>
<li>Translations to other languages (e.g. python) can be automated.<br /></li>
<li>Then add higher level mapping manually.</li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/19/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/17/updates-on-managing-references.html">Updates On Managing References</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 17 Jul 2013</p>

<p style="font-style:italic"> pageviews: 10 </p>

<article>
<div class="excerpt">
<h3 id="mendeley-api-for-jekyll">Mendeley API for Jekyll</h3>
<p>Finally finished off my <a href="https://github.com/cboettig/labnotebook/blob/7f7137fbdd4ca2eed86765b61561c283b4cd5811/_plugins/jekyll-labnotebook-plugins/mendeley.rb">Mendeley API for Jekyll</a> (The top uses the Ruby API implementation from Michael Bus, copied here since it is not yet provided as a gem. scroll to the bottom for my plugin code). The Mendeley API implementation is remarkably clumsy. For instance, rather than allowing queries to the API that specify a user’s folder and a sort-by field, we are forced to first query the folder for a list of IDs and then query document details <em>on every document in the folder</em>, after which we can sort locally. Hence to show recent additions to a folder with 100s of documents requires 100s of API calls, instead of one single call. Straight-forward to program but incredibly slow and frustrating. Apparently this might be addressed in an eventual update of the API.</p>
<p>Meanwhile, at least I know have a reasonably powerful tool to add reading lists by subject area directly from my library rather than pulling from my rather more aggregated group RSS feeds. Due to the ridiculous number of API calls, the plugin writes out the data in JSON, which provides a nice local database of the publication metadata we can read in from instead. The plugin will use local JSON file for the data if it exists – so one must delete the JSON file to trigger an update.</p>
<p>Currently the plugin lists the <code>n</code> most recent papers in a given folder. Further adaptation might be nice to allow for richer filtering, e.g. by tag,author, etc.</p>
<h3 id="direct-access-to-the-local-sql-database">Direct access to the local SQL database</h3>
<p>Accessing personal Mendeley data from the API is actually quite silly when the data is already available in a local SQLite database.</p>
<p>Using R to explore and query my local Mendeley sqlite database directly:</p>
<p>Make a copy of the database first</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cp</span> ~/.local/share/data/Mendeley Ltd./Mendeley Desktop/cboettig@gmail.com@www.mendeley.com.sqlite tmp.sqlite</code></pre>
<p>Then in R, read in the database and show available tables (see <a href="http://sandymuspratt.blogspot.com/2012/11/r-and-sqlite-part-1.html">example tutorial</a>):</p>
<p><code>{r} library('sqldf') db &lt;- dbConnect(SQLite(), dbname=&quot;tmp.sqlite&quot;) dbListTables(db)</code></p>
<p>Most metadata for all documents in the library appears in table <code>Documents</code>. <code>dbListFields</code> shows the column headers, though alternatively we can just read this information into an R <code>data.frame</code> (This makes sense for reasonable size databases; for large databases we will want to query the data via SQL commands directly rather than attempting to load it all into memory).</p>
<p><code>{r} dbListFields(db, &quot;Documents&quot;) documents &lt;- dbReadTable(db, &quot;Documents&quot;)</code></p>
<p>Several useful fields are not included in this data frame. In particular, Author, Tag, Folder and Keywords. Some extra leg-work is needed to add this data. For instance <code>DocumentFolders</code> gives us the map of each article <code>id</code> to a folder <code>id</code>, and then <code>Folders</code> gives the details of that Folder.</p>
<p>Using the <code>data.table</code> package, indexing by <code>id</code>, and using join methods would probably be the most efficient way to do this. (example may follow when I get the chance).</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/17/updates-on-managing-references.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/07/11/visit-to-WHOI.html">Visit To Whoi</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 11 Jul 2013</p>

<p style="font-style:italic"> pageviews: 12 </p>

<article>
<div class="excerpt">
<p>Presented my work from the <a href="/tags/#nonparametric-bayes">nonparametric-bayes</a> project, <em>Avoiding tipping points in the management of ecological systems: a non-parametric Bayesian approach</em> at the <a href="http://www.whoi.edu/">Woods Hole Oceangraphic Institution</a> as part of their weekly Biology Department Seminar Series.</p>
<ul>
<li>9 - 9:30 <a href="http://www.jbkellner.com/">Julie Kellner</a></li>
<li>9:30-10:15 Ben Jones, (Kellner group)</li>
<li>10:15 - 11:00 Math ecology group</li>
<li>11:00 - 11:30 <a href="http://www.whoi.edu/hpb/Site.do?id=14853">Ian Carroll</a></li>
<li>11:30-12 Seminar prep</li>
<li>12-1 my seminar</li>
<li>1-2 Lunch with the Marine Policy Center (<a href="http://www.whoi.edu/oceanus/viewArticle.do?id=3992">Mike Fogarty</a>, <a href="http://www.whoi.edu/hpb/Site.do?id=1692">Hauke Kite-Powell</a>, <a href="http://www.whoi.edu/hpb/Site.do?id=172">Di Jin</a>), Julie Kellner</li>
</ul>
<p>Enjoyable visit, with some particularly great questions after talk: including challenges/opportunities in multi-dimensional examples, concerns about temporal heterogeneity, partially observed systems / working with catch data in place of stock assessment data. Less push-back than I anticipated on using the Gaussian process; probably aided by the SDP context. Quite broad interest in SDP in particular too, estimating MSY following regression models seems common. Interesting discussion about the challenge introduced by today’s emphasis on ecosystem management; how to still retain focus on tractable models.</p>
<p>Also some good discussion on the rOpenSci goals of large-scale data access and integration.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/11/visit-to-WHOI.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/10/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 10 Jul 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<h3 id="prsb-paper-is-out.">PRSB paper is out.</h3>
<ul>
<li><p>TODO: Post to arxiv now that Drake’s side is out. Update vita. Reply to Drake. Clean up prosectors-fallacy repository. Post on paper(?)</p></li>
<li><p>Note: github record of earlier version of paper turns out handy in reply, (showed earlier wording about not necessary).</p></li>
<li><p>Also, some quick significance testing of <a href="https://github.com/cboettig/prosecutors-fallacy/blob/092b004616e2859c1ab1a2d64892c641c7a527d0/inst/doc/sig-tests.md">distributional differences</a> associated with the patterns we show in Figure 1, for what it’s worth (in reply to query - yay for clean workflow making this easy).</p></li>
<li><p>one day, follow up with Sebastian about the Freidlin-Wentzell work. “Yes, I’d be interested in trying a little dynamic programming for the discrete-time variational problem; it would be educational for me at least, and potentially applicable to some more specific research questions.” (me, April 23).</p></li>
<li><p>Read remaining articles in special issue of TE on warning signals, write out notes. Clear out ‘unsorted’ backlog</p></li>
<li><p>forking the r-journal; see separate post.</p></li>
</ul>
<h3 id="some-thoughts-on-publishing-reviews.">Some thoughts on publishing reviews.</h3>
<ul>
<li><p><a href="http://smallpondscience.com/2013/04/15/transparency-in-research-publish-your-reviews/">Excellent and provactive example by McGlynn</a>, who links all reviews on papers to which he is a first author (the full cascade, not just reviews from the journal that ultimately accepted the paper). Comments</p></li>
<li><p>More thoughts and some comments from me on <a href="http://timotheepoisot.fr/2013/07/10/sharing-reviews">Tim’s blog</a></p></li>
<li><p>Celebrating seven years of Biology Direct <a href="http://dx.doi.org/10.1186/1745-6150-8-11">Koonin et al 2013</a></p></li>
</ul>
<p><a href="http://figshare.com/articles/Moving_toward_a_sustainable_ecological_science_don_t_let_data_go_to_waste_/693745">Poisot et al</a> on data sharing. Nice piece, but the value of a metadata standard for distributing data is clearly not appreciated. Authors advocate data shared over an API without any notion of schema, and do not discuss the challenges and coarse-graining that goes into such vertically integrated repositories.</p>
<h3 id="reviewing-the-new-broader-impacts-guidelines">Reviewing the new Broader impacts guidelines</h3>
<p>2013 changes removes the examples, adds three new potential areas, and now uses the same assessment guidelines as the merit review does – importantly: requiring metrics of success.</p>
<p><em>Include but not limited to:</em></p>
<ol type="1">
<li>full participation of women, persons with disabilities, and underrepresented minorities in science, technology, engineering, and mathematics (STEM); (diversity)</li>
<li>improved STEM education and educator development at any level; (education)</li>
<li>increased public scientific literacy and public engagement with science and technology; (dissemination)</li>
<li>improved well-being of individuals in society; (society)</li>
<li>increased partnerships between academia, industry, and others; (partnerships) (new)</li>
<li>improved national security; (security) (new)</li>
<li>increased economic competitiveness of the United States; (economy) (new)</li>
<li>enhanced infrastructure for research and education. (infrastructure)</li>
</ol>
<p><a href="http://www.nsf.gov/pubs/policydocs/pappguide/nsf13001/gpg_2.jsp#IIC2d">official text in the GPG</a></p>
<p><strong>Criteria</strong></p>
<ol type="1">
<li>What is the potential for the proposed activity to:</li>
</ol>
<ol type="a">
<li>Advance knowledge and understanding within its own field or across different fields (Intellectual Merit); and</li>
<li>Benefit society or advance desired societal outcomes (Broader Impacts)?</li>
</ol>
<ol start="2" type="1">
<li>To what extent do the proposed activities suggest and explore creative, original, or potentially transformative concepts?</li>
<li>Is the plan for carrying out the proposed activities well-reasoned, well-organized, and based on a sound rationale? Does the plan incorporate a mechanism to assess success?</li>
<li>How well qualified is the individual, team, or organization to conduct the proposed activities?</li>
<li>Are there adequate resources available to the PI (either at the home organization or through collaborations) to carry out the proposed activities?</li>
</ol>
<p><a href="http://www.nsf.gov/pubs/policydocs/pappguide/nsf13001/gpg_3.jsp#IIIA2">official text in the GPG</a></p>
<h3 id="packages">packages</h3>
<ul>
<li>revisit Noam’s <code>knitcitations</code> pull request? (<a href="https://github.com/cboettig/knitcitations/pull/45">#45</a>)</li>
<li>pushed documentation to RMendeley. <code>authored</code> works for me, funny error for Noam…</li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/10/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/10/forking-the-R-journal.html">Forking The R Journal</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 10 Jul 2013</p>

<p style="font-style:italic"> pageviews: 107 </p>

<article>
<div class="excerpt">
<p>All I really wanted to do is cite a paper in <em>the R Journal</em>, which is peer-reviewed, open access (CC-BY), LaTeX based and without author charges. Sure, I could do that already, but I like being able to programmatically generate citation metadata from a URL – we do have this thing called the internet now, and citations are just links, right? Unfortunately, nice as it is, the R journal doesn’t have HTML landing pages for articles that embed the metadata.</p>
<p>This makes it harder for Google Scholar to index the articles, and means that we cannot extract citation metadata from the URL using a tool like <a href="http://greycite.knowledgeblog.org">greycite</a>. Until now.</p>
<p>I wrote to Editor-in-Chief Hadley Wickam about this, who responded in the best way possible: making the <a href="ihttps://github.com/rjournal/rjournal.github.io">journal website’s Github repository</a> public. A fork and a little hacking later, and voila, we have html landing pages for <em>the R Journal</em> with embedded metadata (pending a <a href="https://github.com/rjournal/rjournal.github.io/pull/1">pull request</a>). Check out <a href="https://github.com/cboettig/rjournal.github.io/commit/e70e84e6e53e6c04ec9864af162d7ba58439d4d5">the source code</a> for how this works – it’s really quite straight forward since the metadata is already available in <code>_config.yaml</code>. The main step involves a <a href="https://github.com/cboettig/rjournal.github.io/blob/e70e84e6e53e6c04ec9864af162d7ba58439d4d5/_plugins/article_html_pages.rb">Generator plugin</a> which builds a page for each article and makes the relevant article metadata available to the page. Then we can write a <a href="https://github.com/cboettig/rjournal.github.io/blob/e70e84e6e53e6c04ec9864af162d7ba58439d4d5/_layouts/article.html">page template</a> using Liquid code to import the metadata.</p>
<p>The really exciting thing about this is the basic idea of forking a journal website and improving it. There’s a lot we could do to improve the R journal. I think that ideally we’d have HTML5 versions of the articles as well, something that would be straight forward if authors used knitr of course, but I realize that’s a bigger shift. One could also automatically enhance the HTML with quite a bit more semantic content, we could have animations, links the Rnw files, etc etc. I’d be happy to help and I’m sure others would too. R-journal is great but it could be so much more as an example and test-bed of innovation in statistical publishing.</p>
<p>Well, now that the journal is open for pull requests, hack away!</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/10/forking-the-R-journal.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/07/09/reviewing-software-revisited.html">Reviewing Software Revisited</a></h4></header>
<p><span>Great ideas from the Journal of Open Research Software</span></p>
<p style="font-style:italic"> 09 Jul 2013</p>

<p style="font-style:italic"> pageviews: 44 </p>

<article>
<div class="excerpt">
<p>Several weeks ago I wrote a <a href="/2013/06/13/what-I-look-for-in-software-papers.html">post</a> describing issues I commonly raise when reviewing software papers. This provoked a lively discussion among authors, editors, and other reviewers of software papers (most of us having been on at least both reviewing and authoring end), with quite varied perspectives as to whether such criteria were appropriate in this context. As I describe there, I view the concept of dedicated software papers as a somewhat necessary “hack” of the publishing system, and hope a more mature system will eventually come into place, as is now happening for data through efforts such as <a href="http://datadryad.org">Dryad</a> and associated journal archiving requirements.</p>
<p>Sustainability of the software is important for the same reason archiving the literature is important – without this, we cannot build on existing work. Building on “researcher/developer” software (e.g. the kind of software typically covered in ‘software papers’) is currently an uncommon and risky, as eloquently argued by Brian O’Meara in the comments of <a href="http://carlboettiger.info/2013/04/23/we-need-more-object-oriented-design-in-comparative-methods.html">an earlier post</a> describing just such a failure.</p>
<p>Through that discussion, I’ve come to see the goal of reviewing software papers as one of software sustainability. After all, the publication becomes a part of the researcher’s permanent record and part of the permanent scholarly archive, designed to outlive the journal itself. It seems reasonable that the software it describes should at least exist and do something five years later. Though this sentiment underlies most of my criteria, my phrasing is often too specific and a little too grounded in language explicit to software development.</p>
<p>As I learned from Neil Chue Hong (<a href="http://www.software.ac.uk/">Software Sustainability Institute</a>) in a follow up from that discussion, The Journal of Open Research Software has recently revised its guidelines making much of this more explicit and much better worded than I did. They provide separate lists for the reviewer to comment on both the paper and the software – making it clear that the review goes beyond the paper. The questions for the paper are every bit as valuable as those from the software section: for instance, reference to sections on “Reuse”, “Quality Control”, and “Implementation and Architecture”. While these things might not get their own section headings in an MEE paper, the ideas should certainly be addressed. I think these guidelines provide an excellent checklist for promoting sustainable software without putting undo emphasis on questions of performance, scalability, extensibility, etc.</p>
<h2 id="journal-of-open-research-softwares-guidelines">Journal of Open Research Software’s Guidelines</h2>
<p>(content below <a href="http://creativecommons.org/licenses/by/3.0/">CC-BY</a>, quoted from <a href="http://openresearchsoftware.metajnl.com/">Journal of Open Research Software</a>)</p>
<p><em>Please provide a list of your recommendations, indicating which are compulsory for acceptance and which are optional but would improve the quality of the paper or the reusability of the software.</em></p>
<h3 id="the-paper">The paper</h3>
<ol type="a">
<li>Is the title of the paper descriptive and objective?</li>
<li>Does the Abstract give an indication of the software’s functionality, and where it would be used?</li>
<li>Do the keywords enable a reader to search for the software?</li>
<li>Does the Introduction give enough background information to understand the context of the software’s development and use?</li>
<li>Does the Implementation and Architecture section give enough information to get an idea of how the software is designed, and any constraints that may be placed on its use?</li>
<li>Does the Quality Control section adequately explain how the software results can be trusted?</li>
<li>Does the Reuse section provide concrete and useful suggestions for reuse of the software, for instance: other potential applications, ways of extending or modifying the software, integration with other software?</li>
<li>Are figures and diagrams used to enhance the description? Are they clear and meaningful?</li>
<li>Do you believe that another researcher could take the software and use it, or take the software and build on it?</li>
</ol>
<h3 id="the-software">The software</h3>
<ol type="a">
<li>Is the software in a suitable repository? (see <a href="http://openresearchsoftware.metajnl.com/about/repositories/">http://openresearchsoftware.metajnl.com/about/repositories/</a> for more information)</li>
<li>Does the software have a suitable open licence? (see <a href="http://openresearchsoftware.metajnl.com/faq/#q5">http://openresearchsoftware.metajnl.com/faq/#q5</a> for more information)</li>
<li>If the Archive section is filled out, is the link in the form of a persistent identifier, e.g. a DOI? Can you download the software from this link?</li>
<li>If the Code repository section is filled out, does the identifier link to the appropriate place to download the source code? Can you download the source code from this link?</li>
<li>Is the software license included in the software in the repository? Is it included in the source code?</li>
<li>Is sample input and output data provided with the software?</li>
<li>Is the code adequately documented? Can a reader understand how to build/deploy/install/run the software, and identify whether the software is operating as expected?</li>
<li>Does the software run on the systems specified? (if you do not have access to a system with the prerequisite requirements, let us know)</li>
<li>Is it obvious what the support mechanisms for the software are?</li>
</ol>
<hr />
<p>Like my own list, a repository, a license, and sample input-output data are all mentioned explicitly. Software review section suggests there should be a mechanism in place to “identify whether the software is operating as expected” – such as basic unit tests. Extensibility is also mentioned, but without any reference to features that may or may not assist in that – a fault of my own list, which was probably over-specific on that account. I think these are reasonable and solid guidelines, and would love to see other journals that frequently publish software papers (Looking at you, MEE, JSS, Bioinformatics, R Journal, PLoS Comp Bio) adopt similar criteria.</p>
<p>Of course I’d love to hear what others think about this.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/09/reviewing-software-revisited.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/26/starting-an-EML-package-for-R.html">Starting An Eml Package For R</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 26 Jun 2013</p>

<p style="font-style:italic"> pageviews: 38 </p>

<article>
<div class="excerpt">
<p>Started <a href="https://github.com/ropensci/reml"><code>reml</code></a>, and R package for reading, writing, manipulating and publishing EML files. More details and progress so far on the Github page; in particular flushing out design goals in the <a href="https://github.com/ropensci/reml/issues">issues tracker</a>.</p>
<h3 id="initial-goals">Initial goals</h3>
<p>To start, I am aiming for a proof of principle case in which we can write an R <code>data.frame</code> into EML (<a href="https://github.com/ropensci/reml/issues/2">issue #2</a>), publish to a remote repository such as figshare (<a href="https://github.com/ropensci/reml/issues/3">issue #3</a>), and then read the data back in as the original R object types from the EML (<a href="https://github.com/ropensci/reml/issues/6">issue #6</a>). This will require specifying some generic metadata and basic column and unit metadata for the data frame. It would be good to get feedback on what way is most convenient to specify the required metadata in R.</p>
<h3 id="integration-goals-issue-8">Integration goals (issue <a href="https://github.com/ropensci/reml/issues/8">#8</a>)</h3>
<p>This is the <strong>holy grail</strong> of metadata infrastructure and ostensibly the primary purpose of EML, see <a href="http://dx.doi.org/10.1146/annurev.ecolsys.37.091305.110031" title="The New Bioinformatics: Integrating Ecological Data from the Gene to the Biosphere in Annual Review of Ecology, Evolution, and Systematics">Jones et al 2006</a>. Despite that, integration is not actually possible without semantic definitions as well, see <a href="http://dx.doi.org/10.1016/j.tree.2011.11.016" title="Ecoinformatics: supporting ecology as a data-intensive science; in Trends in Ecology &amp; Evolution. doi: 10.1016/j.tree.2011.11.016">Michener &amp; Jones 2012</a>, from which we adapt this minimal example below.</p>
<p>This example provides minimal and sometimes missing semantics; which may make it unresolvable. A complete semantic solution is diagrammed in the figure from <a href="http://dx.doi.org/10.1016/j.tree.2011.11.016" title="Ecoinformatics: supporting ecology as a data-intensive science; in Trends in Ecology &amp; Evolution. doi: 10.1016/j.tree.2011.11.016">Michener &amp; Jones 2012</a>.</p>
<h3 id="dataset-1">Dataset 1</h3>
<pre class="sourceCode ruby"><code class="sourceCode ruby"> dat = data.frame(river=c(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>), 
                   spp = c(<span class="st">&quot;king&quot;</span>, <span class="st">&quot;king&quot;</span>, <span class="st">&quot;ccho&quot;</span>), 
                   stg = c(<span class="st">&quot;smolt&quot;</span>, <span class="st">&quot;parr&quot;</span>, <span class="st">&quot;smolt&quot;</span>),
                   ct =  c(293L, 410L, 210L))

 col_metadata = c(river = <span class="st">&quot;http://dbpedia.org/ontology/River&quot;</span>,
                  spp = <span class="st">&quot;http://dbpedia.org/ontology/Species&quot;</span>, 
                  stg = <span class="st">&quot;Life history stage&quot;</span>,
                  ct = <span class="st">&quot;count&quot;</span>)

 unit_metadata = 
  list(river = c(<span class="dt">SAC</span> = <span class="st">&quot;The Sacramento River&quot;</span>, <span class="dt">AM</span> = <span class="st">&quot;The American River&quot;</span>),
       spp = c(king = <span class="st">&quot;King Salmon&quot;</span>, ccho = <span class="st">&quot;Coho Salmon&quot;</span>),
       stg = c(parr = <span class="st">&quot;third life stage&quot;</span>, smolt = <span class="st">&quot;fourth life stage&quot;</span>),
       ct = <span class="st">&quot;number&quot;</span>)</code></pre>
<h3 id="dataset-2">Dataset 2</h3>
<pre class="sourceCode ruby"><code class="sourceCode ruby"> dat = data.frame(site = c(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>, <span class="st">&quot;AM&quot;</span>), 
                   species = c(<span class="st">&quot;Chinook&quot;</span>, <span class="st">&quot;Chinook&quot;</span>, <span class="st">&quot;Silver&quot;</span>), 
                   smct = c(245L, 511L, 199L),
                   pcnt =  c(290L, 408L, 212L))

 col_metadata = c(site = <span class="st">&quot;http://dbpedia.org/ontology/River&quot;</span>,
                  species = <span class="st">&quot;http://dbpedia.org/ontology/Species&quot;</span>, 
                  smct = <span class="st">&quot;Smolt count&quot;</span>,
                  pcnt = <span class="st">&quot;Parr count&quot;</span>)

 unit_metadata = 
  list(river = c(<span class="dt">SAC</span> = <span class="st">&quot;The Sacramento River&quot;</span>, <span class="dt">AM</span> = <span class="st">&quot;The American River&quot;</span>),
       spp = c(<span class="dt">Chinook</span> = <span class="st">&quot;King Salmon&quot;</span>, <span class="dt">Silver</span> = <span class="st">&quot;Coho Salmon&quot;</span>),
       smct = <span class="st">&quot;number&quot;</span>,
       pcnt = <span class="st">&quot;number&quot;</span>)</code></pre>
<h3 id="figure">Figure</h3>
<figure>
<img src="https://f.cloud.github.com/assets/222586/710800/c0d567b6-de83-11e2-94b1-52090b0c9a5f.png" alt="ontology_synthesis2" /><figcaption>ontology_synthesis2</figcaption>
</figure>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/26/starting-an-EML-package-for-R.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/24/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 24 Jun 2013</p>

<p style="font-style:italic"> pageviews: 42 </p>

<article>
<div class="excerpt">
<h2 id="exploring-xml-schema-in-r">Exploring XML Schema in R</h2>
<p>I danced around the room after reading these lines in the excellent <a href="http://www.springer.com/statistics/computational+statistics/book/978-1-4614-7899-7">XML and Web Technologies for Data Sciences with R</a> book:</p>
<blockquote>
<p>We can generate S4 class definitions corresponding to the data structures that are represented in the XML document. We can programmatically generate R code that will map an XML document adhering to this schema to corresponding R data structures. Similarly, we can automate the creation of code to serialize, or write, an R object of a particular type to a suitable XML format corresponding to the schema</p>
</blockquote>
<p>Wow. Automatically turn EML objects into R objects and vice versa! Generate a NeXML parser for R automatically. The world is your oyster now!</p>
<h2 id="diving-in">Diving in</h2>
<p>Using <code>install_github(&quot;XMLSchema&quot;, &quot;omegahat&quot;)</code>.</p>
<p>Trouble so far.</p>
<h3 id="eml">EML</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">readSchema</span>(<span class="st">&quot;eml.xsd&quot;</span>)</code></pre>
<p>From Duncan:</p>
<blockquote>
<p>As I suspected from just the description of the problem, there is a recursive circularity in the schema files - eml-documentation.xsd imports eml-text.xsd which imports eml-documentation.xsd. I’ll try to find some time to handle this.</p>
</blockquote>
<p>Staying tuned…</p>
<h3 id="nexml">NexML</h3>
<p>This is a schema for phylogenetic trees. Being able to use this schema would be a big improvement over the current approach of our treebase package, which queries a vertically integrated database that is far less expressive then the NeXML schema.</p>
<p>I tried this schema, but cannot download the file for some reason:</p>
<pre class="sourceCode r"><code class="sourceCode r">nex &lt;- <span class="kw">readSchema</span>(<span class="st">&quot;http://www.nexml.org/2009/nexml.xsd&quot;</span>)
failed to load HTTP resource
Error: <span class="dv">1</span>: failed to load HTTP resource
&gt; </code></pre>
<p>I download a copy from here and set its xsd/ directory as my working directory:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt; nex &lt;- <span class="kw">readSchema</span>(<span class="st">&quot;nexml.xsd&quot;</span>)
Error: XML content does not seem to be XML: <span class="st">&#39;./continuous.xsd&#39;</span></code></pre>
<p>because the path should be <code>characters/continuous.xsd</code>. Not sure why I’m not managing to navigate the paths correctly here.</p>
<h3 id="kml-example-from-book">KML example from book</h3>
<p>I cannot load the KML schema directly from the website:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt; kx = <span class="st">&quot;https://developers.google.com/kml/schema/kml21.xsd&quot;</span>
&gt; kml = <span class="kw">readSchema</span>(kx)
Error: XML content does not seem to be XML: <span class="st">&#39;https://developers.google.com/kml/schema/kml21.xsd&#39;</span></code></pre>
<p>Likewise,</p>
<pre class="sourceCode r"><code class="sourceCode r">kml = <span class="kw">readSchema</span>( <span class="st">&quot;http://schemas.opengis.net/kml/2.2.0/ogckml22.xsd&quot;</span>)
failed to load HTTP resource
Error: <span class="dv">1</span>: failed to load HTTP resource</code></pre>
<p>fails, though the other example works</p>
<pre class="sourceCode r"><code class="sourceCode r">pmml = <span class="kw">readSchema</span>(<span class="st">&quot;http://www.dmg.org/v4-1/pmml-4-1.xsd&quot;</span>)</code></pre>
<p>If I wget the xsd and save it locally first, then I can readSchema on the local file. If I have <code>options(error = recover)</code> set, then</p>
<pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">defineClasses</span>(kml)</code></pre>
<p>gets me deep into the debugger. If I run it with <code>options(error=NULL)</code>, then I get warnings:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt; <span class="kw">defineClasses</span>(kml)
<span class="kw">character</span>(<span class="dv">0</span>)
Warning messages:
<span class="dv">1</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.Object is an existing class
<span class="dv">2</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.Feature is an existing class
<span class="dv">3</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.LookAt is an existing class
<span class="dv">4</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.Geometry is an existing class
<span class="dv">5</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.StyleSelector is an existing class
<span class="dv">6</span>: <span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
  http<span class="er">://</span>earth.google.com/kml/<span class="fl">2.1</span>.TimePrimitive is an existing class
&gt; </code></pre>
<p>I load an minimal <a href="http://en.wikipedia.org/wiki/Keyhole_Markup_Language">example file</a> and try fromXML, but hit an error:</p>
<pre><code>&gt; doc &lt;- xmlParse(&quot;kmlexample.xml&quot;)
&gt; fromXML(doc)
Error in UseMethod(&quot;xmlAttrs&quot;, node) : 
  no applicable method for &#39;xmlAttrs&#39; applied to an object of class &quot;c(&#39;XMLInternalDocument&#39;, &#39;XMLAbstractDocument&#39;)&quot;</code></pre>
<p>Stay tuned for more from Duncan on getting this rolling..</p>
<h3 id="reading">Reading</h3>
<ul>
<li><p>Early warning signals in multivariate dynamics: <code>r citet(&quot;http://arxiv.org/abs/1306.3465&quot;, &quot;discusses&quot;)</code>. Population size grows as approaches collapse, multidemensional data needed to detect slowing down (eigenvalue estimates from multidimensional autoregressive model)</p></li>
<li><p>Little grants are more bang for the buck than big grants: <code>r citet(&quot;10.1371/journal.pone.0065263&quot;, &quot;discusses&quot;)</code></p></li>
</ul>
<h3 id="code-tricks">Code tricks</h3>
<p>From Egan, on getting a useful version of rvm on Ubuntu 12.04 (lets me install <code>bundle install</code> on the lab-notebook gemfile so I can deploy from <code>zero.ucdavis</code>, etc.)</p>
<blockquote>
<p>RVM has to be installed locally because the packaged version is broken. Before installation, the following packages need to be installed globally (unless you have root access in which case the installer will take care of them):</p>
</blockquote>
<pre><code>    bison
    libyaml-dev
    libgdbm-dev
    gawk
    sqlite</code></pre>
<blockquote>
<p>To install RVM, run:</p>
</blockquote>
<pre><code>    \curl -L https://get.rvm.io | bash -s stable --ruby --autolibs=enable --auto-dotfiles</code></pre>
<blockquote>
<p>which takes a while because it downloads and builds a lot. The installer takes care of the configuration automatically and all you need to do after installation is log into a fresh shell to source the required functions and environment.</p>
</blockquote>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/24/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

