<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-07-31T22:00:31-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-07-31T22:00:31-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-07-31T22:00:31-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
    <div class="excerpt">
      <div class="scroll">
        <ul><li>cboettig pushed to master at cboettig/labnotebook: <em>update site draft update layout tag</em> <a href="https://github.com/cboettig/labnotebook/compare/340462f1bf...be276f076c">05:05 2013/12/06</a></li><li>cboettig pushed to master at ropensci/reml: <em>eml to rdf</em> <a href="https://github.com/ropensci/reml/compare/44c81de9c9...9e55c1e9b9">11:46 2013/12/05</a></li><li>cboettig pushed to master at ropensci/reml: <em>example of how hf205.xml might be built in reml</em> <a href="https://github.com/ropensci/reml/compare/14d30700b8...44c81de9c9">11:45 2013/12/05</a></li><li>cboettig commented on issue ropensci/reml#62: <em>@emhart that would be awesome. I've just figured out how to get the text: library(RWordXML) f <- wordDoc("inst/examples/methods.docx") doc <- methods…</em> <a href="https://github.com/ropensci/reml/issues/62#issuecomment-29940701">09:37 2013/12/05</a></li><li>cboettig opened issue ropensci/reml#62: <em>Parse a .docx file to get `methods` and other text</em> <a href="https://github.com/ropensci/reml/issues/62">09:23 2013/12/05</a></li></ul>
      </div>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
      <div class="scroll">
       <ul><li><p>RT @_inundata: Excited to announce that we’re (@ropensci) writing a book on open science. ETA July 2014. 
<a href="http://t.co/TyIY1WN5aE">http://t.co/TyIY1WN5aE</a></p>
 <a href="http://twitter.com/cboettig/statuses/407577061805416448">10:28 2013/12/02</a> </li><li><p>RT @researchremix: Great read on CC0 vs CC-BY by @dancohen: <a href="http://t.co/7EeR8XdO7Q">http://t.co/7EeR8XdO7Q</a> #openaccess #opendata</p>
 <a href="http://twitter.com/cboettig/statuses/407576734993612801">10:27 2013/12/02</a> </li><li><p>RT @pebourne: Who is willing to measure the reproducibility of research in their own lab? We did in this PLOS ONE paper <a href="http://t.co/hKZIfhO%E2%80%A6">http://t.co/hKZIfhO…</a></p>
 <a href="http://twitter.com/cboettig/statuses/407576055218581504">10:24 2013/12/02</a> </li><li><p>@davidjayharris I&#39;m so glad someone actually looked at the paper links, they are far better than my babble.  That one is delightful.</p>
 <a href="http://twitter.com/cboettig/statuses/405532265054097408">07:03 2013/11/26</a> </li><li><p>RT @davidjayharris: An actual paper title, via @cboettig: Are exercises like this a good use of anybody&#39;s time? </p>

<p>Worth a careful reading h…</p>
 <a href="http://twitter.com/cboettig/statuses/405532125987758081">07:02 2013/11/26</a> </li></ul>
      </div>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <div class="scroll">
      <ul>
<li>James Wilson White, Louis W Botsford, Alan Hastings et al. 2013. Stochastic models reveal conditions for cyclic dominance in sockeye salmon populations <em>Ecological Monographs</em> <strong></strong>  <a href="http://dx.doi.org/10.1890/12-1796.1">10.1890/12-1796.1</a> </li>
<li>David Lindenmayer, Gene E. Likens. 2013. Benchmarking Open Access Science Against Good Science <em>Bulletin of the Ecological Society of America</em> <strong>94</strong> 4 <a href="http://dx.doi.org/10.1890/0012-9623-94.4.338">10.1890/0012-9623-94.4.338</a> </li>
<li>Elizabeth Eli Holmes, John L Sabo, Steven Vincent Viscido et al. 2007. A statistical approach to quasi-extinction forecasting. <em>Ecology letters</em> <strong>10</strong> 12 <a href="http://dx.doi.org/10.1111/j.1461-0248.2007.01105.x">10.1111/j.1461-0248.2007.01105.x</a> </li>
<li>Jan Esper, Ulf Büntgen, David C Frank et al. 2007. 1200 Years of Regular Outbreaks in Alpine Insects. <em>Proceedings. Biological sciences / The Royal Society</em> <strong>274</strong> 1610 <a href="http://dx.doi.org/10.1098/rspb.2006.0191">10.1098/rspb.2006.0191</a> </li>
<li>Santiago Salinas, Simon C. Brown, Marc Mangel et al. 2013. Non-genetic inheritance and changing environments <em>Non-Genetic Inheritance</em> <strong>1</strong>  <a href="http://dx.doi.org/10.2478/ngi-2013-0005">10.2478/ngi-2013-0005</a> </li>

      </div>
    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/11/19/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 19 Nov 2013</p>

<p style="font-style:italic"> pageviews: 8 </p>

<article>
<div class="excerpt">
<h3 id="rnexml">RNeXML</h3>
<ul>
<li>Feedback from Rutger, need to add <code>about</code> attributes so that RDFa abstraction references the right level of the DOM (issue <a href="https://github.com/ropensci/RNeXML/issues/35">#35</a>).<br /></li>
<li>Looking for strategy for distilling RDF from RDFa in R, see <a href="http://stackoverflow.com/questions/20026540/distill-rdfa-into-rdf-xml-in-r">my question on SO</a>. Hopefully don’t have to wrap some C library myself…</li>
</ul>
<h3 id="nonparametric-bayes">nonparametric-bayes</h3>
<p>Writing writing.</p>
<ul>
<li><p>Update pandoc templates to use yaml metadata for author, affiliation, abstract, etc. Avoids having to manually edit the <code>elsarticle.latex</code> template with this metadata. Added <a href="https://github.com/cboettig/pandoc-templates/">fork</a> for my templates, e.g. see my <a href="https://github.com/cboettig/pandoc-templates/blob/f348513558c9189ef28f0aef06fe3031d7a17d96/elsarticle.latex">elsarticle.latex</a>. Example metadata in <a href="https://github.com/cboettig/nonparametric-bayes/blob/a1801c525142d4bb64555721403d869ecfbde8d4/inst/doc/nonparametric-bayes.Rmd">manuscript</a>.</p></li>
<li><p>fixing xtable caption (as argument)</p></li>
<li><p>Extended discussion. Adjustments to figures. See commit log /diffs for details.</p></li>
</ul>
<h2 id="reading">Reading</h2>
<p><span class="showtooltip" title="Mace G (2013). &#39;Global Change: Ecology Must Evolve.&#39; Nature, 503, pp. 191-192. ISSN 0028-0836."><a href="http://dx.doi.org/10.1038/503191a">Mace (2013)</a></span> , e.g.</p>
<blockquote>
<p>a new kind of ecology is needed that is predicated on scaling up efforts, data sharing and collaboration</p>
</blockquote>
<p>hear hear.</p>
<ul>
<li><p>PNAS with a somewhat confused take on error rates, suggesting a revised threshold p-value…</p></li>
<li><p>AmNat <a href="http://w3.biosci.utexas.edu/amnatasilomar/files/AsilomarMeetingProgramNov172013.pdf">Asilomar schedule</a> (pdf) is up.</p></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/19/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/11/17/RNeXML-metadata-queries.html">Rnexml Metadata Queries</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 17 Nov 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>(From issue <a href="https://github.com/ropensci/RNeXML/issues/20">#20</a>)</p>
<h3 id="basic-metadata-queries">Basic metadata queries</h3>
<p>a question of how the user queries that metadata. Currently we have a <code>metadata</code> function that simply extracts all the metadata at the specified level (nexml, otus, trees, tree, etc) and returns a named character string in which the name corresponds to the <code>rel</code> or <code>property</code> and the value corresponds to the <code>content</code> or <code>href</code>, e.g.:</p>
<pre class="sourceCode r"><code class="sourceCode r">birds &lt;-<span class="st"> </span><span class="kw">read.nexml</span>(<span class="st">&quot;birdOrders.xml&quot;</span>)
meta &lt;-<span class="st"> </span><span class="kw">get_metadata</span>(birds) </code></pre>
<p>prints the named string with the top-level (default-level) metadata elements as so:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt;<span class="st"> </span>meta 
##                                             dc:date 
##                                        &quot;2013-11-17&quot; 
##                                          cc:license 
## &quot;http://creativecommons.org/publicdomain/zero/1.0/&quot;</code></pre>
<p>Which we can subset by name, e.g. <code>meta[&quot;dc:date&quot;]</code>. This is probably simplest to most R users; though exactly what the namespace prefix means may be unclear if they haven’t worked with namespaces before. (The user can always print a summary of the namespaces and prefixes in the nexml file using <code>birds@namespaces</code>).</p>
<p>This approach is simple, albeit a bit limited.</p>
<h3 id="xpath-queries">XPath queries</h3>
<p>For instance, the R user has a much more natural and powerful way to handle these issues of prefixes and namespaces using either the XML or rrdf libraries. For instance, if we extract meta nodes into RDF-XML, we could handle the queries like so:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">xpathSApply</span>(meta, <span class="st">&quot;//dc:title&quot;</span>, xmlValue)</code></pre>
<p>which uses the namespace prefix defined in the nexml; or</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">xpathSApply</span>(meta, <span class="st">&quot;//x:title&quot;</span>, xmlValue, <span class="dt">namespaces=</span><span class="kw">c</span>(<span class="dt">x =</span> <span class="st">&quot;http://purl.org/dc/elements/1.1/&quot;</span>))</code></pre>
<p>defining the custom prefix <code>x</code> to the URI</p>
<h3 id="sparql-queries">Sparql queries</h3>
<p>Pretty exciting that qe can make arbitrary SPARQL queries of the metadata as well.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rrdf)
<span class="kw">sparql.rdf</span>(ex, <span class="st">&quot;SELECT ?title WHERE { ?x &lt;http://purl.org/dc/elements/1.1/title&gt; ?title })</span></code></pre>
<p>Obviously the XPath or SPARQL queries are more expressive / powerful than drawing out the metadata from the S4 structure directly. On the other hand, because both of these approaches use just the distilled metadata, the original connection between metadata elements and the structure of the XML tree is lost unless stated explicitly. An in-between solution is to use XPath on the nexml XML instead, though I think we cannot make use of the namespaces in that case, since they appear in attribute values rather than structure.</p>
<p>Anyway, it’s nice to have these options in R, particularly for more complex queries where we might want to make some use of the ontology as well. On the other hand, simple presentation of basic metadata is probably necessary for most users.</p>
<p>Would be nice to illustrate with a query that required some logical deduction from the ontology.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/17/RNeXML-metadata-queries.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/11/08/MBI-day-five-notes.html">Mbi Day Five Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 08 Nov 2013</p>

<p style="font-style:italic"> pageviews: 27 </p>

<article>
<div class="excerpt">
<p>Panel discussion</p>
<ul>
<li>Hugh’s question on the usefulness of dynamic vs static models: do we have dynamical systems envy?</li>
<li>Chris: are temporal dynamics historical artefact, and space the new frontier?</li>
<li>Hugh: though decision theory is fundamentally temporal. really question of sequential decision vs single decision</li>
<li>Hugh, on what would be his priority if he had time for new question: Solve the 2 player, 2 step SDP competition closed form.<br /></li>
<li>Paul: the narrow definitions of “math biology” with 1980s flavor.<br /></li>
<li><span class="citation" data-cites="mathbiopaul">@mathbiopaul</span>: Formulating the hard problems arising in application in an appropriate abstraction that mathematicians will attack.</li>
<li>Leah raises issue of publishing software and reproducibility</li>
<li>Julia mentions Environmental modeling and software journal</li>
</ul>
<h3 id="notes-from-some-discussions-today">Notes from some discussions today</h3>
<ul>
<li><p>Sandy Leibold mentions potential for exploring early warning signals in larch budmoth data: <span class="showtooltip" title="Esper J, Buntgen U, Frank D, Nievergelt D and Liebhold A (2007). &#39;1200 Years of Regular Outbreaks in Alpine Insects.&#39; Proceedings of The Royal Society B: Biological Sciences, 274, pp. 671-679. ISSN 0962-8452."><a href="http://dx.doi.org/10.1098/rspb.2006.0191">Esper et al. (2007)</a></span> , <code>r citet(&quot;10.1111/j.1469 -8137.2009.02825.x&quot;)</code>.</p></li>
<li><p>Bill Fagan mentions paper with Eli Holmes on quasi-extinction forecasting <span class="showtooltip" title="Holmes E, Sabo J, Viscido S and Fagan W (2007). &#39;A Statistical Approach to Quasi-Extinction Forecasting.&#39; Ecology Letters, 10, pp. 1182-1198. ISSN 1461-023X."><a href="http://dx.doi.org/10.1111/j.1461-0248.2007.01105.x">Holmes et al. (2007)</a></span> and an analysis of the GPDD <span class="showtooltip" title="Fagan u, Meir u, Prendergast u, Folarin u and Karieva u (2001). &#39;Characterizing Population Vulnerability For 758 Species.&#39; Ecology Letters, 4, pp. 132-138. ISSN 1461-023X."><a href="http://dx.doi.org/10.1046/j.1461-0248.2001.00206.x">Fagan et al. (2001)</a></span> as potentially interesting applications for our Bayesian nonparametrics work.</p></li>
<li><p>Matt Potts mentions DoD program in <a href="http://www.serdp.org/">Strategic Environmental Research and Development Program (SERDP)</a>. E.g. see the current call for <a href="http://www.serdp.org/content/download/22120/228239/version/2/file/FY15_RCSON-15-01_New+Conservation+Paradigms_Final.pdf">New Paradigms for Managing Species and Ecosystems in a Non-Stationary World (pdf)</a></p></li>
</ul>
<h3 id="pdg-control">pdg-control</h3>
<p>Trying to understand pattern of increasing ENPV with increasing stochasticity. Despite having the same optimal policy inferred under increasing stochasticity (i.e. still in Reed’s self-sustaining criterion, below <span class="math">\(\sigma_g\)</span> of 0.2 or so) the average over simulated replicates is higher. We don’t seem to obtain the theoretical ENPV, but something less, in either case. See code <a href="https://github.com/cboettig/pdg_control/blob/38bccb71b08f1bbd02b7a5ca55188a739eea5109/inst/examples/noise_effects.Rmd">noise_effects.md</a>.</p>
<h3 id="ropensci">ropensci</h3>
<p>Schema.org defines a vocabulary for <a href="http://schema.org/Dataset">datasets</a> (microdata/rdfa)</p>
<p>Rutger gives a one-liner solution for tolweb to nexml using <a href="https://github.com/rvosa/bio-phylo">bio-phylo</a> perl library:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">perl</span> -MBio::Phylo::IO=parse -e <span class="st">&#39;print parse-&gt;to_xml&#39;</span> format tolweb as_project 1 url <span class="st">&#39;http://tolweb.org/onlinecontributors/app?service=external&amp;page=xml/TreeStructureService&amp;node_id=52643&#39;</span></code></pre>
<p>Hmm, there’s a journal of <a href="http://www.sciencedirect.com/science/journal/15749541">Ecological Informatics</a>.</p>
<h2 id="references">References</h2>
<ul>
<li>Rebecca S. Epanchin-Niell, Robert G. Haight, Ludek Berec, John M. Kean, Andrew M. Liebhold, Helen Regan, (2012) Optimal Surveillance And Eradication of Invasive Species in Heterogeneous Landscapes. <em>Ecology Letters</em> <strong>15</strong> 803-812 <a href="http://dx.doi.org/10.1111/j.1461-0248.2012.01800.x">10.1111/j.1461-0248.2012.01800.x</a></li>
<li>Rebecca S. Epanchin-Niell, James E. Wilen, (2012) Optimal Spatial Control of Biological Invasions. <em>Journal of Environmental Economics And Management</em> <strong>63</strong> 260-270 <a href="http://dx.doi.org/10.1016/j.jeem.2011.10.003">10.1016/j.jeem.2011.10.003</a></li>
<li>J. Esper, U. Buntgen, D. C Frank, D. Nievergelt, A. Liebhold, (2007) 1200 Years of Regular Outbreaks in Alpine Insects. <em>Proceedings of The Royal Society B: Biological Sciences</em> <strong>274</strong> 671-679 <a href="http://dx.doi.org/10.1098/rspb.2006.0191">10.1098/rspb.2006.0191</a></li>
<li>unknown Fagan, unknown Meir, unknown Prendergast, unknown Folarin, unknown Karieva, (2001) Characterizing Population Vulnerability For 758 Species. <em>Ecology Letters</em> <strong>4</strong> 132-138 <a href="http://dx.doi.org/10.1046/j.1461-0248.2001.00206.x">10.1046/j.1461-0248.2001.00206.x</a></li>
<li>A. R. Hall, A. D. Miller, H. C. Leggett, S. H. Roxburgh, A. Buckling, K. Shea, (2012) Diversity-Disturbance Relationships: Frequency And Intensity Interact. <em>Biology Letters</em> <strong>8</strong> 768-771 <a href="http://dx.doi.org/10.1098/rsbl.2012.0282">10.1098/rsbl.2012.0282</a></li>
<li>Elizabeth Eli Holmes, John L. Sabo, Steven Vincent Viscido, William Fredric Fagan, (2007) A Statistical Approach to Quasi-Extinction Forecasting. <em>Ecology Letters</em> <strong>10</strong> 1182-1198 <a href="http://dx.doi.org/10.1111/j.1461-0248.2007.01105.x">10.1111/j.1461-0248.2007.01105.x</a></li>
<li>Brian Leung, Nuria Roura-Pascual, Sven Bacher, Jaakko Heikkilä, Lluis Brotons, Mark A. Burgman, Katharina Dehnen-Schmutz, Franz Essl, Philip E. Hulme, David M. Richardson, Daniel Sol, Montserrat Vilà, Marcel Rejmanek, (2012) Teasing Apart Alien Species Risk Assessments: A Framework For Best Practices. <em>Ecology Letters</em> <strong>15</strong> 1475-1493 <a href="http://dx.doi.org/10.1111/ele.12003">10.1111/ele.12003</a></li>
<li>A. D. Miller, S. H. Roxburgh, K. Shea, (2011) How Frequency And Intensity Shape Diversity-Disturbance Relationships. <em>Proceedings of The National Academy of Sciences</em> <strong>108</strong> 5643-5648 <a href="http://dx.doi.org/10.1073/pnas.1018594108">10.1073/pnas.1018594108</a></li>
<li>Adam David Miller, Stephen H. Roxburgh, Katriona Shea, (2011) Timing of Disturbance Alters Competitive Outcomes And Mechanisms of Coexistence in an Annual Plant Model. <em>Theoretical Ecology</em> <strong>5</strong> 419-432 <a href="http://dx.doi.org/10.1007/s12080-011-0133-1">10.1007/s12080-011-0133-1</a></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/08/MBI-day-five-notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/11/07/MBI-day-four-notes.html">Mbi Day Four Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 07 Nov 2013</p>

<p style="font-style:italic"> pageviews: 17 </p>

<article>
<div class="excerpt">
<h2 id="morning-session">Morning session</h2>
<h4 id="iadine-chades">Iadine Chades</h4>
<ul>
<li><p>Chadès, I., Carwardine, J., Martin, T.G., Nicol, S., Sabbadin, R. &amp; Buffet, O. (2012) MOMDPs: a solution for modelling adaptive management problems. The Twenty-Sixth AAAI Conference on Artificial Intelligence (AAAI-12), pp. 267-273. Toronto, Canada.</p></li>
<li><p>10.1098/rspb.2013.0325 Migratory connectivity magnifies the consequences of habitat loss from sea-level rise for shorebird populations</p></li>
</ul>
<h4 id="jake-lariviera">Jake LaRiviera</h4>
<p>presents the challenges of the uncertainty table. Additional challenges in making an apples-to-apples comparison of the benefit of decreasing noise of different systems (e.g. in pricing information?)</p>
<h4 id="me">Me</h4>
<p>Some good questions following talk, primarily on BNP part.</p>
<ul>
<li>Where does the risk-adverse vs risk-prone behavior come from? Adjusting curvature of the uncertainty appropriately.<br /></li>
<li>Any lessons after stock collapsed, e.g. Rebuild a stock rather than mantain it? (Perhaps, but may face hysteresis in a way the intial collapse does not).<br /></li>
<li>Brute-force passive learning?</li>
</ul>
<h2 id="afternoon-discussion">Afternoon discussion</h2>
<ol type="1">
<li>Is an active learning approach more or less valuable in a changing environment</li>
<li>Embracing surprise: how do we actually mathematically do this.</li>
<li>Limitations due to constraints on frequency of updating. (e.g. we don’t get to change harvest, we get to set a TAC once every ten years).<br /></li>
<li>Uncertainty affecting net present value vs affecting model behavior</li>
</ol>
<!--
### Misc comments on L & L

Brian, great post here.  

To be honest, it sounds like L&L are primarily concerned about bad statistical practices rather than about open data, ecoinformatics, etc.  They speculate on a rising number of publications that make mistakes in multiple comparisons, in ignoring heterogeneity in data, in failing to have well-posed questions first.  This blog has often spoken to the issue of pervasiveness of statistical errors in the ecological literature, which is certainly neither a new problem.  As many commenters have pointed out, those ecologists not collecting their own data are perhaps not the leading culprits in making these statistical errors.  

I completely appreciate the deep knowledge of context that comes with first-hand collection of data, but at least to the extent that the context informs the conclusions, that context should be made part of the metadata.  Do we not move towards, not away, from enabling junk science when we permit conclusions to be drawn only by those who designed the experiment and collect the data?  

It is too easy to read L&L as a defense of data hugging.  You can't use my data because you cannot understand it.  If publications and citations didn't play the role they do in our academic economy, I suspect we would hear the exact same sentiments about publications themselves: if I publish my ideas, people who lack my superior understanding of context will misunderstand and misuse them!  When these contributions are rewarded, it is easier to appreciate the benefits having these ideas be seen and debated among the scientific community outweigh the risk of misintepretation.  The same might be said of data.  

Open data is often represented as benefiting the parasites (through "free" publications) at a cost to the original contributors (in "lost" publications).  If the production and citation of one's original datasets became the major metric of scientific contribution, this would reward all those collecting data while leaving the parasites without any recognized contribution - a valuation that sounds more in line with how L&L view the reliatve merits of these groups.  Given their position, it seems they would be great advocates for data publication and citation, which rewards the things they value most, and overhauling the current publication system so favorable to the existence of parasites.  


Phylogenetic data is perhaps the most pervasive example across ecological and evolutionary work where people now use sequences or phylogeny data they did not collect themselves.  And yes, the literature is rife with examples where this data has been misapplied - such as branch lengths being misinterpreted or uncertainty in branch lengths ignored.  Would fewer errors be made if we had no genbank and everyone had to collect all their own sequence data?  Yes, because most of these studies would become infeasible anyhow.  Creating a time-calibrated phylogeny from sequence data and fossils you collected all yourself might give you a greater appreciation for the uncertainty, but only to the extent that you learn the theoretical statistical underpinnings of the inference along the way.  

-->


 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/07/MBI-day-four-notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/11/06/MBI-day-three.html">Mbi Day Three</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 06 Nov 2013</p>

<p style="font-style:italic"> pageviews: 17 </p>

<article>
<div class="excerpt">
<h2 id="mbi-workshop">MBI Workshop</h2>
<h4 id="becky-epanchin-niell">Becky Epanchin-Niell</h4>
<ul>
<li><p>Motivating example of Rabies spatial control in Switzerland</p></li>
<li><p>2012 Eco Let: should New Zealand survey for bark beetle? Cost of survellience, control, and damage. <span class="showtooltip" title="Epanchin-Niell R, Haight R, Berec L, Kean J, Liebhold A and Regan H (2012). &#39;Optimal Surveillance And Eradication of Invasive Species in Heterogeneous Landscapes.&#39; Ecology Letters, 15, pp. 803-812. ."><a href="http://dx.doi.org/10.1111/j.1461-0248.2012.01800.x">Epanchin-Niell et al. (2012)</a></span></p></li>
<li><p>2012 JEEM spatial spread of star thistle: control spread and eradication as an integer programming problem in deterministic context. (with Wilen) <span class="showtooltip" title="Epanchin-Niell R and Wilen J (2012). &#39;Optimal Spatial Control of Biological Invasions.&#39; Journal of Environmental Economics And Management, 63, pp. 260-270. ISSN 00950696."><a href="http://dx.doi.org/10.1016/j.jeem.2011.10.003">Epanchin-Niell &amp; Wilen (2012)</a></span></p></li>
<li><p>Breaking landscape into individual landowners makes it less valuable to control early.</p></li>
</ul>
<h4 id="kat-shea">Kat Shea</h4>
<ul>
<li><p>IDH varying both frequency and intensity. <span class="showtooltip" title="Miller A, Roxburgh S and Shea K (2011). &#39;How Frequency And Intensity Shape Diversity-Disturbance Relationships.&#39; Proceedings of The National Academy of Sciences, 108, pp. 5643-5648. ISSN 0027-8424."><a href="http://dx.doi.org/10.1073/pnas.1018594108">Miller et al. (2011)</a></span> , experimental example in <span class="showtooltip" title="Hall A, Miller A, Leggett H, Roxburgh S, Buckling A and Shea K (2012). &#39;Diversity-Disturbance Relationships: Frequency And Intensity Interact.&#39; Biology Letters, 8, pp. 768-771. ISSN 1744-9561."><a href="http://dx.doi.org/10.1098/rsbl.2012.0282">Hall et al. (2012)</a></span></p></li>
<li>Changing of timing matters more than frequency (e.g. hit weed a key phase of life cycle) <span class="showtooltip" title="Miller A, Roxburgh S and Shea K (2011). &#39;Timing of Disturbance Alters Competitive Outcomes And Mechanisms of Coexistence in an Annual Plant Model.&#39; Theoretical Ecology, 5, pp. 419-432. ISSN 1874-1738."><a href="http://dx.doi.org/10.1007/s12080-011-0133-1">Miller et al. (2011b)</a></span> .<br /></li>
<li><p>changing autocorrelation of distrubance as a managment strategy</p></li>
</ul>
<h4 id="brian-leung">Brian Leung</h4>
<p>“Data, uncertainty and risk in biological invasions”</p>
<ul>
<li>Alien risk assessments <span class="showtooltip" title="Leung B, Roura-Pascual N, Bacher S, Heikkilä J, Brotons L, Burgman M, Dehnen-Schmutz K, Essl F, Hulme P, Richardson D, Sol D, Vilà M and Rejmanek M (2012). &#39;Teasing Apart Alien Species Risk Assessments: A Framework For Best Practices.&#39; Ecology Letters, 15, pp. 1475-1493. ."><a href="http://dx.doi.org/10.1111/ele.12003">Leung et al. (2012)</a></span> , shows a dominance of rank scoring over truely quantitative approaches. Limitations to each.<br /></li>
<li>It’s not the model complexity, but the implementation interface that poses the real barrier. Also need better integration needs.</li>
</ul>
<hr />
<h4 id="breakout-session-spatial-strategies-for-management">Breakout session: Spatial strategies for management</h4>
<p>Scott Barrett</p>
<h3 id="pdg-control">PDG Control</h3>
<ul>
<li>back-and-forth with Paul: is second column redundant? Seems not: NPV when paying penalty that doesn’t exist is: profit under penalty (<span class="math">\(\Pi_0(x,h)\)</span>) minus zero (adjustment cost), while scaling is set such that profit under penalty <em>minus adjustment cost</em>, <span class="math">\(\Pi_0(x,h) - \Pi_1(h, c_2)\)</span>. Also, better to normalize everything against the adjustment-free ENPV (<span class="math">\(Pi_0\)</span>) than to normalize by the truth/simulation model (which differs in different cases).</li>
</ul>
<p>See <a href="https://github.com/cboettig/pdg_control/commits/6e81656ce60a5c9f53bb8bd383b388266d72fcf3">commit log</a> for updated versions.</p>
<h2 id="references">References</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bibliography</span>()</code></pre>
<ul>
<li>Rebecca S. Epanchin-Niell, Robert G. Haight, Ludek Berec, John M. Kean, Andrew M. Liebhold, Helen Regan, (2012) Optimal Surveillance And Eradication of Invasive Species in Heterogeneous Landscapes. <em>Ecology Letters</em> <strong>15</strong> 803-812 <a href="http://dx.doi.org/10.1111/j.1461-0248.2012.01800.x">10.1111/j.1461-0248.2012.01800.x</a></li>
<li>Rebecca S. Epanchin-Niell, James E. Wilen, (2012) Optimal Spatial Control of Biological Invasions. <em>Journal of Environmental Economics And Management</em> <strong>63</strong> 260-270 <a href="http://dx.doi.org/10.1016/j.jeem.2011.10.003">10.1016/j.jeem.2011.10.003</a></li>
<li>A. R. Hall, A. D. Miller, H. C. Leggett, S. H. Roxburgh, A. Buckling, K. Shea, (2012) Diversity-Disturbance Relationships: Frequency And Intensity Interact. <em>Biology Letters</em> <strong>8</strong> 768-771 <a href="http://dx.doi.org/10.1098/rsbl.2012.0282">10.1098/rsbl.2012.0282</a></li>
<li>Brian Leung, Nuria Roura-Pascual, Sven Bacher, Jaakko Heikkilä, Lluis Brotons, Mark A. Burgman, Katharina Dehnen-Schmutz, Franz Essl, Philip E. Hulme, David M. Richardson, Daniel Sol, Montserrat Vilà, Marcel Rejmanek, (2012) Teasing Apart Alien Species Risk Assessments: A Framework For Best Practices. <em>Ecology Letters</em> <strong>15</strong> 1475-1493 <a href="http://dx.doi.org/10.1111/ele.12003">10.1111/ele.12003</a></li>
<li>A. D. Miller, S. H. Roxburgh, K. Shea, (2011) How Frequency And Intensity Shape Diversity-Disturbance Relationships. <em>Proceedings of The National Academy of Sciences</em> <strong>108</strong> 5643-5648 <a href="http://dx.doi.org/10.1073/pnas.1018594108">10.1073/pnas.1018594108</a></li>
<li>Adam David Miller, Stephen H. Roxburgh, Katriona Shea, (2011) Timing of Disturbance Alters Competitive Outcomes And Mechanisms of Coexistence in an Annual Plant Model. <em>Theoretical Ecology</em> <strong>5</strong> 419-432 <a href="http://dx.doi.org/10.1007/s12080-011-0133-1">10.1007/s12080-011-0133-1</a></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/06/MBI-day-three.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/11/05/MBI-day-two.html">Mbi Day Two</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 05 Nov 2013</p>

<p style="font-style:italic"> pageviews: 21 </p>

<article>
<div class="excerpt">
<h2 id="mbi-workshop">MBI Workshop</h2>
<h4 id="paul-armsworth">Paul Armsworth</h4>
<p>Very nice example of control in creating a market for ecosystem services for landowners. Key feature is that multiple land-owners respond by adjusting their prices, and so payments can be divided into fraction going into subsidy and fraction going into conservation. When one land-owner controls land that is particularly good cost per conservation accomplished, they also stand to gain largest value.</p>
<p>Also looked at auction mechanism and impact of cooperation among landowners to create hold outs.</p>
<h4 id="hugh-possingham">Hugh Possingham</h4>
<p>Hot spot assignment. Marxan and success while ignoring dynamics. {to what extent is data substitute for dynamics}</p>
<p>Dynamics (but no heterogeneity)</p>
<h4 id="bill-fagan">Bill Fagan</h4>
<p>Linking individual movements to population dynamics. Home range vs migration vs nomadic motion.</p>
<h4 id="afternoon-breakout">Afternoon breakout</h4>
<p>Looking at role of institutions and tractability of implementation problems. Interesting observation from Lou in the long-tail effect of certain individuals in explaining heterogeneity in implementation success.</p>
<hr />
<h3 id="pdg-control">PDG Control</h3>
<p>Working on table following <a href="https://github.com/cboettig/pdg_control/issues/41">issue #41</a></p>
<p>Based on <a href="https://github.com/cboettig/pdg_control/blob/5c18c1108dd7502a54bb39c1750d9be74c974ba0/inst/examples/policycosts/errors_table.Rmd">errors_table.Rmd</a>, plot from <a href="https://github.com/cboettig/pdg_control/blob/49b3faf7cd4430aa8ba18cb5a36862b5d07f1e0f/inst/examples/policycosts/plot_table.Rmd">plot_table.Rmd</a></p>
<p>shows the effect of greater noise actually reducing the impact of being wrong (either by ignoring adjustment costs that exist or assuming adjustment costs that don’t exist). Bigger induced reduction in NPV (higher cost) naturally decreases value.</p>
<p>Functional form doesn’t matter when assuming costs that don’t exist, since these are calibrated to be equivalent by selecting the coefficients in order to have equal reduction in NPV. Obviously functional form does matter when ignoring penalties that do exist, and it seems that ignoring L2 penalties is most damaging, ignoring L1 penalties the least damaging?</p>
<figure>
<img src="http://farm8.staticflickr.com/7371/10693106325_9cb0c2c8bb_o.png" alt="plot of chunk unnamed-chunk-1" /><figcaption>plot of chunk unnamed-chunk-1</figcaption>
</figure>
<TABLE border=1>
<TR> <TH>  </TH> <TH> 
penalty_fn
</TH> <TH> 
ignore_cost
</TH> <TH> 
ignore_fraction
</TH> <TH> 
assume_cost
</TH> <TH> 
assume_fraction
</TH> <TH> 
sigma_g
</TH> <TH> 
reduction
</TH>  </TR>
  <TR> <TD align="right"> 
1
</TD> <TD> 
L1
</TD> <TD align="right"> 
14536.09
</TD> <TD align="right"> 
1.00
</TD> <TD align="right"> 
16857.61
</TD> <TD align="right"> 
1.00
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.10
</TD> </TR>
    <TR> <TD align="right"> 
2
</TD> <TD> 
L2
</TD> <TD align="right"> 
11020.10
</TD> <TD align="right"> 
0.76
</TD> <TD align="right"> 
15538.44
</TD> <TD align="right"> 
0.92
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.10
</TD> </TR>
      <TR> <TD align="right"> 
3
</TD> <TD> 
fixed
</TD> <TD align="right"> 
13584.97
</TD> <TD align="right"> 
0.92
</TD> <TD align="right"> 
17582.78
</TD> <TD align="right"> 
1.05
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.10
</TD> </TR>
        <TR> <TD align="right"> 
4
</TD> <TD> 
L1
</TD> <TD align="right"> 
9273.66
</TD> <TD align="right"> 
1.09
</TD> <TD align="right"> 
17561.81
</TD> <TD align="right"> 
1.04
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.20
</TD> </TR>
          <TR> <TD align="right"> 
5
</TD> <TD> 
L2
</TD> <TD align="right"> 
11020.10
</TD> <TD align="right"> 
0.76
</TD> <TD align="right"> 
15538.44
</TD> <TD align="right"> 
0.92
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.20
</TD> </TR>
            <TR> <TD align="right"> 
6
</TD> <TD> 
fixed
</TD> <TD align="right"> 
8332.45
</TD> <TD align="right"> 
0.84
</TD> <TD align="right"> 
17401.82
</TD> <TD align="right"> 
1.03
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.20
</TD> </TR>
              <TR> <TD align="right"> 
7
</TD> <TD> 
L1
</TD> <TD align="right"> 
-641.78
</TD> <TD align="right"> 
0.21
</TD> <TD align="right"> 
15451.25
</TD> <TD align="right"> 
0.92
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.30
</TD> </TR>
                <TR> <TD align="right"> 
8
</TD> <TD> 
L2
</TD> <TD align="right"> 
-272586.11
</TD> <TD align="right"> 
-24.90
</TD> <TD align="right"> 
11567.12
</TD> <TD align="right"> 
0.69
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.30
</TD> </TR>
                  <TR> <TD align="right"> 
9
</TD> <TD> 
fixed
</TD> <TD align="right"> 
-1213.01
</TD> <TD align="right"> 
-0.46
</TD> <TD align="right"> 
15519.95
</TD> <TD align="right"> 
0.92
</TD> <TD align="right"> 
0.05
</TD> <TD align="right"> 
0.30
</TD> </TR>
                    <TR> <TD align="right"> 
10
</TD> <TD> 
L1
</TD> <TD align="right"> 
18281.97
</TD> <TD align="right"> 
1.01
</TD> <TD align="right"> 
20973.56
</TD> <TD align="right"> 
0.99
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.10
</TD> </TR>
                      <TR> <TD align="right"> 
11
</TD> <TD> 
L2
</TD> <TD align="right"> 
12462.51
</TD> <TD align="right"> 
0.72
</TD> <TD align="right"> 
19392.85
</TD> <TD align="right"> 
0.91
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.10
</TD> </TR>
                        <TR> <TD align="right"> 
12
</TD> <TD> 
fixed
</TD> <TD align="right"> 
16691.94
</TD> <TD align="right"> 
1.03
</TD> <TD align="right"> 
20143.80
</TD> <TD align="right"> 
0.95
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.10
</TD> </TR>
                          <TR> <TD align="right"> 
13
</TD> <TD> 
L1
</TD> <TD align="right"> 
12788.95
</TD> <TD align="right"> 
0.96
</TD> <TD align="right"> 
20663.21
</TD> <TD align="right"> 
0.97
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.20
</TD> </TR>
                            <TR> <TD align="right"> 
14
</TD> <TD> 
L2
</TD> <TD align="right"> 
12462.51
</TD> <TD align="right"> 
0.72
</TD> <TD align="right"> 
19392.85
</TD> <TD align="right"> 
0.91
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.20
</TD> </TR>
                              <TR> <TD align="right"> 
15
</TD> <TD> 
fixed
</TD> <TD align="right"> 
12399.01
</TD> <TD align="right"> 
0.88
</TD> <TD align="right"> 
21289.21
</TD> <TD align="right"> 
1.00
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.20
</TD> </TR>
                                <TR> <TD align="right"> 
16
</TD> <TD> 
L1
</TD> <TD align="right"> 
9099.48
</TD> <TD align="right"> 
0.73
</TD> <TD align="right"> 
19999.07
</TD> <TD align="right"> 
0.94
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                  <TR> <TD align="right"> 
17
</TD> <TD> 
L2
</TD> <TD align="right"> 
-6337.76
</TD> <TD align="right"> 
-0.39
</TD> <TD align="right"> 
17765.33
</TD> <TD align="right"> 
0.83
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                    <TR> <TD align="right"> 
18
</TD> <TD> 
fixed
</TD> <TD align="right"> 
7399.01
</TD> <TD align="right"> 
0.62
</TD> <TD align="right"> 
21969.84
</TD> <TD align="right"> 
1.03
</TD> <TD align="right"> 
0.20
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                      <TR> <TD align="right"> 
19
</TD> <TD> 
L1
</TD> <TD align="right"> 
30227.43
</TD> <TD align="right"> 
1.07
</TD> <TD align="right"> 
31377.07
</TD> <TD align="right"> 
0.94
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.10
</TD> </TR>
                                        <TR> <TD align="right"> 
20
</TD> <TD> 
L2
</TD> <TD align="right"> 
33472.18
</TD> <TD align="right"> 
1.00
</TD> <TD align="right"> 
33472.18
</TD> <TD align="right"> 
1.00
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.10
</TD> </TR>
                                          <TR> <TD align="right"> 
21
</TD> <TD> 
fixed
</TD> <TD align="right"> 
28926.73
</TD> <TD align="right"> 
1.09
</TD> <TD align="right"> 
30500.85
</TD> <TD align="right"> 
0.91
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.10
</TD> </TR>
                                            <TR> <TD align="right"> 
22
</TD> <TD> 
L1
</TD> <TD align="right"> 
24157.93
</TD> <TD align="right"> 
1.10
</TD> <TD align="right"> 
31767.92
</TD> <TD align="right"> 
0.95
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.20
</TD> </TR>
                                              <TR> <TD align="right"> 
23
</TD> <TD> 
L2
</TD> <TD align="right"> 
19091.31
</TD> <TD align="right"> 
0.93
</TD> <TD align="right"> 
29225.46
</TD> <TD align="right"> 
0.87
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.20
</TD> </TR>
                                                <TR> <TD align="right"> 
24
</TD> <TD> 
fixed
</TD> <TD align="right"> 
23573.19
</TD> <TD align="right"> 
1.02
</TD> <TD align="right"> 
30069.80
</TD> <TD align="right"> 
0.90
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.20
</TD> </TR>
                                                  <TR> <TD align="right"> 
25
</TD> <TD> 
L1
</TD> <TD align="right"> 
21086.12
</TD> <TD align="right"> 
1.12
</TD> <TD align="right"> 
32035.93
</TD> <TD align="right"> 
0.96
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                                    <TR> <TD align="right"> 
26
</TD> <TD> 
L2
</TD> <TD align="right"> 
19091.31
</TD> <TD align="right"> 
0.93
</TD> <TD align="right"> 
29225.46
</TD> <TD align="right"> 
0.87
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                                      <TR> <TD align="right"> 
27
</TD> <TD> 
fixed
</TD> <TD align="right"> 
17209.56
</TD> <TD align="right"> 
0.85
</TD> <TD align="right"> 
32946.24
</TD> <TD align="right"> 
0.98
</TD> <TD align="right"> 
0.50
</TD> <TD align="right"> 
0.30
</TD> </TR>
                                                         </TABLE>


<hr />
<h2 id="ropensci-ecoinformatics">rOpenSci / ecoinformatics</h2>
<ul>
<li>Exploring strategies for addressing certificate authentication workflow.<br /></li>
<li>Plans to merge <code>dataone</code> and <code>rdataone</code>, shedding the current <code>rJava</code> dependency and dealing with existing vs new namespacing <a href="https://github.com/ropensci/rdataone/issues/1">#1</a></li>
</ul>
<h2 id="phylogenetics">phylogenetics</h2>
<ul>
<li>Some discussion of <span class="showtooltip" title="Quintero I, Wiens J and Harmon L (2013). &#39;Rates of Projected Climate Change Dramatically Exceed Past Rates of Climatic Niche Evolution Among Vertebrate Species.&#39; Ecology Letters, 16, pp. 1095-1103. ."><a href="http://dx.doi.org/10.1111/ele.12144">Quintero et al. (2013)</a></span></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/05/MBI-day-two.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/11/04/MBI-notes.html">Mbi Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 04 Nov 2013</p>

<p style="font-style:italic"> pageviews: 20 </p>

<article>
<div class="excerpt">
<h2 id="introductions">Introductions</h2>
<h2 id="hastings">Hastings</h2>
<h2 id="lou-gross">Lou Gross</h2>
<ul>
<li>Big picture: big data, computational challenges.<br /></li>
<li>NEON, DataONE, ADW, iDigBio</li>
<li>“Convergence” as the new interdisciplinary (National Academies)</li>
<li>Rise of synthesis centers</li>
<li>“Enabling architecture for next gen life science research” – National Academies report Lou Gross (2013)</li>
<li><p>Comp Science for Natural Resource Management – Fuller, Wang, Gross (2007)</p></li>
<li><p>language barriers: What’s a model? Mouse, drosophila? logistic, ricker? GIS map layers? …</p></li>
</ul>
<p>Contraints frequently dominate, not the control or the state equation.</p>
<p>Let stakeholders make their own rankings. Scenario analysis vs optimal control. Uncertainies! “Relative assessment protocol” Fuller, Gross, Duke-Sylvester, Palmer. “Testing the robustness of management descions under uncertainty” (ATLSS modeling)</p>
<h3 id="breakouts">Breakouts</h3>
<p>notes from my breakout session:</p>
<h4 id="questions">Questions</h4>
<ul>
<li>Sensitivity analysis of Scenario Rankings</li>
<li>What does a Resilience approach add</li>
<li>Generalities</li>
</ul>
<h4 id="tools-for-decision-under-uncertainty">Tools for decision under uncertainty</h4>
<ul>
<li>optimization</li>
<li>threshold planning</li>
<li>scenario planning</li>
<li>resilience thinking</li>
</ul>
<p>(Polasky et al 2011 TREE)</p>
<h4 id="optimization">optimization</h4>
<ul>
<li>Info gap “theory”</li>
<li>Satisfiability / mini-max</li>
<li>model approximation methods</li>
<li>dynamic programming</li>
</ul>
<h4 id="leading-challenges">Leading Challenges</h4>
<ul>
<li>To what extent are these approaches different sides of the same coin?</li>
<li>Are there truly non-optimization based approaches?</li>
<li>Almost-optimal approaches</li>
<li>Including constraints</li>
</ul>
<h4 id="what-we-do-well">what we do well</h4>
<ul>
<li>Optimize easy problems</li>
<li>open loop</li>
</ul>
<h4 id="state-of-the-art">State-of-the-art</h4>
<ul>
<li>Starting to: simulate optimal solutions to simple problems under more realistic circumstances</li>
<li>Starting to find multiple “optima”</li>
<li>feedback control (SDP)</li>
<li>Dual-control / adaptive control / active adaptive management / learning</li>
<li>large state space</li>
</ul>
<h4 id="open-challenges-what-we-do-poorly">Open challenges / what we do poorly</h4>
<ul>
<li>Dual control under parameter uncertainty (without restrictive assumptions on parameters)</li>
<li>high-dimensional problems</li>
<li>multiple stake-holders / game-theory solutions (outside fisheries)</li>
<li>mapping between control and implementation (partial controllability)</li>
<li>large action space</li>
<li>(multiple) delayed effect actions</li>
</ul>
<h4 id="open-challenges-multiple-stake-holder-games">Open challenges: Multiple stake-holder games</h4>
<ul>
<li>beyond 2-player differential games (with feedback)</li>
<li>simultaneous player actions</li>
</ul>
<h4 id="open-challenges-adaptive-management-timescales">Open challenges: adaptive management timescales</h4>
<ul>
<li>frequency of revisiting decisions</li>
<li>biological timescales</li>
<li>political timescales</li>
</ul>
<h4 id="known-nuisances">Known nuisances</h4>
<ul>
<li>curse of dimensionality</li>
<li>data collection methods</li>
<li>numerical methods</li>
<li>local vs global</li>
</ul>
<h4 id="missed-things">Missed things</h4>
<ul>
<li>spatial data, using rich data under the curse of dimensionality</li>
</ul>
<hr />
<h2 id="ropensci">ropensci</h2>
<p>Writing out proof-of-principle interface to the dataone REST API, see <a href="https://github.com/ropensci/rdataone/">rdataone</a> and <a href="https://github.com/ropensci/rdataone/issues/1">Introduction to the package</a></p>
<p>Key things:</p>
<ul>
<li>We can accomplish handling of certificates from httr, just add <code>config = list(sslcert = )</code>; see <code>?httr::config</code>, e.g. archive a file with:</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r">httr::<span class="kw">PUT</span>(<span class="kw">paste0</span>(<span class="st">&quot;https://knb.ecoinformatics.org/knb/d1/mn/v1/archive/&quot;</span>, 
                         <span class="st">&quot;urn:uuid:0bf9cf50-6b6b-45ad-b22c-ffd0ee2a21c3&quot;</span>), 
                                        <span class="dt">config=</span><span class="kw">config</span>(<span class="dt">sslcert =</span> <span class="st">&quot;/tmp/x509up_u1000&quot;</span>))</code></pre>
<p>Posting new data requires writing a system metadata XML file. Currently have a crude minimal version of this, <a href="https://github.com/ropensci/rdataone/blob/6b0a8e9d53768c4e739179923681a053c13552a8/R/write_sysmeta.R">write_sysmeta.R</a>, should see how <code>dataone</code> package is handling this.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/11/04/MBI-notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/10/17/RNeXML-semantic-considerations.html">Rnexml Semantic Considerations</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 17 Oct 2013</p>

<p style="font-style:italic"> pageviews: 18 </p>

<article>
<div class="excerpt">
<p>Some very good feedback from Hilmar as I tackle some of the semantic capabilities of NeXML in RNeXML. As the complete discussion is already archived in the Github issues tracking, (see in particular <a href="https://github.com/ropensci/RNeXML/issues/26">#26</a> and <a href="https://github.com/ropensci/RNeXML/issues/24">#24</a>) I only paraphrase here.</p>
<p>One of the central advantages we can offer in programmatic generation of NeXML in the R environment is the ability to validate names and enhance the metadata included in the resulting nexml file using programmatic queries to taxonomic name resolution services such as ITIS, EOL, NCBI, as provided in the <a href="https://github.com/ropensci/taxize_">taxize</a> package.</p>
<p>A subtly to this approach is discussed in issue <a href="https://github.com/ropensci/RNeXML/issues/24">#24</a>. Whenever we provide new data, we should also provide future users the appropriate metadata describing where it came from, if we had found an exact match or maybe only a close match (perhaps an alternative spelling of the species name was used). Specifying the provenance exactly can become quite verbose, for each taxonomic unit we have:</p>
<figure>
<img src="https://raw.github.com/phylotastic/ontologies/master/tnrs/tnrs-instance-1otu.png" />
</figure>
<p>which corresponds to adding RDFa to the NeXML looking something like:</p>
<pre class="sourceCode xml"><code class="sourceCode xml"><span class="kw">&lt;otus</span><span class="ot"> id=</span><span class="st">&quot;tax1&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;otu</span><span class="ot"> label=</span><span class="st">&quot;Struthioniformes&quot;</span><span class="ot"> id=</span><span class="st">&quot;t1&quot;</span><span class="kw">&gt;</span>
      <span class="kw">&lt;meta</span><span class="ot"> xmlns:rdf=</span><span class="st">&quot;http://www.w3.org/1999/02/22-rdf-syntax-ns#&quot;</span>
<span class="ot">     xmlns:rdfs=</span><span class="st">&quot;http://www.w3.org/2000/01/rdf-schema#&quot;</span>
<span class="ot">     xmlns=</span><span class="st">&quot;http://www.w3.org/1999/xhtml&quot;</span>
<span class="ot">     xmlns:obo=</span><span class="st">&quot;http://purl.obolibrary.org/obo/&quot;</span>
<span class="ot">     xmlns:tc=</span><span class="st">&quot;http://rs.tdwg.org/ontology/voc/TaxonConcept#&quot;</span>
<span class="ot">     xmlns:dc=</span><span class="st">&quot;http://purl.org/dc/elements/1.1/&quot;</span>
<span class="ot">     xmlns:dcterms=</span><span class="st">&quot;http://purl.org/dc/terms/&quot;</span>
<span class="ot">     xmlns:tnrs=</span><span class="st">&quot;http://phylotastic.org/terms/tnrs.rdf#&quot;</span>
<span class="ot">     xmlns:xsd=</span><span class="st">&quot;http://www.w3.org/2001/XMLSchema#&quot;</span>
<span class="ot">     class=</span><span class="st">&quot;rdf2rdfa&quot;</span><span class="kw">&gt;</span>
   <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> about=</span><span class="st">&quot;http://phylotastic.org/terms/tnrs-instance.rdf#otu5&quot;</span>
<span class="ot">        typeof=</span><span class="st">&quot;obo:CDAO_0000138&quot;</span><span class="kw">&gt;</span>
      <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;rdfs:label&quot;</span><span class="ot"> content=</span><span class="st">&quot;Panthera tigris HQ263408&quot;</span><span class="kw">/&gt;</span>
      <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:resolvesAs&quot;</span><span class="kw">&gt;</span>
         <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> typeof=</span><span class="st">&quot;tnrs:NameResolution&quot;</span><span class="kw">&gt;</span>
            <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:matchCount&quot;</span><span class="ot"> content=</span><span class="st">&quot;2&quot;</span><span class="kw">/&gt;</span>
            <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:matches&quot;</span><span class="kw">&gt;</span>
               <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> typeof=</span><span class="st">&quot;tnrs:Match&quot;</span><span class="kw">&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:acceptedName&quot;</span><span class="ot"> content=</span><span class="st">&quot;Panthera tigris&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:matchedName&quot;</span><span class="ot"> content=</span><span class="st">&quot;Panthera tigris&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:score&quot;</span><span class="ot"> content=</span><span class="st">&quot;1.0&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tc:toTaxon&quot;</span><span class="ot"> resource=</span><span class="st">&quot;http://www.ncbi.nlm.nih.gov/taxonomy/9694&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:usedSource&quot;</span><span class="kw">&gt;</span>
                     <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> about=</span><span class="st">&quot;http://www.ncbi.nlm.nih.gov/taxonomy&quot;</span>
<span class="ot">                          typeof=</span><span class="st">&quot;tnrs:ResolutionSource&quot;</span><span class="kw">&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;dc:description&quot;</span><span class="ot"> content=</span><span class="st">&quot;NCBI Taxonomy&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:hasRank&quot;</span><span class="ot"> content=</span><span class="st">&quot;3&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:sourceStatus&quot;</span><span class="ot"> content=</span><span class="st">&quot;200: OK&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;dc:title&quot;</span><span class="ot"> content=</span><span class="st">&quot;NCBI&quot;</span><span class="kw">/&gt;</span>
                     <span class="kw">&lt;/meta&gt;</span>
                  <span class="kw">&lt;/meta&gt;</span>
               <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:matches&quot;</span><span class="kw">&gt;</span>
               <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> typeof=</span><span class="st">&quot;tnrs:Match&quot;</span><span class="kw">&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:acceptedName&quot;</span><span class="ot"> content=</span><span class="st">&quot;Megalachne&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:matchedName&quot;</span><span class="ot"> content=</span><span class="st">&quot;Pantathera&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:score&quot;</span><span class="ot"> content=</span><span class="st">&quot;0.47790686999749&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tc:toTaxon&quot;</span><span class="ot"> resource=</span><span class="st">&quot;http://www.tropicos.org/Name/40015658&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:usedSource&quot;</span><span class="kw">&gt;</span>
                     <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span><span class="ot"> about=</span><span class="st">&quot;http://tnrs.iplantcollaborative.org/&quot;</span>
<span class="ot">                          typeof=</span><span class="st">&quot;tnrs:ResolutionSource&quot;</span><span class="kw">&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;dc:description&quot;</span>
<span class="ot">                             content=</span><span class="st">&quot;The iPlant Collaborative TNRS provides parsing and fuzzy matching for plant taxa.&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:hasRank&quot;</span><span class="ot"> content=</span><span class="st">&quot;2&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:sourceStatus&quot;</span><span class="ot"> content=</span><span class="st">&quot;200: OK&quot;</span><span class="kw">/&gt;</span>
                        <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;dc:title&quot;</span><span class="ot"> content=</span><span class="st">&quot;iPlant Collaborative TNRS v3.0&quot;</span><span class="kw">/&gt;</span>
                     <span class="kw">&lt;/meta&gt;</span>
                  <span class="kw">&lt;/meta&gt;</span>
               <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;dcterms:source&quot;</span><span class="kw">&gt;</span>
               <span class="kw">&lt;meta</span><span class="ot"> class=</span><span class="st">&quot;description&quot;</span>
<span class="ot">                    about=</span><span class="st">&quot;http://phylotastic.org/terms/tnrs-instance.rdf#request&quot;</span>
<span class="ot">                    typeof=</span><span class="st">&quot;tnrs:ResolutionRequest&quot;</span><span class="kw">&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:submitDate&quot;</span><span class="ot"> content=</span><span class="st">&quot;Mon Jun 11 20:25:16 2012&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:usedSource&quot;</span><span class="ot"> resource=</span><span class="st">&quot;http://tnrs.iplantcollaborative.org/&quot;</span><span class="kw">/&gt;</span>
                  <span class="kw">&lt;meta</span><span class="ot"> rel=</span><span class="st">&quot;tnrs:usedSource&quot;</span><span class="ot"> resource=</span><span class="st">&quot;http://www.ncbi.nlm.nih.gov/taxonomy&quot;</span><span class="kw">/&gt;</span>
               <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;/meta&gt;</span>
            <span class="kw">&lt;meta</span><span class="ot"> property=</span><span class="st">&quot;tnrs:submittedName&quot;</span><span class="ot"> content=</span><span class="st">&quot;Panthera tigris&quot;</span><span class="kw">/&gt;</span>
         <span class="kw">&lt;/meta&gt;</span>
      <span class="kw">&lt;/meta&gt;</span>
   <span class="kw">&lt;/meta&gt;</span>
<span class="kw">&lt;/meta&gt;</span>
<span class="kw">&lt;/otu&gt;</span></code></pre>
<p>Why provenance? This can be particularly important in tracking down errors or inconsistencies. For instance, image the taxanomic barcode id number we provide for a taxon is later divided into multiple species. Ideally this would be reflected in the updated entries of the barcode service, establishing new id numbers for the split members and identifying that the old id was split – after all, a barcode system is supposed to facilitate addressing these kinds of issues.</p>
<p>Still, this appears much more verbose than say, what TreeBase does when automatically adding identifiers as annotations to the NeXML otu nodes.</p>
<p>Meanwhile, in more concrete terms, we seem to have some consensus on using NCBI taxonomic ids, since its API queries are pretty fast:</p>
<pre class="sourceCode xml"><code class="sourceCode xml"><span class="kw">&lt;otus</span><span class="ot"> id=</span><span class="st">&quot;tax1&quot;</span><span class="kw">&gt;</span>
    <span class="kw">&lt;otu</span><span class="ot"> label=</span><span class="st">&quot;Struthioniformes&quot;</span><span class="ot"> id=</span><span class="st">&quot;t1&quot;</span><span class="kw">&gt;</span>
      <span class="kw">&lt;meta</span><span class="ot"> xsi:type=</span><span class="st">&quot;ResourceMeta&quot;</span><span class="ot"> href=</span><span class="st">&quot;http://ncbi.nlm.nih.gov/taxonomy/8798&quot;</span><span class="ot"> rel=</span><span class="st">&quot;tc:toTaxon&quot;</span><span class="kw">/&gt;</span>
    <span class="kw">&lt;/otu&gt;</span></code></pre>
<p>note this example uses the <code>tc: http://rs.tdwg.org/ontology/voc/TaxonConcept#</code> <code>toTaxon</code> concept to provide a ontological definition of the link as a taxon identier.</p>
<p>NCBI does not do partial matching, so we simply warn when a user’s taxonomic names do not match an NCBI id, giving them a chance to correct them if in error (either manually or automatically using the partial name matching functions in <code>taxize</code>)</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/10/17/RNeXML-semantic-considerations.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/10/11/is-it-time-to-retire-pagels-lambda.html">Is it time to retire Pagel's lambda?</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 11 Oct 2013</p>

<p style="font-style:italic"> pageviews: 482 </p>

<article>
<div class="excerpt">
<p>Pagel’s <span class="math">\(\lambda\)</span> (lambda), introduced in <a href="http://doi.org/10.1038/44766" title="Inferring the historical patterns of biological evolution. in Nature.">Pagel 1999</a> as a potential measure of “phylogenetic signal,” the extent to which correlations in traits reflect their shared evolutionary history (as approximated by Brownian motion).</p>
<p>Numerous critiques and ready alternatives have not appeared to decrease it’s popularity. There are many issues with the statistic, some of which I attempt to summarise below.</p>
<p>The <span class="math">\(\lambda\)</span> statistic is defined by the Brownian motion model together with a transformation of the branch lengths: multiply all internal branches by <span class="math">\(\lambda\)</span>. The motivation for the definition is obvious: <span class="math">\(\lambda = 1\)</span> the tree is unchanged and the model equivalent to Brownian motion, while for <span class="math">\(\lambda = 0\)</span> the tree becomes a star phylogeny and the model is equivalent to completely independent random walks. <span class="math">\(0 &lt; \lambda &lt; 1\)</span> provides an intermediate range where the correlations are weaker than expected.</p>
<p><strong>Problem 1: It is biological nonsense to treat tips different from other edges.</strong></p>
<p>All other problems arise from this. While it is okay that a statistic does not have a corresponding evolutionary model, being part of an explicit model might have helped avoid this sillyness. Technically <span class="math">\(\lambda\)</span> is a model, but one that treats evolution along “tips” as special, as if evolution should follow completely different rules for a species alive today relative to it’s former evolutionary history. Sounds almost creationist.</p>
<p><strong>Problem 2: The statistic doesn’t measure what is says it measures.</strong></p>
<p>To demonstrate this, we can consider two cases in which phylogeny has the identical effect of explaining trait correlations, and yet have very different lambdas. Consider that Researcher 1 examines the phylogeny in Figure 1 and estimates very little phylogenetic signal, <span class="math">\(\lambda = 0.1\)</span>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ape)
<span class="kw">cat</span>(<span class="st">&quot;(((A_sp:10,B_sp:10):1,C_sp:11):1,D_sp:12);&quot;</span>, <span class="dt">file =</span> <span class="st">&quot;ex.tre&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
ex &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="st">&quot;ex.tre&quot;</span>)
<span class="kw">plot</span>(ex)</code></pre>
<figure>
<img src="http://farm4.staticflickr.com/3708/10715190003_f2f21044be_o.png" alt="Figure 1" /><figcaption>Figure 1</figcaption>
</figure>
<p>Now Researcher 2 discovers closely related sister species of some of the taxa originally studied, as in Figure 2.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="st">&quot;((((A_sp:1, A2_sp:1):9,(B_sp:1, B2_sp:1):9):1,(C_sp:1, C2_sp:1):10):1,(D_sp:1, D2_sp:1):11);&quot;</span>, 
    <span class="dt">file =</span> <span class="st">&quot;ex2.tre&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>)
ex2 &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="st">&quot;ex2.tre&quot;</span>)
<span class="kw">plot</span>(ex2)</code></pre>
<figure>
<img src="http://farm6.staticflickr.com/5482/10715001046_f914f6ecee_o.png" alt="Figure 2" /><figcaption>Figure 2</figcaption>
</figure>
<p>There traits of sister taxa are very similar (indeed let us assume the sister species are hard to distinguish morphologically - perhaps why they were overlooked by Researcher 1). The OU or BM model estimates made by researcher 1 will closely agree with with those of Researcher 1, since the sister taxa have quite similar traits. Yet the <span class="math">\(\lambda\)</span> estimates differs greatly – all of a sudden the phylogenetic signal must be quite high!</p>
<p>And yet the underlying evolutionary process by which we have simulated the data has been unchanged! The difference arises because what formerly appeared as long tips have become short tips. How do we intepret a metric that depends so heavily on whether or not all sister species are present in the data? As noted, this problem does not impact other phylogenetic comparative methods to nearly the same extent.</p>
<p><strong>Problem 3</strong> The statistic has no notion of timescale or depth in the phylogeny.</p>
<p>In <span class="math">\(\lambda\)</span> (and other definitions such as Blomberg’s K), phylogenetic signal is an all-or-nothing proposition. If we find that really recently diverged species that happen to resemble each-other, while species that have diverged for longer than, say, a couple million years show no correlation – is this phylogenetic signal or not? This ‘extinction’ of phylogenetic signal as we go far enough back in time seems like a biologically reasonable concept that is perfectly well expressed in the <span class="math">\(\alpha\)</span> parameter of the OU model, but is lost in the consideration of <span class="math">\(\lambda\)</span>. If folks really want to estimate a continuous quantity to measure phylogenetic signal, I suggest <span class="math">\(\alpha\)</span> is a far more meaningful number (note that it has units! (1/time or 1/branch length)).</p>
<p>Consider the returning force alpha in the OU model (i.e. stabilizing selection). When alpha is near zero, the model is essentially Brownian, (i.e. ‘strong phylogenetic signal,’ where more recently diverged species are more similar on average than distantly related ones). When alpha is very large, traits reflect the selective constraint of the environment rather than their history, and so recently diverged species are no more or less likely to be similar than distant ones (provided all species in question are under the same OU model / same selection strength for the trait in question). The size of alpha gives the timescale over which ‘phylogenetic signal’ is lost (in units of the branch length). Two very recently diverged sister-taxa may thus show some phylogenetic correlation because their divergence time is of order 1/alpha, while those with longer divergence times act phylogeneticly independent, such as in our Figure 2 above. I find this an imperfect but reasonable meaning of phylogenetic signal.</p>
<p>If we restrict <span class="math">\(\lambda\)</span> to be strictly 1 or 0 these problems are alliviated, though then it is unnessary to define the statistic as such as we may instead consider a star tree (sometimes called the “white noise” model of evolution).</p>
<h4 id="other-such-statistics">other such statistics</h4>
<p>Pagel’s <span class="math">\(\delta\)</span> is a transformation on node depth, which is again problematic as there is no meaningfully consistent way to describe what is a node (think about deep speciation events with no present day ancestor.) I believe <span class="math">\(\kappa\)</span> would also be problematic to interpret as it is a nonlinear transform of branch length – raises branch length to a power – and thus would have a rather different effect depending on the units in which branch length were measured. (For instance, consider the case where the tree is scaled to length unity, so all branch lengths are less than one and thus become shorter with large exponents, vs one in which lengths are all larger than one). Fortunately these statistics are far less popular than <span class="math">\(\lambda\)</span>.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/10/11/is-it-time-to-retire-pagels-lambda.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

