<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-04-12T08:53:44-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-04-12T08:53:44-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-04-12T08:53:44-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
     <div class="excerpt">
        <ul><li>cboettig pushed to master at cboettig/labnotebook: <em>drafting drafting drafting</em> <a href="https://github.com/cboettig/labnotebook/compare/b40dc9394e...e2aeea034b">08:14 2013/07/16</a></li><li>cboettig pushed to master at cboettig/sandbox: <em>calc Moore neighborhood in R; answers @DrCraigMc query</em> <a href="https://github.com/cboettig/sandbox/compare/786ae0df27...160695821b">08:36 2013/07/15</a></li><li>cboettig pushed to master at cboettig/knitcitations: <em>roll back to original doi pattern #45</em> <a href="https://github.com/cboettig/knitcitations/compare/6650b5e056...7ba14c7bf2">03:08 2013/07/12</a></li><li>cboettig commented on issue ropensci/RMendeley#13: <em>@dfalster Glad you were able figure things out despite the dated documentation, sorry for the trouble. We'll correct it sometime -- and a la @yihui …</em> <a href="https://github.com/ropensci/RMendeley/issues/13#issuecomment-20855747">02:55 2013/07/12</a></li><li>cboettig pushed to master at cboettig/labnotebook: <em>drafting</em> <a href="https://github.com/cboettig/labnotebook/compare/e49bd74905...b40dc9394e">03:38 2013/07/11</a></li></ul>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
       <ul><li><p>@waltonjones sometimes, e.g. see notes: <a href="http://t.co/FcugrRfHSY">http://t.co/FcugrRfHSY</a></p>
 <a href="http://twitter.com/cboettig/statuses/357324484618891264">07:22 2013/07/16</a> </li><li><p>RT @rOpenSci: We&#39;re super excited to announce our newest board member MacKenzie Smith - University Librarian at @ucdavis - <a href="http://t.co/pTKZ%E2%80%A6">http://t.co/pTKZ…</a></p>
 <a href="http://twitter.com/cboettig/statuses/357213455725510656">12:01 2013/07/16</a> </li><li><p>@waltonjones actually I build the html from the md locally by running Jekyll, then push the html to github (bc I have custom plugins)</p>
 <a href="http://twitter.com/cboettig/statuses/357173698106638337">09:23 2013/07/16</a> </li><li><p>Upcoming @NSF deadline: Advances in Bioinformatics <a href="http://t.co/N6ZJwQjm4M">http://t.co/N6ZJwQjm4M</a> Hoping folks will submit #ecoinformatics ideas! #notjustATCG</p>
 <a href="http://twitter.com/cboettig/statuses/356870051447578628">01:17 2013/07/15</a> </li><li><p>@kaythaney @recology_ awe thanks!</p>
 <a href="http://twitter.com/cboettig/statuses/356869139324866561">01:13 2013/07/15</a> </li></ul>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <ul><li>Early warning signals: the charted and uncharted territories: Theoretical Ecology (2013). Carl Boettiger, Noam Ross, Alan Hastings et al. <a href="http://www.mendeley.com/research/early-warning-signals-charted-uncharted-territories-1/">12:32 2013/06/21</a></li><li>Theory of early warning signals of disease emergenceand leading indicators of elimination: Theoretical Ecology (2013). Suzanne M. O’Regan, John M Drake et al. <a href="http://www.mendeley.com/research/theory-early-warning-signals-disease-emergenceand-leading-indicators-elimination/">12:32 2013/06/21</a></li><li>Complex decisions made simple: a primer on stochastic dynamic programming: Methods in Ecology and Evolution (2013). Pages: n/a-n/a. Lucile Marescot, Guillaume Chapron, Iadine Chadès, Paul L. Fackler, Christophe Duchamp, Eric Marboutin, Olivier Gimenez et al. <a href="http://www.mendeley.com/research/complex-decisions-made-simple-primer-stochastic-dynamic-programming/">12:31 2013/06/21</a></li><li>Stochastic multiplicative population growth predicts and interprets Taylor's power law of fluctuation scaling.: Proceedings. Biological sciences / The Royal Society (2013). Volume: 280, Issue: 1757. Pages: 20122955. Joel E Cohen, Meng Xu, William S F Schuster et al. <a href="http://www.mendeley.com/research/stochastic-multiplicative-population-growth-predicts-interprets-taylors-power-law-fluctuation-scalin-2/">12:31 2013/06/21</a></li></ul>
    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/07/11/visit-to-WHOI.html">Visit To Whoi</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 11 Jul 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>Presented my work from the <a href="/tags/#nonparametric-bayes">nonparametric-bayes</a> project, <em>Avoiding tipping points in the management of ecological systems: a non-parametric Bayesian approach</em> at the <a href="http://www.whoi.edu/">Woods Hole Oceangraphic Institution</a> as part of their weekly Biology Department Seminar Series.</p>
<ul>
<li>9 - 9:30 <a href="http://www.jbkellner.com/">Julie Kellner</a></li>
<li>9:30-10:15 Ben Jones, (Kellner group)</li>
<li>10:15 - 11:00 Math ecology group</li>
<li>11:00 - 11:30 <a href="http://www.whoi.edu/hpb/Site.do?id=14853">Ian Carroll</a></li>
<li>11:30-12 Seminar prep</li>
<li>12-1 my seminar</li>
<li>1-2 Lunch with the Marine Policy Center (<a href="http://www.whoi.edu/oceanus/viewArticle.do?id=3992">Mike Fogarty</a>, <a href="http://www.whoi.edu/hpb/Site.do?id=1692">Hauke Kite-Powell</a>, <a href="http://www.whoi.edu/hpb/Site.do?id=172">Di Jin</a>), Julie Kellner</li>
</ul>
<p>Enjoyable visit, with some particularly great questions after talk: including challenges/opportunities in multi-dimensional examples, concerns about temporal heterogeneity, partially observed systems / working with catch data in place of stock assessment data. Less push-back than I anticipated on using the Gaussian process; probably aided by the SDP context. Quite broad interest in SDP in particular too, estimating MSY following regression models seems common. Interesting discussion about the challenge introduced by today’s emphasis on ecosystem management; how to still retain focus on tractable models.</p>
<p>Also some good discussion on the rOpenSci goals of large-scale data access and integration.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/11/visit-to-WHOI.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/10/forking-the-R-journal.html">Forking The R Journal</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 10 Jul 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>All I really wanted to do is cite a paper in <em>the R Journal</em>, which is peer-reviewed, open access (CC-BY), LaTeX based and without author charges. Sure, I could do that already, but I like being able to programmatically generate citation metadata from a URL – we do have this thing called the internet now, and citations are just links, right? Unfortunately, nice as it is, the R journal doesn’t have HTML landing pages for articles that embed the metadata.</p>
<p>This makes it harder for Google Scholar to index the articles, and means that we cannot extract citation metadata from the URL using a tool like <a href="http://greycite.knowledgeblog.org">greycite</a>. Until now.</p>
<p>I wrote to Editor-in-Chief Hadley Wickam about this, who responded in the best way possible: making the <a href="ihttps://github.com/rjournal/rjournal.github.io">journal website’s Github repository</a> public. A fork and a little hacking later, and voila, we have html landing pages for <em>the R Journal</em> with embedded metadata (pending a <a href="https://github.com/rjournal/rjournal.github.io/pull/1">pull request</a>). Check out <a href="https://github.com/cboettig/rjournal.github.io/commit/e70e84e6e53e6c04ec9864af162d7ba58439d4d5">the source code</a> for how this works – it’s really quite straight forward since the metadata is already available in <code>_config.yaml</code>. The main step involves a <a href="https://github.com/cboettig/rjournal.github.io/blob/e70e84e6e53e6c04ec9864af162d7ba58439d4d5/_plugins/article_html_pages.rb">Generator plugin</a> which builds a page for each article and makes the relevant article metadata available to the page. Then we can write a <a href="https://github.com/cboettig/rjournal.github.io/blob/e70e84e6e53e6c04ec9864af162d7ba58439d4d5/_layouts/article.html">page template</a> using Liquid code to import the metadata.</p>
<p>The really exciting thing about this is the basic idea of forking a journal website and improving it. There’s a lot we could do to improve the R journal. I think that ideally we’d have HTML5 versions of the articles as well, something that would be straight forward if authors used knitr of course, but I realize that’s a bigger shift. One could also automatically enhance the HTML with quite a bit more semantic content, we could have animations, links the Rnw files, etc etc. I’d be happy to help and I’m sure others would too. R-journal is great but it could be so much more as an example and test-bed of innovation in statistical publishing.</p>
<p>Well, now that the journal is open for pull requests, hack away!</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/10/forking-the-R-journal.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/26/starting-an-EML-package-for-R.html">Starting An Eml Package For R</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 26 Jun 2013</p>

<p style="font-style:italic"> pageviews: 26 </p>

<article>
<div class="excerpt">
<p>Started <a href="https://github.com/ropensci/reml"><code>reml</code></a>, and R package for reading, writing, manipulating and publishing EML files. More details and progress so far on the Github page; in particular flushing out design goals in the <a href="https://github.com/ropensci/reml/issues">issues tracker</a>.</p>
<h3 id="initial-goals">Initial goals</h3>
<p>To start, I am aiming for a proof of principle case in which we can write an R <code>data.frame</code> into EML (<a href="https://github.com/ropensci/reml/issues/2">issue #2</a>), publish to a remote repository such as figshare (<a href="https://github.com/ropensci/reml/issues/3">issue #3</a>), and then read the data back in as the original R object types from the EML (<a href="https://github.com/ropensci/reml/issues/6">issue #6</a>). This will require specifying some generic metadata and basic column and unit metadata for the data frame. It would be good to get feedback on what way is most convenient to specify the required metadata in R.</p>
<h3 id="integration-goals-issue-8">Integration goals (issue <a href="https://github.com/ropensci/reml/issues/8">#8</a>)</h3>
<p>This is the <strong>holy grail</strong> of metadata infrastructure and ostensibly the primary purpose of EML, see <a href="http://dx.doi.org/10.1146/annurev.ecolsys.37.091305.110031" title="The New Bioinformatics: Integrating Ecological Data from the Gene to the Biosphere in Annual Review of Ecology, Evolution, and Systematics">Jones et al 2006</a>. Despite that, integration is not actually possible without semantic definitions as well, see <a href="http://dx.doi.org/10.1016/j.tree.2011.11.016" title="Ecoinformatics: supporting ecology as a data-intensive science; in Trends in Ecology &amp; Evolution. doi: 10.1016/j.tree.2011.11.016">Michener &amp; Jones 2012</a>, from which we adapt this minimal example below.</p>
<p>This example provides minimal and sometimes missing semantics; which may make it unresolvable. A complete semantic solution is diagrammed in the figure from <a href="http://dx.doi.org/10.1016/j.tree.2011.11.016" title="Ecoinformatics: supporting ecology as a data-intensive science; in Trends in Ecology &amp; Evolution. doi: 10.1016/j.tree.2011.11.016">Michener &amp; Jones 2012</a>.</p>
<h3 id="dataset-1">Dataset 1</h3>
<pre class="sourceCode ruby"><code class="sourceCode ruby"> dat = data.frame(river=c(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>), 
                   spp = c(<span class="st">&quot;king&quot;</span>, <span class="st">&quot;king&quot;</span>, <span class="st">&quot;ccho&quot;</span>), 
                   stg = c(<span class="st">&quot;smolt&quot;</span>, <span class="st">&quot;parr&quot;</span>, <span class="st">&quot;smolt&quot;</span>),
                   ct =  c(293L, 410L, 210L))

 col_metadata = c(river = <span class="st">&quot;http://dbpedia.org/ontology/River&quot;</span>,
                  spp = <span class="st">&quot;http://dbpedia.org/ontology/Species&quot;</span>, 
                  stg = <span class="st">&quot;Life history stage&quot;</span>,
                  ct = <span class="st">&quot;count&quot;</span>)

 unit_metadata = 
  list(river = c(<span class="dt">SAC</span> = <span class="st">&quot;The Sacramento River&quot;</span>, <span class="dt">AM</span> = <span class="st">&quot;The American River&quot;</span>),
       spp = c(king = <span class="st">&quot;King Salmon&quot;</span>, ccho = <span class="st">&quot;Coho Salmon&quot;</span>),
       stg = c(parr = <span class="st">&quot;third life stage&quot;</span>, smolt = <span class="st">&quot;fourth life stage&quot;</span>),
       ct = <span class="st">&quot;number&quot;</span>)</code></pre>
<h3 id="dataset-2">Dataset 2</h3>
<pre class="sourceCode ruby"><code class="sourceCode ruby"> dat = data.frame(site = c(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>, <span class="st">&quot;AM&quot;</span>), 
                   species = c(<span class="st">&quot;Chinook&quot;</span>, <span class="st">&quot;Chinook&quot;</span>, <span class="st">&quot;Silver&quot;</span>), 
                   smct = c(245L, 511L, 199L),
                   pcnt =  c(290L, 408L, 212L))

 col_metadata = c(site = <span class="st">&quot;http://dbpedia.org/ontology/River&quot;</span>,
                  species = <span class="st">&quot;http://dbpedia.org/ontology/Species&quot;</span>, 
                  smct = <span class="st">&quot;Smolt count&quot;</span>,
                  pcnt = <span class="st">&quot;Parr count&quot;</span>)

 unit_metadata = 
  list(river = c(<span class="dt">SAC</span> = <span class="st">&quot;The Sacramento River&quot;</span>, <span class="dt">AM</span> = <span class="st">&quot;The American River&quot;</span>),
       spp = c(<span class="dt">Chinook</span> = <span class="st">&quot;King Salmon&quot;</span>, <span class="dt">Silver</span> = <span class="st">&quot;Coho Salmon&quot;</span>),
       smct = <span class="st">&quot;number&quot;</span>,
       pcnt = <span class="st">&quot;number&quot;</span>)</code></pre>
<h3 id="figure">Figure</h3>
<figure>
<img src="https://f.cloud.github.com/assets/222586/710800/c0d567b6-de83-11e2-94b1-52090b0c9a5f.png" alt="ontology_synthesis2" /><figcaption>ontology_synthesis2</figcaption>
</figure>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/26/starting-an-EML-package-for-R.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/06/24/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 24 Jun 2013</p>

<p style="font-style:italic"> pageviews: 25 </p>

<article>
<div class="excerpt">
<h2 id="exploring-xml-schema-in-r">Exploring XML Schema in R</h2>
<p>I danced around the room after reading these lines in the excellent <a href="http://www.springer.com/statistics/computational+statistics/book/978-1-4614-7899-7">XML and Web Technologies for Data Sciences with R</a> book:</p>
<blockquote>
<p>We can generate S4 class definitions corresponding to the data structures that are represented in the XML document. We can programmatically generate R code that will map an XML document adhering to this schema to corresponding R data structures. Similarly, we can automate the creation of code to serialize, or write, an R object of a particular type to a suitable XML format corresponding to the schema</p>
</blockquote>
<p>Wow. Automatically turn EML objects into R objects and vice versa! Generate a NeXML parser for R automatically. The world is your oyster now!</p>
<h2 id="diving-in">Diving in</h2>
<p>Using <code>install_github(&quot;XMLSchema&quot;, &quot;omegahat&quot;)</code>.</p>
<p>Trouble so far.</p>
<h3 id="eml">EML</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">readSchema</span>(<span class="st">&quot;eml.xsd&quot;</span>)</code></pre>
<p>From Duncan:</p>
<blockquote>
<p>As I suspected from just the description of the problem, there is a recursive circularity in the schema files - eml-documentation.xsd imports eml-text.xsd which imports eml-documentation.xsd. I’ll try to find some time to handle this.</p>
</blockquote>
<p>Staying tuned…</p>
<h3 id="nexml">NexML</h3>
<p>This is a schema for phylogenetic trees. Being able to use this schema would be a big improvement over the current approach of our treebase package, which queries a vertically integrated database that is far less expressive then the NeXML schema.</p>
<p>I tried this schema, but cannot download the file for some reason:</p>
<pre class="sourceCode r"><code class="sourceCode r">nex &lt;-<span class="st"> </span><span class="kw">readSchema</span>(<span class="st">&quot;http://www.nexml.org/2009/nexml.xsd&quot;</span>)
failed to load HTTP resource
Error:<span class="st"> </span><span class="dv">1</span>:<span class="st"> </span>failed to load HTTP resource
&gt;<span class="st"> </span></code></pre>
<p>I download a copy from here and set its xsd/ directory as my working directory:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt;<span class="st"> </span>nex &lt;-<span class="st"> </span><span class="kw">readSchema</span>(<span class="st">&quot;nexml.xsd&quot;</span>)
Error:<span class="st"> </span>XML content does not seem to be XML:<span class="st"> &#39;./continuous.xsd&#39;</span></code></pre>
<p>because the path should be <code>characters/continuous.xsd</code>. Not sure why I’m not managing to navigate the paths correctly here.</p>
<h3 id="kml-example-from-book">KML example from book</h3>
<p>I cannot load the KML schema directly from the website:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt;<span class="st"> </span>kx =<span class="st"> &quot;https://developers.google.com/kml/schema/kml21.xsd&quot;</span>
&gt;<span class="st"> </span>kml =<span class="st"> </span><span class="kw">readSchema</span>(kx)
Error:<span class="st"> </span>XML content does not seem to be XML:<span class="st"> &#39;https://developers.google.com/kml/schema/kml21.xsd&#39;</span></code></pre>
<p>Likewise,</p>
<pre class="sourceCode r"><code class="sourceCode r">kml =<span class="st"> </span><span class="kw">readSchema</span>( <span class="st">&quot;http://schemas.opengis.net/kml/2.2.0/ogckml22.xsd&quot;</span>)
failed to load HTTP resource
Error:<span class="st"> </span><span class="dv">1</span>:<span class="st"> </span>failed to load HTTP resource</code></pre>
<p>fails, though the other example works</p>
<pre class="sourceCode r"><code class="sourceCode r">pmml =<span class="st"> </span><span class="kw">readSchema</span>(<span class="st">&quot;http://www.dmg.org/v4-1/pmml-4-1.xsd&quot;</span>)</code></pre>
<p>If I wget the xsd and save it locally first, then I can readSchema on the local file. If I have <code>options(error = recover)</code> set, then</p>
<pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">defineClasses</span>(kml)</code></pre>
<p>gets me deep into the debugger. If I run it with <code>options(error=NULL)</code>, then I get warnings:</p>
<pre class="sourceCode r"><code class="sourceCode r">&gt;<span class="st"> </span><span class="kw">defineClasses</span>(kml)
<span class="kw">character</span>(<span class="dv">0</span>)
Warning messages:
<span class="dv">1</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.Object is an existing class
<span class="dv">2</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.Feature is an existing class
<span class="dv">3</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.LookAt is an existing class
<span class="dv">4</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.Geometry is an existing class
<span class="dv">5</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.StyleSelector is an existing class
<span class="dv">6</span>:<span class="st"> </span><span class="kw">In</span> (function (className, baseType, <span class="dt">where =</span> <span class="kw">globalenv</span>())  :
<span class="st">  </span>http:<span class="er">//</span>earth.google.com/kml/<span class="fl">2.1</span>.TimePrimitive is an existing class
&gt;<span class="st"> </span></code></pre>
<p>I load an minimal <a href="http://en.wikipedia.org/wiki/Keyhole_Markup_Language">example file</a> and try fromXML, but hit an error:</p>
<pre><code>&gt; doc &lt;- xmlParse(&quot;kmlexample.xml&quot;)
&gt; fromXML(doc)
Error in UseMethod(&quot;xmlAttrs&quot;, node) : 
  no applicable method for &#39;xmlAttrs&#39; applied to an object of class &quot;c(&#39;XMLInternalDocument&#39;, &#39;XMLAbstractDocument&#39;)&quot;</code></pre>
<p>Stay tuned for more from Duncan on getting this rolling..</p>
<h3 id="reading">Reading</h3>
<ul>
<li><p>Early warning signals in multivariate dynamics: <code>r citet(&quot;http://arxiv.org/abs/1306.3465&quot;, &quot;discusses&quot;)</code>. Population size grows as approaches collapse, multidemensional data needed to detect slowing down (eigenvalue estimates from multidimensional autoregressive model)</p></li>
<li><p>Little grants are more bang for the buck than big grants: <code>r citet(&quot;10.1371/journal.pone.0065263&quot;, &quot;discusses&quot;)</code></p></li>
</ul>
<h3 id="code-tricks">Code tricks</h3>
<p>From Egan, on getting a useful version of rvm on Ubuntu 12.04 (lets me install <code>bundle install</code> on the lab-notebook gemfile so I can deploy from <code>zero.ucdavis</code>, etc.)</p>
<blockquote>
<p>RVM has to be installed locally because the packaged version is broken. Before installation, the following packages need to be installed globally (unless you have root access in which case the installer will take care of them):</p>
</blockquote>
<pre><code>    bison
    libyaml-dev
    libgdbm-dev
    gawk
    sqlite</code></pre>
<blockquote>
<p>To install RVM, run:</p>
</blockquote>
<pre><code>    \curl -L https://get.rvm.io | bash -s stable --ruby --autolibs=enable --auto-dotfiles</code></pre>
<blockquote>
<p>which takes a while because it downloads and builds a lot. The installer takes care of the configuration automatically and all you need to do after installation is log into a fresh shell to source the required functions and environment.</p>
</blockquote>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/24/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/23/notes-on-leveraging-the-ecological-markup-language.html">Notes On Leveraging The Ecological Markup Language</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 23 Jun 2013</p>

<p style="font-style:italic"> pageviews: 19 </p>

<article>
<div class="excerpt">
<p>This post creates new tag <a href="/tags.html#eml">eml</a> for posts discussing the Ecological Markup Langauge or the implementation of an R parser thereof (<a href="https://github.com/ropensci/reml">reml</a>).</p>
<h3 id="why-eml">Why EML?</h3>
<p><a href="http://10.1146/annurev.ecolsys.37.091305.110031">Jones et al.(2005)</a> give an excellent explanation describing the relative advantage of the descriptive metadata approach over vertically integrated repositories (such as Genbank, TreeBase, or many other repostories that have been the focus of rOpenSci development to date).</p>
<p>In short, the EML approach allows archiving of very heterogeneous data without having to standardize everything into a narrow and pre-defined syntax.</p>
<h3 id="providers-of-eml">Providers of EML</h3>
<p>As far as I can tell, the long-term ecological research program sites are responsible for almost all available EML files. XML files in specificed schemas involve strict criteria and are thus best generated by software.</p>
<h3 id="use-cases-of-eml">Use cases of EML</h3>
<p>These sites seem to do a better job generating this data than making any use of it – a <a href="http://dx.doi.org/10.1007/978-90-481-8782-9_5">review of the project</a> mentions “a common metadata” standard in passing and without citation. While providing nice examples of the breadth of ecological data as published by a range of researchers and journals, it provides little obvious example of synthesis. I just reread Jones et al. review to confirm that it provides no examples of papers involving data synthesis based on EML (instead citing a single example of manual data synthesis to illustrate the challenges involved).</p>
<h3 id="generating-eml">Generating EML</h3>
<p><a href="http://dx.doi.org/10.1109/4236.957896">Metacat</a> a GPL program written in Java by EML authors Matt and co, describes itself as a database that uses EML to store data. It might be a useful tool for generating EML, but looks pretty intimidating to me. It probably provides the “<a href="http://knb.ecoinformatics.org/knb/cgi-bin/register-dataset.cgi?cfg=knb">Web Software Interface</a>” that provides rather minimal EML</p>
<p><a href="http://knb.ecoinformatics.org/morphoportal.jsp">Morpho</a> is aimed at documenting species trait data, and looks like a rather useful if tedious tool for generating EML files. Unfortunately, without the ability to script inputs or automatically detect existing data structures, we are forced through the rather arduous process of adding all metadata annotation each time.</p>
<h2 id="what-can-a-package-offer">What can a package offer?</h2>
<p>Translating xpath commands into R function wrappers probably provides little utility. Do we really need a <code>get_authors</code> function in place of <code>getNodeSet(doc, &quot;//creator&quot;)</code>? Of course we might want to convert the returned nodes into characters strings or R <code>person</code> objects. Or providing larger chunks of metadata extracted into a single R object, or text-based summary (e.g. markdown or yaml).</p>
<p>A package could also provide utilities to generate EML from R objects, leveraging the metadata implicit in R objects that is not present in a CSV (in which there is no built-in notion of whether a column is numeric or character string, what missing value characters it uses, or really if it is consistent at all. Avoiding manual specification of these things makes the metadata annotation less tedious as well.</p>
<p>Of course such ideas are just the beginning. Ideally a package could help tackle the grand challenge of heterogeneous data integration, though this might have to wait for more semantics than are natively found EML.</p>
<ul>
<li>Annual Review discussing “ecoinformatics” and the EML approach <span class="showtooltip" title="Jones M, Schildhauer M, Reichman O and Bowers S (2006). 'The New Bioinformatics: Integrating Ecological Data From The Gene to The Biosphere.' Annual Review of Ecology, Evolution, And Systematics, 37, pp. 519-544. ISSN 1543-592X."><a href="http://dx.doi.org/10.1146/annurev.ecolsys.37.091305.110031" rel="http://purl.org/spar/cito/discusses" >Jones et al. (2006)</a></span> .</li>
<li>The use of ontologies in ecology <span class="showtooltip" title="Madin J, Bowers S, Schildhauer M and Jones M (2008). 'Advancing Ecological Research With Ontologies.' Trends in Ecology &amp; Evolution, 23, pp. 159-168. ISSN 01695347."><a href="http://dx.doi.org/10.1016/j.tree.2007.11.007" rel="http://purl.org/spar/cito/usesMethodIn" >Madin et al. (2008)</a></span><br /></li>
<li>Updated, shorter review in TREE <span class="showtooltip" title="Michener W and Jones M (2012). 'Ecoinformatics: Supporting Ecology as A Data-Intensive Science.' Trends in Ecology &amp; Evolution, 27, pp. 85-93. ISSN 01695347."><a href="http://dx.doi.org/10.1016/j.tree.2011.11.016" rel="http://purl.org/spar/cito/discusses" >Michener &amp; Jones (2012)</a></span></li>
<li>Simple EML introduction in <em>Frontiers</em> <span class="showtooltip" title="Fegraus E, Andelman S, Jones M and Schildhauer M (2005). 'Maximizing The Value of Ecological Data With Structured Metadata: an Introduction to Ecological Metadata Language (Eml) And Principles For Metadata Creation.' Bulletin of The Ecological Society of America, 86, pp. 158-168. ISSN 0012-9623."><a href="http://dx.doi.org/10.1890/0012-9623(2005)86[158:MTVOED]2.0.CO;2" rel="http://purl.org/spar/cito/discusses" >Fegraus et al. (2005)</a></span></li>
<li>metacat paper, <span class="showtooltip" title="Jones M, Berkley C, Bojilova J and Schildhauer M (2001). 'Managing Scientific Metadata.' Ieee Internet Computing, 5, pp. 59-68. ISSN 10897801."><a href="http://dx.doi.org/10.1109/4236.957896" rel="http://purl.org/spar/cito/discusses" >Jones et al. (2001)</a></span></li>
<li>LTER paper, <span class="showtooltip" title="Müller F, Baessler C, Schubert H, Klotz S, Gosz J, Waide R and Magnuson J (unknown). 'Long-Term Ecological Research.' Unknown."><a href="" rel="http://purl.org/spar/cito/discusses" >Müller et al. (unknown)</a></span></li>
</ul>
<h2 id="references">References</h2>
<ul>
<li>Eric H. Fegraus, Sandy Andelman, Matthew B. Jones, Mark Schildhauer, (2005) Maximizing The Value of Ecological Data With Structured Metadata: an Introduction to Ecological Metadata Language (Eml) And Principles For Metadata Creation. <em>Bulletin of The Ecological Society of America</em> <strong>86</strong> 158-168 <a href="http://dx.doi.org/10.1890/0012-9623(2005)86[158:MTVOED]2.0.CO;2">10.1890/0012-9623(2005)86[158:MTVOED]2.0.CO;2</a></li>
<li>M.B. Jones, C. Berkley, J. Bojilova, M. Schildhauer, (2001) Managing Scientific Metadata. <em>Ieee Internet Computing</em> <strong>5</strong> 59-68 <a href="http://dx.doi.org/10.1109/4236.957896">10.1109/4236.957896</a></li>
<li>Matthew B. Jones, Mark P. Schildhauer, O.J. Reichman, Shawn Bowers, (2006) The New Bioinformatics: Integrating Ecological Data From The Gene to The Biosphere. <em>Annual Review of Ecology, Evolution, And Systematics</em> <strong>37</strong> 519-544 <a href="http://dx.doi.org/10.1146/annurev.ecolsys.37.091305.110031">10.1146/annurev.ecolsys.37.091305.110031</a></li>
<li>Joshua S. Madin, Shawn Bowers, Mark P. Schildhauer, Matthew B. Jones, (2008) Advancing Ecological Research With Ontologies. <em>Trends in Ecology &amp; Evolution</em> <strong>23</strong> 159-168 <a href="http://dx.doi.org/10.1016/j.tree.2007.11.007">10.1016/j.tree.2007.11.007</a></li>
<li>William K. Michener, Matthew B. Jones, (2012) Ecoinformatics: Supporting Ecology as A Data-Intensive Science. <em>Trends in Ecology &amp; Evolution</em> <strong>27</strong> 85-93 <a href="http://dx.doi.org/10.1016/j.tree.2011.11.016">10.1016/j.tree.2011.11.016</a></li>
<li>Felix Müller, Cornelia Baessler, Hendrik Schubert, Stefan Klotz, James R. Gosz, Robert B. Waide, John J. Magnuson, (unknown) Long-Term Ecological Research. <em>Unknown</em></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/23/notes-on-leveraging-the-ecological-markup-language.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/18/figure-for-summarizing-simulations.html">Figure For Summarizing Simulations</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 18 Jun 2013</p>

<p style="font-style:italic"> pageviews: 73 </p>

<article>
<div class="excerpt">
<p>Trying to find a better way to summarize outputs of simulations,</p>
<figure>
<img src="http://farm6.staticflickr.com/5448/9021275460_cbc5de534d_o.png" />
</figure>
<p>Heavy over-plotting makes the simulation trajectories themselves hard to read, even with transparent curves. Instead, we can plot the cumulative profit (technically should be discounted first):</p>
<figure>
<img src="http://farm6.staticflickr.com/5334/9019048845_1e55f47129_o.png" />
</figure>
<p>We can normalize these profits as a fraction of the theoretical optimum (true), providing a comparison that is less dependent on model parameters.</p>
<figure>
<img src="http://farm4.staticflickr.com/3672/9093236637_273d3f3974_o.png" />
</figure>
<p>We can then attempt to plot these on a single plot, either as stacked histograms:</p>
<figure>
<img src="http://farm6.staticflickr.com/5531/9093236713_f4737d525e_o.png" />
</figure>
<p>or stacked density plots:</p>
<figure>
<img src="http://farm6.staticflickr.com/5328/9093236773_168c3a160b_o.png" />
</figure>
<p>Even after (or because of) some tuning of binwidth I’m not sure I’m happy. Will probably go with the histogram. The density plot fails when distribution is purely a delta function, giving a uniform distribution instead (and is thus maximally wrong); this can be avoided by adding a small jitter.</p>
<ul>
<li>Code: see complete code <a href="https://github.com/cboettig/nonparametric-bayes/blob/fc63651f1b1ef44b0b53817d35351a11997c6ce8/inst/examples/BUGS/external-chunks.R">nonparametric-bayes/fc6365</a></li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r">## @knitr Figure3
<span class="kw">ggplot</span>(sims_data) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(time, fishstock, <span class="dt">group=</span><span class="kw">interaction</span>(reps,method), <span class="dt">color=</span>method), <span class="dt">alpha=</span>.<span class="dv">1</span>) +
<span class="st">    </span><span class="kw">scale_colour_manual</span>(<span class="dt">values=</span>colorkey, <span class="dt">guide =</span> <span class="kw">guide_legend</span>(<span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">alpha =</span> <span class="dv">1</span>)))


## @knitr profits
Profit &lt;-<span class="st"> </span>sims_data[, <span class="kw">sum</span>(profit), by=<span class="kw">c</span>(<span class="st">&quot;reps&quot;</span>, <span class="st">&quot;method&quot;</span>)]
tmp &lt;-<span class="st"> </span><span class="kw">dcast</span>(Profit, reps ~<span class="st"> </span>method)
<span class="co">#tmp$Allen &lt;- tmp[,&quot;Allen&quot;] + rnorm(dim(tmp)[1], 0, 1) # jitter for plotting</span>
tmp &lt;-<span class="st"> </span>tmp /<span class="st"> </span>tmp[,<span class="st">&quot;True&quot;</span>]
tmp &lt;-<span class="st"> </span><span class="kw">melt</span>(tmp[<span class="dv">2</span>:<span class="kw">dim</span>(tmp)[<span class="dv">2</span>]])
actual_over_optimal &lt;-<span class="kw">subset</span>(tmp, variable !=<span class="st"> &quot;True&quot;</span>)


## @knitr Figure4 
<span class="kw">ggplot</span>(actual_over_optimal, <span class="kw">aes</span>(value)) +<span class="st"> </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>variable)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">facet_wrap</span>(~variable, <span class="dt">scales =</span> <span class="st">&quot;free_y&quot;</span>)  +<span class="st"> </span><span class="kw">guides</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>) +
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Total profit by replicate&quot;</span>) +<span class="st"> </span><span class="kw">scale_fill_manual</span>(<span class="dt">values=</span>colorkey)

<span class="kw">ggplot</span>(actual_over_optimal, <span class="kw">aes</span>(value)) +<span class="st"> </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">fill=</span>variable), <span class="dt">binwidth=</span><span class="fl">0.1</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Total profit by replicate&quot;</span>)+<span class="st"> </span><span class="kw">scale_fill_manual</span>(<span class="dt">values=</span>colorkey)

<span class="kw">ggplot</span>(actual_over_optimal, <span class="kw">aes</span>(value, <span class="dt">fill=</span>variable, <span class="dt">color=</span>variable)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">stat_density</span>(<span class="kw">aes</span>(<span class="dt">y=</span>..density..), <span class="dt">position=</span><span class="st">&quot;stack&quot;</span>, <span class="dt">adjust=</span><span class="dv">3</span>, <span class="dt">alpha=</span>.<span class="dv">9</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">&quot;Total profit by replicate&quot;</span>)+<span class="st"> </span><span class="kw">scale_fill_manual</span>(<span class="dt">values=</span>colorkey)+<span class="st"> </span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span>colorkey)</code></pre>
<h2 id="misc">Misc</h2>
<ul>
<li><p>Via Noam. <a href="http://snikolov.wordpress.com/2012/11/14/early-detection-of-twitter-trends/">Interesting</a></p></li>
<li><p>Richard interview</p></li>
<li><p><a href="https://github.com/scholmd/scholmd/wiki/Todo-list-from-workshop">scholarly markdown</a></p></li>
<li><p>Exchange with Martin on <a href="http://blog.martinfenner.org/2013/06/17/what-is-scholarly-markdown/#comment-934745570">links are citations</a> <strong>Edit</strong>: it appears comments have been lost, but see discussion on the follow-up post, <a href="blog.martinfenner.org/2013/06/19/citations-in-scholarly-markdown/">citations in scholarly markdown</a>.</p></li>
<li><p>Wow, the scaling unit <code>em</code> in LaTeX, CSS, etc. actually comes from the width of the character <code>M</code>, dating back to printing-press era, when that character was used to define the spacing <a href="http://tex.stackexchange.com/questions/119068/meaning-of-quad?newsletter=1&amp;nlcode=50792%7ca76d">quadratone</a></p></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/18/figure-for-summarizing-simulations.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/06/13/what-I-look-for-in-software-papers.html">What I look for in 'Software Papers'</a></h4></header>
<p><span>pet peeves and faux pas</span></p>
<p style="font-style:italic"> 13 Jun 2013</p>

<p style="font-style:italic"> pageviews: 649 </p>

<article>
<div class="excerpt">
<p>I am more and more frequently reviewing ‘software papers:’ which I define as publications whose primary purpose is to publicize a piece of scientific software and provide a traditional research product with hopes that it will receive citations and recognition from other researchers in grant and job reviews. To me this feels very much like hacking the publication recognition system rather than the ideal way to recognize and track the role of software in research communities, but a very practical one in the current climate. I have written two myself, so I have been on both ends of this issue. In this post, I share what I look for in such papers and what omissions I see most frequently.</p>
<h2 id="reviewing-software">Reviewing software</h2>
<p>If we are going to employ this hack of the publication system for software, we should at least use it to maximal advantage. As a reviewer, I feel that means reviewing not just the submitted manuscript but the software itself. If we can agree on nothing else, we as a community should at least be able to say:</p>
<blockquote>
<p>my review of a software paper is a review of the software</p>
</blockquote>
<p>I assume most other authors, reviewers and editors on this content share this implicit assumption, but I’d love to hear first hand from anyone else. For instance: as an editor, what would you do if your reviewers only commented on the paper directly and not on the software distributed?</p>
<p>We are not really taught to review software, any more than we are taught to write it in the first place. Most journals offer little guidance on this (though see the Journal of Open Research Software <a href="http://openresearchsoftware.metajnl.com/peer-review/">guidelines for peer review of software</a>, all though they are still rather minimal.) In the absence of a culture on software reviewing, I thought I would lay out my own perspective with the hope of hearing feedback and push back from others. Perhaps through such dialogs we can develop clearer expectations for this rapidly expanding genre.</p>
<h2 id="reviewing-the-software-paper">Reviewing the software paper</h2>
<p>I don’t include “to document” the software as a purpose, since none do so very comprehensively, and besides, documentation belongs in the software, not in a journal. “Publicize” usually includes some motivating examples that could convince many readers that the software does something useful for them without too much effort. As such, I expect the paper to:</p>
<ul>
<li>provide the journal’s audience with a clear motivation for why the package is useful, * and have at least one functioning “wow” example that I can run (by copy-paste) and understand without difficulty (e.g. without referring to code comments or the package manual to understand the function calls and their arguments).</li>
</ul>
<p>This is an intentionally low bar that I hope helps promote these kinds of contributions. Despite this, papers frequently fail the copy-paste test or the plain text explanations of the function calls. Come on people. Meanwhile, I try to focus the rest of my review on the software itself.</p>
<h2 id="my-partial-list-of-criteria">My partial list of criteria</h2>
<p>As I am almost always reviewing R packages, the software already meets some very basic standards required by submission to CRAN: dependencies and license stated, built-in documentation, passing some minimal automatic checks, etc. (See the <a href="http://cran.r-project.org/web/packages/policies.html">CRAN Policies</a> and the <a href="http://cran.r-project.org/doc/manuals/R-exts.html">Writing R Extensions Manual</a> for details). This is great, as it clears the first few hurdles of installation, etc. without much fuss, but still provides a bar that is by itself unacceptably low for published scientific software. Here is a list of the things I see that most often frustrate me. This isn’t intended as a style-guide or a comprehensive list of best practices, just my own pet peeves. I have somewhat tongue-in-cheek labeled them by severity of the review I might give; which like any other use of these terms is more of a measure of how annoyed I am then anything else. Critiques and suggestions welcome.</p>
<h3 id="edit">Edit:</h3>
<p>The comments from other reviewers, authors, and editors have been fantastic, thank you all. I particularly appreciate the opportunity to have reviewing styles critiqued, something that does not happen in normal peer review.</p>
<p>Just a note on my headings here. I do not see any of these things as “gatekeeping requirements” and have intentionally omitted the option of “<em>Reject</em>”. I would reject such a paper for methodological flaws, etc., but not for any of the reasons on my list below. The list is intended only to improve, not prevent, software publication.</p>
<p>I believe any of the decisions below typically result in a revision to the same journal, that authors judiciously choose how to respond to reviewer comments guided by the editor’s own feedback, and that it is ultimately the editor’s decision whether any of this is relevant. <code>&lt;/edit&gt;</code></p>
<h2 id="reject-and-resubmit">“Reject and resubmit”</h2>
<h3 id="automatic-tests">Automatic tests</h3>
<p>A scientific R package <em>must must must</em> have some automated tests that are run by <code>R CMD check</code>. Even if further development of the package doesn’t break anything (most likely only if further development doesn’t happen), changes to the package dependencies could still break things, and so it is crucial to have a mechanism in place to detect these problems when they arise. For code that runs quickly, the simplest way to do this is through the examples in the documentation. I don’t expect all scientific software to have a complete test suite with 100% coverage covering all the weird things that can happen if a user passes in a matrix when the function expects a data frame or has some unanticipated missing values, etc. Just some tests to make sure the basic examples execute and I’ll be happy. Longer running functions or those that need calls to external web resources shouldn’t be run in the examples (too much of a burden for CRAN’s automatic testing) so they should be marked <code>dontrun</code> and put in a separate test suite or vignette as it says in the manual.</p>
<h3 id="passing-optional-arguments">Passing optional arguments</h3>
<p>I see authors write functions like this all the time:</p>
<pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">myfunction</span>(f, p){ 
  <span class="co">#  stuff</span>
  o &lt;-<span class="st"> </span><span class="kw">optim</span>(f, p)
  <span class="co">#  stuff</span>
}</code></pre>
<p>calling an existing library function like <code>optim</code> that has a whole host of very useful optional arguments that have a significant impact on how the algorithm functions. Whenever you a rich function like <code>optim</code>, please have the courtesy to make it’s arguments available to future users and developers through your function call. Yes, most users will just want the default arguments, (or your default arguments, if different), and that can be handled just fine by providing default values as optional arguments. R has a fantastic mechanism for this exact issue: the <code>...</code> argument. The above code could be fixed simply by using:</p>
<pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">myfunction</span>(f, p, ...){ 
  <span class="co">#  stuff</span>
  o &lt;-<span class="st"> </span><span class="kw">optim</span>(f, p, ...)
  <span class="co">#  stuff</span>
}</code></pre>
<p>which works just they way you think it would. If you have more than one such function (ask yourself if you can write shorter functions first and then) pass optional arguments as lists,</p>
<pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">myfunction</span>(f, p, optim_options, fn2_options){
  <span class="co"># stuff</span>
  o &lt;-<span class="st"> </span><span class="kw">do.call</span>(optim, <span class="kw">as.list</span>(<span class="kw">c</span>(f, p, optim_options)))
  <span class="co"># stuff</span>
  b &lt;-<span class="st"> </span><span class="kw">do.call</span>(fn2, fn2_options)
  <span class="co"># stuff </span>
}</code></pre>
<p>arguments can also be extracted with <code>list(...)$arg1</code> etc.</p>
<p>A converse of this issue is not providing default arguments where it might be natural to do so. This does not bother me so much, as it is probably useful to force the user to think how many iterations <code>n</code> are appropriate for their problem rather than just assuming that <code>100</code> is good because it is the default. The only time this case is annoying is when the argument will not be changing – such as a user’s authentication token to access a web resource. Don’t make me manually pass the token to every function in the library please.</p>
<h3 id="development-site-and-bug-tracker">Development site and bug tracker</h3>
<p>I would really like to see a link to the software development page, such as r-forge or Github. The primary asset in this context is pointing reviewers to an address with a bug tracking system where issues can be assigned ticket numbers and readers can transparently see if a package is being actively maintained. A reader who comes across the paper years later who has only an email address that may or may not work has little way to determine what the latest version of the code is, whether it is actively maintained, or whether earlier versions that may have been in used in previous publications suffered from any significant bugs.</p>
<h3 id="cite-your-dependencies">Cite your dependencies!</h3>
<p>We write software papers with the sometimes vain hope that they will be cited by users, so authors of such papers should at least follow these best practices themselves. R includes a native mechanism for providing citations to packages, <code>citation(packagename)</code>, including the information for any software paper published along with it. Be sure to add your own software papers to the <code>CITATION</code> file. More information can be found in my post on <a href="http://purl.org/cboettig/2012/03/20/citing-r-packages.html">Citing R packages</a>.</p>
<h2 id="major-revisions">“Major Revisions”</h2>
<p>These are other things that commonly frustrate me, but fall on a bit more of a continuum of style rather than gross oversights. As such I’m not sure that any one of these things would justify rejection.</p>
<h3 id="functionalize-the-code">Functionalize the code</h3>
<p>Style guides will tell you to keep functions short, not more than a screen or 20 lines. Breaking large functions into a series of smaller functions and documenting those smaller functions – even if they are only used internally – is a great help to a reviewer trying to understand what a function is supposed to do and also test that it does what it says. Anyone building the code base later (most often yourself) will appreciate the reusable modules.</p>
<h3 id="stable-clean-and-complete-return-objects">Stable, clean, and complete return objects</h3>
<p>An extension of providing optional arguments to functions is to also provide access to all of their return information. To extend the example from wrapping <code>optim</code>, this would involve returning the convergence information. Using object classes and helper functions for return objects helps keep code stable and lets users leverage existing code for similar objects, such as fitting or plotting routines. More discussion on this topic based on my own experiences in the post, <a href="http://carlboettiger.info/2013/04/23/we-need-more-object-oriented-design-in-comparative-methods.html#comment-878249659">we need more object oriented design in comparative methods</a></p>
<h3 id="state-a-license">State a license</h3>
<p>Because CRAN requires this through the DESCRIPTION file, R package authors rarely neglect this entirely. A sometimes misconception is that because R itself is primarily dual-licensed under GPL-2 and GPL-3 that R packages must use a GPL license due to the “viral” clause of the GPL. This clause only applies if you are modifying existing GPL functions directly and is not a requirement for R packages, which recognize a large array of licenses. My own recommendation for authors seeking to maximize the impact of their work is to use MIT, BSD (2 clause), or CC0 license for the package. CC0 has the advantage of being suitable for and data or documentation included, but authors should do there homework and decide what is best for them.</p>
<h2 id="minor-revisions">“Minor Revisions”</h2>
<p>Consistent use of coding style, good documentation, clear examples, intelligent reuse of code, and other best practices are all areas in which any work could improve. While we can all become better developers by highlighting these issues in our reviews, they are probably best developed over time and in dialog with the user community. I also put anything extending the scope of functionality into this category – I do not have any concept of minimal contribution as long as the code meets the criteria above. Meanwhile, there’s always a few pet peeves I just cannot help mentioning. Here’s one which is particular to R packages and so commonly overlooked.</p>
<h3 id="imports-not-depends">IMPORTS not DEPENDS</h3>
<p>Many developers overlook that package dependencies that provide functions your functions will use internally should be listed as under IMPORTS rather than DEPENDS. This keeps the users namespace cleaner and avoids collisions of functions having the same name. Use DEPENDS only for those packages whose functions will be used by the end user as well.</p>
<hr />
<p>If you are an author, editor, or reviewer of R software packages, what are your pet peeves?</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/13/what-I-look-for-in-software-papers.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/13/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 13 Jun 2013</p>

<p style="font-style:italic"> pageviews: 15 </p>

<article>
<div class="excerpt">
<p>Assembling various loose ends and some ropensci work.</p>
<h2 id="misc-things-to-work-out-and-write-at-some-date">Misc things to work out and write at some date</h2>
<ul>
<li>Catch up on literature reading (unsorted list in Mendeley). Get entries into notebook.<br /></li>
<li>Write thoughts on selecting links anchor text in the notebook</li>
<li>knitr script workflow: externalizing chunks, cache and figure prefixes.<br /></li>
<li>branching projects and unique repositories</li>
<li>source code for all manuscripts into repositories</li>
<li>clean up directories. in particular, establish separate prosecutors-fallacy repository</li>
<li>Comments on workflow: Date-oriented (notebook) vs File-oriented (Github) explanation. (Or is Github both, via history button, and the notebook just the summary that links to it?)</li>
</ul>
<h2 id="ropensci">ropensci</h2>
<p>As per <a href="http://ropensci.org/blog/2013/06/12/sloan/">yesterday’s announcement</a>, we are officially funded by a generous grant from the Alfred P. Sloan Foundation!</p>
<p>Today:</p>
<ul>
<li>Writing feedback on goals, comments to advisors</li>
</ul>
<h3 id="taxize">taxize</h3>
<ul>
<li>My friendly <a href="https://github.com/ropensci/taxize_/issues/151">review</a> of Scott’s taxize <a href="https://github.com/ropensci/taxize_/tree/dev_ms/inst/doc">manuscript</a> (<a href="https://github.com/ropensci/taxize_/blob/dev_ms/inst/doc/taxize_withcode_plos.pdf?raw=true">pdf</a>)</li>
</ul>
<h3 id="ropensci-website">ropensci website</h3>
<p>Scott has a good question about automatic TOC sidebars.</p>
<p>Can generate TOC with redcarpet filter, but it grabs all headers. Unclear how to modify so that it grabs only headers above a certain level. See <a href="https://github.com/vmg/redcarpet/issues/267">my attempt as an issue in redcarpet repo</a>. Similarly pandoc only provides <code>--toc</code> as a compile flag, which does not give us the same control over it’s placement in the HTML.</p>
<h3 id="misc-semantics">misc semantics</h3>
<ul>
<li>Interesting <a href="http://serena.macs.hw.ac.uk/serena/discover-me-semantically/process.php">example</a> of RDF visualization and linking to external linked-data resources. In particular, good use/example of linking to DBpedia information. See <a href="https://github.com/cboettig/labnotebook/issues/63">issue #63</a></li>
</ul>
<h2 id="misc">Misc</h2>
<ul>
<li>Open peer review on <a href="http://publications.copernicus.org/services/public_peer_review.html">copernicus platform</a></li>
<li><p>Smith fellows 2014 <a href="http://www.conbio.org/mini-sites/smith-fellows/apply/proposal-guidelines">applications open</a></p></li>
<li><p>Neil’s <a href="http://www.software.ac.uk/blog/2013-04-09-five-stars-research-software">five stars of research software</a></p></li>
<li><p>Titus’s <a href="http://ivory.idyll.org/blog/ladder-of-academic-software-notsuck.html">academic software notsuck</a></p></li>
<li><p>Nature Chemistry’s take on <a href="http://blogs.nature.com/thescepticalchymist/2013/06/all-you-can-tweet-the-blog-version.html/?WT.mc_id=TWT_NatureBlogs">Twitter for publishers</a></p></li>
<li><p>A Nature Chemistry blog reflects on <a href="http://blogs.nature.com/thescepticalchymist/2013/06/speaking-frankly-the-allure-of-pasteurs-quadrant.html">Pasteur’s Quadrant</a></p></li>
<li><p>G8 Science ministers <a href="https://www.gov.uk/government/news/g8-science-ministers-statement">call for</a> global research infrastructure, open scientific research data and expanded access to results</p></li>
</ul>
<h4 id="how-i-use-git-with-svn-repositories">How I use git with svn repositories</h4>
<p>I don’t use SVN much, but in my opinion “commits” mean something different in SVN then in Git anyway, so I’d rather not entangle the commit history. Because merging is hard in SVN and all local changes are slave to the master central repository, commits tend to be larger, more polished things. Commits and merges are cheap and easy in git, so I commit much more often. By having to commit twice, I am deliberate about whether I’m committing to SVN, with all the overhead that involves, or whether I’m just committing to git.</p>
<p>I think my approach is much easier than <a href="http://cameron.bracken.bz/git-with-r-forge">this</a></p>
<h2 id="my-replies-to-data-quality-survey">My replies to data quality survey</h2>
<h5 id="what-constitutes-data-quality">What constitutes data quality?</h5>
<p><strong>Statistical quality</strong> Measurement error, the degree to which potential covariates are either controlled for or co-measured, and the sample size.</p>
<p><strong>Computational quality</strong> Open and standardized format (csv, xml in defined schema of the field, etc) Thorough metadata conforming to tightly defined and verifiable schema. Terms defined as URIs / linked data. Openly accessible under permissive licensing.</p>
<h5 id="how-might-data-quality-be-assessed-generally">How might data quality be assessed generally?</h5>
<p>Statistical quality depends on what the researcher wants to do with it, and is generally easy for the researcher to assess. All the elements listed under statistical quality should simply be included in metadata, and researchers can draw their own conclusions about sample size, covariates, etc.</p>
<p>Computational quality measurement should reflect data management practices more broadly. Something like Tim Berners Lee’s five-star mechanism for linked data, or badges that indicate if data includes verification tools, etc.</p>
<h5 id="datacites-role">DataCite’s Role?</h5>
<p>Good quality data should be incentivised the same way good publications are. If good quality data can make a career, researchers will rise to the challenge with a little help for proofreading and formatting. Ease of depositing data in a useful format and sharing that data, along with metrics to reflect the impact it has, can go a long way in this regard.</p>
<!-- Text that didn't make it into my post 

## Where to submit

Actually, I don't care.  Almost any journal publishing methodology recieves software papers, which often provides a clear target audience.  For instance, though the [JSS]() is best adapted to publishing papers involving R packages, the editorial board is not one of ecologists or evolutionary biologists, who more frequently publish software papers in [MEE](), [Oxford Bioinformatics]() or [PLoS Bioinformatics]().  The JISC's Journal of Open Research Software ... 


-->


 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/13/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/06/10/mansucript-reviews-on-github.html">manuscript reviews on github?</a></h4></header>
<p><span>examples and questions</span></p>
<p style="font-style:italic"> 10 Jun 2013</p>

<p style="font-style:italic"> pageviews: 340 </p>

<article>
<div class="excerpt">
<p>I was recently impressed to learn of Trevor Bedford’s strategy of seeking <a href="https://twitter.com/trvrb/status/334310856982671361">pre-approval</a> for posting his reviewer’s comments as Github issues. Beyond providing links to the data and source-code, I generally don’t advertise the open science nature of papers I submit – I guess I assume that if the reader or reviewers care, it should be easy enough for them to discover it. Consequently I am usually immediately frustrated to realize that upon receiving my reviews I have to create a second, private repository for the review material, our replies to reviewers, etc., as I don’t have permission to disclose that information. <sup><a href="#fn1" class="footnoteRef" id="fnref1">1</a></sup> I have recently stumbled across <a href="http://www.steinsaltz.me.uk/pnas.html">several</a> <a href="http://theseamonster.net/2013/05/are-unreasonably-harsh-reviewers-retarding-the-pace-of-coral-reef-science/">examples</a> of authors publishing to the web anonymous reviews they have received. Though anonymous, I feel the practice potentially murky without explicit permission, so I would appreciate any insight others have on this.</p>
<h3 id="asking-permission">Asking permission</h3>
<p>Trevor’s approach suggests I should consider broaching this question when first submitting my review, so I am puzzling over the best way to do so. One option would be to include such a request in the cover letter. For example,</p>
<blockquote>
<p>The authors of this manuscript would like to request that the editor and reviewers indicate in their replies if they consent or decline to have their comments posted anonymously in the public <a href="#">Issues Tracker</a> of this paper.</p>
</blockquote>
<p>Does that need more explanation? A link to examples like <a href="https://github.com/trvrb/flux/issues?labels=reviewer+1">Trevor’s</a> that have done this before? Do I need to explain the value of having this kind of transparent provenance for the paper? Should I mention how this could give the reviewer more transparent credit? Encourage them to comment directly on Github from their own account?</p>
<p>Does this need the blessing of the journal? How would you feel about such a clause as a reviewer or editor? For a recent review I had done of a paper that was similarly written on Github, I obtained the Journal’s permission to post my review as an <a href="https://github.com/weecology/data-sharing-paper/issues/71">issue</a> in their repository. I would love to see more examples of this kind of thing, if anyone has come across them.</p>
<h2 id="cover-letters-for-open-science-manuscripts">Cover letters for open science manuscripts?</h2>
<p>While I lean towards a minimal statement such as the one above, perhaps a cover letter would be a good place to document some of the other open and reproducible features of the manuscript? Or perhaps such statements should be added to the manuscript itself? Among the options, I might point out:</p>
<ol type="1">
<li>The manuscript has been written on Github. Consequently the full drafting and <strong>revision history</strong> is available, along with graphs of author contributions (which omit authors without Github accounts and may be distorted by trivial line changes)</li>
<li>The manuscript has been written with all the code necessary to repoduce the results embedded as a <a href="http://yihui.name/knitr">knitr</a> <strong>dynamic document</strong>. This helps ensure the analysis is always in synch with the results presented in the manuscript and the that the research is reproducible. The analysis, figures, and manuscript can be reassembled from scratch by typing <code>make pdf</code> in the repository directory.<br /></li>
<li><strong>Code</strong> to replicate the analysis and produce each of the figures shown can be found at: (Version-stable lnk to the appropriate Github pages? Deposit in Figshare/Dryad first?)<br /></li>
<li><strong>Data</strong> to replicate the analysis and data shown in each of the figures can be found at: (Easiest to link to Github, since the code and data already reside there.<br /><em>Alternatively I could deposit these in <a href="http://figshare.com">Figshare</a> or <a href="http://datadryad.org">Dryad</a> first…)</em></li>
<li>The manuscript, code, data, and documentation are available <strong>as an R package in the Github repository</strong>.<br /></li>
<li>The <strong>issues tracker</strong> associated with the manuscript’s repository provides a record of this research, including lines of investigation that were resolved into the results presented here, lines that were closed as dead-ends or null results, and outstanding issues for further investigation.<br /></li>
<li>The <strong>daily lab notebook entries</strong> accompanying this research can be found under the <a href="/tags">project-tag</a> between dates of XX and XX.</li>
</ol>
<p>Listing all of these would make for a somewhat lengthy cover letter, which might be a bit overwhelming to be useful (or seem more promotional than valuable). Are any of the seven things above worth highlighting in particular?</p>
<p>Perhaps these details could be deferred to a README file in the project’s Github repo, and the cover letter could simply provide a link to the project repository? What, if anything, would appear most useful and accessible to a reviewer unfamiliar with this approach or its potential value? Though elements of this approach have been discussed in the published literature, e.g. <span class="showtooltip" title="Gentleman R and Temple Lang D (2007). Statistical Analyses And
Reproducible Research. _Journal of Computational And Graphical
Statistics_, *16*. ISSN 1061-8600, 
http://dx.doi.org/10.1198/106186007X178663."><a href="http://dx.doi.org/10.1198/106186007X178663" rel="http://purl.org/spar/cito/citesAsEvidence" >Gentleman &amp; Temple Lang (2007)</a></span> ‘compendium’ concept, <span class="showtooltip" title="Stodden V (2009). Enabling Reproducible Research: Open Licensing
for Scientific Innovation. 
http://papers.ssrn.com/sol3/papers.cfm?abstract_id=1362040. 
http://papers.ssrn.com/sol3/papers.cfm?abstract_id=1362040."><a href="http://papers.ssrn.com/sol3/papers.cfm?abstract_id=1362040" rel="http://purl.org/spar/cito/citesAsEvidence" >Stodden (2009)</a></span> RRS concept, or <span class="showtooltip" title="Peng R (2011). Reproducible Research in Computational Science.
_Science_, *334*. ISSN 0036-8075, 
http://dx.doi.org/10.1126/science.1213847."><a href="http://dx.doi.org/10.1126/science.1213847" rel="http://purl.org/spar/cito/citesAsEvidence" >Peng (2011)</a></span> reproducible papers in the Journal of Biostatistics, I’m unsure if pointing a reviewer to these references would be more valuable or more confusing. Let me know what you think.</p>
<h2 id="references">References</h2>
<ul>
<li>Robert Gentleman, Duncan Temple Lang, (2007) Statistical Analyses And Reproducible Research. <em>Journal of Computational And Graphical Statistics</em> <strong>16</strong> <a href="http://dx.doi.org/10.1198/106186007X178663">10.1198/106186007X178663</a></li>
<li>R. D. Peng, (2011) Reproducible Research in Computational Science. <em>Science</em> <strong>334</strong> <a href="http://dx.doi.org/10.1126/science.1213847">10.1126/science.1213847</a></li>
<li>Victoria Stodden, (2009) Enabling Reproducible Research: Open Licensing for Scientific Innovation. <a href="http://papers.ssrn.com/sol3/papers.cfm?abstract_id=1362040">http://papers.ssrn.com/sol3/papers.cfm?abstract_id=1362040</a></li>
</ul>
<section class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Strictly speaking the edits to the manuscript in the open repository could also be considered confidential, though at that stage I haven’t yet signed the copyright agreements that come with publication, which tend to be quite reasonable even for the traditional subscription based journals I work with<a href="#fnref1">↩</a></p></li>
</ol>
</section>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/06/10/mansucript-reviews-on-github.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

