<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-07-31T22:00:31-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-07-31T22:00:31-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-07-31T22:00:31-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
    <div class="excerpt">
      <div class="scroll">
        <ul><li>cboettig pushed to master at cboettig/labnotebook: <em>draft new post post</em> <a href="https://github.com/cboettig/labnotebook/compare/75ec57927d...8867cb0746">11:09 2013/09/16</a></li><li>cboettig closed issue ropensci/rfishbase#13: <em>Environment / Climate / Range</em> <a href="https://github.com/ropensci/rfishbase/issues/13">08:53 2013/09/16</a></li><li>cboettig commented on issue ropensci/reml#47: <em>@mbjones Thanks! Yes, I agree entirely about getting more of the eml-attribute level metadata in, and struggle a bit to find an approach that is b…</em> <a href="https://github.com/ropensci/reml/issues/47#issuecomment-24540927">08:17 2013/09/16</a></li><li>cboettig commented on issue cboettig/labnotebook#108: <em>Yup, HTML comments are still parsed. Liquid has it's own commenting syntax, {% comment %} . glad things are working for you.</em> <a href="https://github.com/cboettig/labnotebook/issues/108#issuecomment-24534304">06:49 2013/09/16</a></li><li>cboettig commented on issue cboettig/knitcitations#46: <em>@trinker That makes sense; not currently possible since bibliography is a print method that invisibly returns bibliography while displaying the for…</em> <a href="https://github.com/cboettig/knitcitations/issues/46#issuecomment-24524661">04:41 2013/09/16</a></li></ul>
      </div>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
      <div class="scroll">
       <ul><li><p>[1307.8250] Critical Transitions in Social Network Activity <a href="http://t.co/SCayT97ebm">http://t.co/SCayT97ebm</a></p>
 <a href="http://twitter.com/cboettig/statuses/379717389073805312">02:24 2013/09/16</a> </li><li><p>Interesting persp. on #openscience. asks &quot;How much pay would you give up to retain the freedom to publish while in… <a href="http://t.co/2rznX2NHRF">http://t.co/2rznX2NHRF</a></p>
 <a href="http://twitter.com/cboettig/statuses/379713582201835521">02:09 2013/09/16</a> </li><li><p><a href="http://t.co/YXXVrA7VZp">http://t.co/YXXVrA7VZp</a></p>
 <a href="http://twitter.com/cboettig/statuses/379647986768166913">09:48 2013/09/16</a> </li><li><p>Toward a comprehensive impact report for every software project <a href="http://t.co/rsgD5jiGIG">http://t.co/rsgD5jiGIG</a></p>
 <a href="http://twitter.com/cboettig/statuses/378972405588193280">01:03 2013/09/14</a> </li><li><p>RT @LeahAWasser: Another brilliant Graphic! RT @ninmah: #ISEES Day Two: Workforce Development, morning conversation (visual notes) http://t…</p>
 <a href="http://twitter.com/cboettig/statuses/377938795917615105">04:36 2013/09/11</a> </li></ul>
      </div>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <div class="scroll">
      <ul>
<li>Stephan B Munch, A Kottas. 2009. A Bayesian modeling approach for determining productivity regimes and their characteristics <em>Ecological Applications</em> <strong>19</strong> 2 <a href="http://dx.doi.org/10.1890/07-2116.1">10.1890/07-2116.1</a> </li>
<li>Mark John Costello, Marta Coll, Roberto Danovaro et al. 2010. A Census of Marine Biodiversity Knowledge, Resources, and Future Challenges <em>PLoS ONE</em> <strong>5</strong> 8 <a href="http://dx.doi.org/10.1371/journal.pone.0012110">10.1371/journal.pone.0012110</a> </li>
<li>R. F. Hechinger, K. D. Lafferty, a. P. Dobson et al. 2011. A Common Scaling Rule for Abundance, Energetics, and Production of Parasitic and Free-Living Species <em>Science</em> <strong>333</strong> 6041 <a href="http://dx.doi.org/10.1126/science.1204337">10.1126/science.1204337</a> </li>
<li>Jonathan B. Shurin, Elizabeth T. Borer, Eric W. Seabloom et al. 2002. A cross-ecosystem comparison of the strength of trophic cascades <em>Ecology Letters</em> <strong>5</strong> 6 <a href="http://dx.doi.org/10.1046/j.1461-0248.2002.00381.x">10.1046/j.1461-0248.2002.00381.x</a> </li>
<li>Wayne M Getz, David Saltz. 2008. A framework for generating and analyzing movement paths on ecological landscapes. <em>Proceedings of the National Academy of Sciences of the United States of America</em> <strong>105</strong> 49 <a href="http://dx.doi.org/10.1073/pnas.0801732105">10.1073/pnas.0801732105</a> </li>

      </div>
    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/09/11/extending-data-frame-class.html">Extending Data Frame Class</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 11 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>I’d like to define a class that acts just like a <code>data.frame</code>, just like the <code>data.table</code> class does, but contains some additional metadata (e.g. the units associated with the columns) and has some additional methods associated with it (e.g. that might do something with those units) while also working with any function that simply knows how to handle <code>data.frame</code> objects.</p>
<p>How might this be done?</p>
<hr />
<p><em>I’m not really sure where to start on this, so below is a summary of my attempt so far and some further puzzles I have stumbled across (e.g. is a <code>data.frame</code> and S4 object?). This seems to work, but might not be the best way to go about it.</em></p>
<hr />
<p>If I understood what the data structure underlying <code>data.frame</code> was, I might have a better idea how to extend it. But I don’t know what makes a <code>data.frame</code> a <code>data.frame</code>, as the exploration below will illustrate:</p>
<p>For instance, it’s not clear if I should try writing my extended object as an S3 or an S4 object that inherits <code>data.frame</code>. When is a <code>data.frame</code> an S4 object and when is it an S3 object?</p>
<pre><code>A &lt;- new(&quot;data.frame&quot;)
isS4(A) </code></pre>
<p>returns <code>TRUE</code>, suggesting that data.frames are S4 objects and that we might start with: <code>setClass(&quot;myclass&quot;, contains=&quot;data.frame&quot;)</code>. However,</p>
<pre><code>B &lt;- data.frame(x = rnorm(10))
isS4(B) </code></pre>
<p>returns <code>FALSE</code>, suggesting that <code>data.frames</code> are S3 classes, and perhaps I should consider starting with</p>
<p>attr(B, “class”) &lt;- c(“myclass”, “data.frame”)</p>
<p>It appears that a data.frame is built upon a list, since</p>
<pre><code>is.list(data.frame(x = rnorm(10)))</code></pre>
<p>gives <code>TRUE</code>.</p>
<p>Despite <code>isS4(B)</code> being <code>FALSE</code>, we find that the S4 method <code>slotNames(B)</code> gives:</p>
<pre><code>&gt; slotNames(B)
[1] &quot;.Data&quot;     &quot;names&quot;     &quot;row.names&quot; &quot;.S3Class&quot; </code></pre>
<p>suggesting that <code>data.frame</code> is an S4 object, with a (unnamed?) list in <code>.Data</code> and nonempty values in <code>names</code>, along with the slot “.S3Class” containing the string “data.frame”.</p>
<p>Interestingly:</p>
<ul>
<li><code>B@names</code> returns an error, while <code>slot(B, &quot;names&quot;)</code> returns <code>&quot;x&quot;</code>. (Suggesting <code>B</code> is S3, with some <code>slot</code> method defined?)<br /></li>
<li><code>B@.Data</code> returns an unnamed length-1 list containing a numeric vector with our 10 random values. (and illustrating a working <code>@</code> accessor, suggesting <code>B</code> is S4.)</li>
</ul>
<p>Very confusing.</p>
<hr />
<p>From this exploration, I conclude that I can promote a <code>list</code> to a <code>data.frame</code> by providing an unnamed list, editing the “class” attribute, and then assigning values to the slots <code>names</code> and <code>row.names</code>:</p>
<pre><code>obj &lt;- list(rnorm(10))
attr(obj, &quot;class&quot;) &lt;- &quot;data.frame&quot;
slot(obj, &quot;names&quot;) &lt;- &quot;x&quot;
slot(obj, &quot;row.names&quot;) &lt;- 1:10</code></pre>
<p>And indeed <code>obj</code> acts like a proper <code>data.frame</code> (at least it prints like one).</p>
<p>This suggests that I could extend <code>data.frame</code> by adding a new slot for my units, like so:</p>
<p>setClass(“myclass”, representation(units = “character”), contains=“data.frame”)</p>
<p>But this does not inherit the slotNames <code>names</code>, <code>row.names</code>, <code>.S3Class</code>. I try assigning these explicitly,</p>
<pre><code>setClass(&quot;myclass&quot;, 
         representation(units = &quot;character&quot;,
                        names = &quot;character&quot;,
                        row.names = &quot;integer&quot;,
                        .S3Class = &quot;character&quot;),
                        contains = &quot;list&quot;)

obj &lt;- new(&quot;myclass&quot;, names=&quot;x&quot;, row.names=1:10, units = &quot;degrees Celsius&quot;)
obj@.Data &lt;- list(rnorm(10))
obj@.S3Class &lt;- &quot;data.frame&quot;</code></pre>
<p>but this new object is not a <code>data.frame</code>. Finally we add</p>
<pre><code>attr(obj, &quot;class&quot;) &lt;- &quot;data.frame&quot; </code></pre>
<p>and we have a successful extension that acts like a data frame. Some weird things of note: It appears that .S3Class needs to be assigned as above with <code>@</code>, and not as a named argument in <code>new</code> or as part of the <code>prototype</code>.</p>
<p>This last step is almost certainly not what we really want to do, since it destroys our new class assigment (<code>myclass</code>). We really want the object to have two classes, <code>myclass</code> and <code>data.frame</code>. However, if we do <code>attr(obj, &quot;class&quot;) &lt;- c(&quot;myclass&quot;, &quot;data.frame&quot;)</code>, then the object does not behave as a <code>data.frame</code> at all.</p>
<p>One might thing that the solution would be to then define all the S3 methods as copies of those avialable for data.frames, e.g. define <code>print.myclass &lt;- print.data.frame</code>. Unfortunately this doesn’t quite work (appears not to get <code>row.names</code> and hence prints column headings but not data. Defining the S4 method for <code>row.names</code> with <code>myclass</code> signature doesn’t fix this either…).</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/11/extending-data-frame-class.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/11/ISEES-Workforce-Development-Day-2.html">Isees Workforce Development Day 2</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 11 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>Near the end of another excellent day of discussions we were asked to put pen to paper to outline our own views in writing. My own contribution was the following notebook scribble.</p>
<hr />
<p><em>My own vision of what ISEES should be doing for workforce development:</em></p>
<p>ISEES would employ a multi-scale approach to improve basic computational expertise across the entire workforce necessary to fully participate in the modern software-immersed world of research through the development of <em>training standards</em>, while using a <em>working-group model</em> focused on emerging researcher-developer communities which would simultaneously improve the quality and synthesis of existing software tools while also delivering essential skills to the most transformative members of the research software communities.</p>
<p>Development and promotion of standards for computational training. A workshop model could allow these to be written and updated by the community. These could serve as a guideline for curriculum development and play a key role in assessment. Curriculum review panels within departments, as well as reviewers of grants which include a component of computational training, would at least have such consensus standards to which they could point, either to reject sub-standard practices or make the case for increased resources to meet these goals. Curriculum should be geared towards inclusion of this training as part of domain-specific classes rather than extramural classes.</p>
<p>Working groups or hack-a-thons bringing together “researcher-developer” domain scientists with software engineering experts would simultaneously provide a direct benefit to “hardening” (improving the interoperability, scalability, and sustainability) of existing software while also training the most influential members of the research community who are most frequently involved in training other domain scientists. By organizing such working groups around existing research communities, the incentive to participate would be higher as the outputs would be more directly aligned with existing research programmes, while at the same time training the researchers most active in software development in their domain and most likely to teach these skills to others.</p>
<p>Thirdly, ISEES should support workshops or small conferences to address developing concerns in software workforce training, including issues such as assessing current training practices, measuring improvement and the impact of the center, or challenges in software career paths.</p>
<hr />
<p>Didn’t really jot many notes down in today’s session, but will instead link the excellent prepared by our awesome expert facilitator <a href="https://twitter.com/ninmah">Rachel Smith</a>.</p>
<figure>
<img src="http://farm4.staticflickr.com/3807/9727344142_8d449003a7_c.jpg" />
</figure>
<figure>
<img src="http://farm4.staticflickr.com/3727/9724949449_a5a7998b6c_c.jpg" />
</figure>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/11/ISEES-Workforce-Development-Day-2.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/10/ISEES-training-workshop-day-1.html">Isees Training Workshop Day 1</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 10 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<h2 id="education-workforce-development-group">Education / Workforce Development Group</h2>
<p>Cannot do grand synthesis opaquely and be science. must be able to scale and evolve.</p>
<p>How did data succeed? What features really made the difference and what were just sideshows?</p>
<ul>
<li>Sticks (Funder and journal mandates?)</li>
<li>Working groups (NCEAS)</li>
<li>Training</li>
<li>external forces: there is more data.</li>
</ul>
<p>Why one center? Why not 15 centers? Divided by discipline rather than region…. synthesize the synthesis centers</p>
<p>(on the ground risk assessment. the huge outside-of-the-academy representation).<br />Computational training - math stats or computer science? Both.</p>
<ul>
<li>Grassroots</li>
<li>Infrastructure</li>
<li>Killer app examples</li>
</ul>
<p>(Relatively) easy to just do it and do it wrong. Grand challenge synthesis – just plug different models and different data together in a terrible way.</p>
<p>So what? step 1 is to try. step 2 is to iterate. must be able to do step 2.</p>
<p>Papers don’t scale. What was once a robust exchange is largely talking past each other. We look at different data,</p>
<p>Levels of software</p>
<ul>
<li>stats</li>
<li>one off models</li>
<li>custom analytics</li>
<li>modeling frameowrks</li>
<li>community models</li>
<li>workflows</li>
<li>compute engines</li>
<li>data management</li>
<li>services (blast)</li>
</ul>
<hr />
<p>The ‘odd’ model framework for individual based models. Hard to define/describe a model when it cannot be consisely stated in mathematics.<br />see <a href="http://jasss.soc.surrey.ac.uk/11/2/3.html">Polhill et al 2008</a></p>
<hr />
<h3 id="stakeholders">Stakeholders</h3>
<ul>
<li>who has the most to lose The student. Perhaps the tax payer</li>
<li>who would have most impact (on training?) Administrators (e.g. university level). Funder. Reviewer.</li>
<li>who has most to contribute Researcher-developer (author).</li>
</ul>
<p>Understanding current skills and practice: http://doi.org/10.1525/bio.2012.62.12.8</p>
<p>Work involving computation is often considered interdisciplinary. We don’t</p>
<p>Literacy</p>
<p>Role of institutional instruction vs the software center.</p>
<p>Teaching directly vs supporting training.</p>
<p>Understanding, development, and publication of community standards. Something for reviewers of papers, grants, departments to point to</p>
<h2 id="day-1-wrapup">Day 1 Wrapup</h2>
<h3 id="organization-and-governance">Organization and Governance</h3>
<ul>
<li>Stakeholders: Students, Educators, Researchers, Professional societies &amp; Institutions, Funders, Data Repositories, Software Community</li>
<li>Client: Researchers.<br /></li>
<li>Where to engage</li>
<li>Values: Results-oriented transparent collaborative community-drive, useful</li>
</ul>
<h2 id="rachel-smiths-visual-notes">Rachel Smith’s Visual notes</h2>
<figure>
<img src="http://farm3.staticflickr.com/2869/9719874150_2e8d3c1b4b_c.jpg" />
</figure>
<figure>
<img src="http://farm8.staticflickr.com/7327/9718138109_0af2579343_c.jpg" />
</figure>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/10/ISEES-training-workshop-day-1.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/09/08/some-rfishbase-updates.html">Some Rfishbase Updates</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 08 Sep 2013</p>

<p style="font-style:italic"> pageviews: 14 </p>

<article>
<div class="excerpt">
<p>Ever since writing the <code>rfishbase</code> package I’ve been frustrated by the lack of a proper API from Fishbase.org that would allow access to that wealth of information that is not exposed in the “Summary XML” files from which <code>rfishbase</code> extracts its content. Since writing the paper in Journal of Fish Biology, I continue to receive user requests asking if or how such-and-such a data element might be extracted. Usually this involves a disappointing boiler-plate reply about how the package is limited to the “Summary XML” data and that improved query potential must wait for a proper API from Fishbase.org at some unspecified time.</p>
<p>I have been in touch with many of the folks at Fishbase.org at various times looking for ways to address these issues. They have been in general keenly supportive of the idea, but I gather that they are largely overworked and under-resourced and as such, implementation of any of these steps has been very slow. Last October another rfishbase user contacted me, who works full time as a software developer. I put him in touch with the UBC team, and he is willing to volunteer the hours and expertise the FishBase development team lacks to help transform the database into a full API. I think this is a very promising development and a nice example of a community effort, but still faces the same fundamental challenges in getting things rolling. So, so far, no <code>rfishbase</code> version 2. (Actually in our semantic versioning it would be the less romantic version 0.3 or 0.4 or so, depending on how many CRAN submits we do before then and provided we don’t break any backwards compatibility)…</p>
<p>One of the more frequent requests involves accessing quantitative trophic level estimates provided for many species on the “Ecology” page, linked from the “Summary page”. The only way to access this data is by scraping the HTML.</p>
<p>Previously I’ve been reluctant to scrape the HTML directly for several reasons:</p>
<ul>
<li><p>It is generally unrealiable/unverifiable. Without any metadata (rdfa or microdata/microformats could have helped…) in the HTML to signpost that we have the correct value, we are generally counting arbitrary header, div and span elements, followed by some Regular Expression matching, to find the element we want.<br />This is error-prone and easily broken by any changes in the html pages either over time or between pages for different species.</p></li>
<li><p>Second, it puts a much greater load on the fishbase servers. Often we cannot even start by querying the page we want, (e.g. since URLs involve things like stock-ids that the end user will not know and I haven’t located a look-up table), so a single call may follow links through several pages before parsing the data it needs. Repeating such queries over many fish programmatically can create a rather substantial burst of traffic that could slow or crash the servers.</p></li>
<li><p>It’s slow. Making all the queries takes time, even for a computer. (Though caching, as we already do in rfishbase, could take care of much this).</p></li>
<li><p>It is potentially time consuming and annoying to code the scaping.</p></li>
</ul>
<p>Okay, the last is just me being lazy. Since this particular query about trophic level was a common email request, and as the data was available from the “Ecology” page of a species in the form of an HTML Table, (and hence somewhat less fragile to robustly identify…), I thought I’d give it a try.</p>
<p>Here’s my approach (or see <a href="">source code</a>). We would follow the same query structure as <code>getSize</code> did <sup><a href="#fn1" class="footnoteRef" id="fnref1">1</a></sup>, and execute several steps:</p>
<ol type="1">
<li>Visit the Summary page of the requested fish</li>
<li>Find it’s Ecology page</li>
<li>Find the Trophic level table</li>
<li>parse the table.</li>
</ol>
<p>In code, that looks like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">getTrophicLevel &lt;- function(<span class="dt">fish.data =</span> <span class="ot">NULL</span>,
                            <span class="dt">path =</span> <span class="ot">NULL</span>,
                            <span class="dt">as_table=</span><span class="ot">FALSE</span>, 
                            <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;diet composition&quot;</span>, <span class="st">&quot;individual food items&quot;</span>), 
                            <span class="dt">unfished =</span> <span class="ot">FALSE</span>,
                            <span class="dt">justSE =</span> <span class="ot">FALSE</span>,
                            <span class="dt">justReference =</span> <span class="ot">FALSE</span>){

  ids &lt;- <span class="kw">getIds</span>(<span class="dt">fish.data =</span> fish.data, <span class="dt">path=</span>path)
  out &lt;- <span class="kw">sapply</span>(ids, function(id){
    summaryPage &lt;- <span class="kw">getSummary</span>(id)
    ecologyPage &lt;- <span class="kw">getEcology</span>(summaryPage)
    tab &lt;- <span class="kw">readTrophicLevel</span>(ecologyPage)
    if(as_table)
      tab
    else
      <span class="kw">parseTrophicLevel</span>(tab, <span class="dt">from=</span>from, <span class="dt">unfished=</span>unfished,<span class="dt">justSE=</span>justSE, <span class="dt">justReference=</span>justReference)
    })
  out
}</code></pre>
<p>Which has a whole bunch of options just to let the user control the format of the output / target value, since there are different numbers.</p>
<p>Each of the steps corresponds to it’s own function:</p>
<pre class="sourceCode r"><code class="sourceCode r">getIds &lt;- function(<span class="dt">fish.data=</span><span class="ot">NULL</span>, <span class="dt">path=</span><span class="ot">NULL</span>){
  if(<span class="kw">is.null</span>(fish.data))
    fish.data &lt;- <span class="kw">loadCache</span>(<span class="dt">path=</span>path)
  ids &lt;- <span class="kw">sapply</span>(fish.data, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&#39;id&#39;</span>)
  species.names &lt;- <span class="kw">sapply</span>(fish.data, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&#39;ScientificName&#39;</span>)
  <span class="kw">names</span>(ids) &lt;- <span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, species.names) <span class="co"># use underscores instead of spaces</span>
  ids
}


getSummary &lt;- function(id){ 
  <span class="co"># Grab and parse page matching id</span>
  url &lt;- <span class="kw">paste0</span>(<span class="st">&quot;http://www.fishbase.org/summary/speciessummary.php?id=&quot;</span>, id)
  summaryPage &lt;- <span class="kw">htmlParse</span>(url) 
}

getEcology &lt;- function(summaryPage){
  link &lt;- <span class="kw">xpathApply</span>(summaryPage, <span class="st">&quot;//*[contains(@href, &#39;/Ecology/FishEcologySummary.php&#39;)][1]&quot;</span>, xmlAttrs)[[<span class="dv">1</span>]][[<span class="st">&quot;href&quot;</span>]]
  ecologyPage &lt;- <span class="kw">htmlParse</span>(<span class="kw">paste0</span>(<span class="st">&quot;http://www.fishbase.org/&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.</span><span class="ch">\\</span><span class="st">./&quot;</span>, <span class="st">&quot;&quot;</span>, link)))
}


readTrophicLevel &lt;- function(ecologyPage){ 
  tab &lt;- <span class="kw">readHTMLTable</span>(ecologyPage)[[<span class="dv">6</span>]]
}</code></pre>
<p>which are all nice and short and potentially self-explanatory. Extracting values from the table we isolate is done rather crudely at the moment and could no doubt be improved, but this is what I came up with:</p>
<pre class="sourceCode r"><code class="sourceCode r">parseTrophicLevel &lt;- function(tab, 
                              <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;diet composition&quot;</span>, <span class="st">&quot;individual food items&quot;</span>), 
                              <span class="dt">unfished =</span> <span class="ot">FALSE</span>,
                              <span class="dt">justSE =</span> <span class="ot">FALSE</span>,
                              <span class="dt">justReference =</span> <span class="ot">FALSE</span>){
  from &lt;- <span class="kw">match.arg</span>(from)
  if(justReference)
    out &lt;- <span class="kw">as.character</span>(tab[<span class="dv">3</span>,<span class="dv">2</span>])
  else{
    col &lt;- <span class="dv">2</span>
    adj &lt;- <span class="dv">0</span>
    if(unfished)
      col &lt;- <span class="dv">4</span>
    if(justSE)
      adj &lt;- <span class="dv">1</span>
    if(from == <span class="st">&quot;diet composition&quot;</span>)
      out &lt;- <span class="kw">as.numeric</span>(<span class="kw">as.character</span>(tab[<span class="dv">2</span>,col+adj]))
    else if(from == <span class="st">&quot;individual food items&quot;</span>)
      out &lt;- <span class="kw">as.numeric</span>(<span class="kw">as.character</span>(tab[<span class="dv">4</span>,col+adj]))
  }
  out 
}</code></pre>
<p>Most of these are internal functions of course, only the top <code>getTrophicLevel</code> is user-facing.<br />The use is straight forward:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rfishbase)
<span class="kw">data</span>(fishbase)

<span class="kw">getTrophicLevel</span>(fish.data[<span class="dv">1</span>:<span class="dv">10</span>])</code></pre>
<p>or any of the various optional configurations, e.g.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getTrophicLevel</span>(fish.data[<span class="dv">1</span>:<span class="dv">10</span>], <span class="dt">as_table=</span><span class="ot">TRUE</span>)</code></pre>
<p>(which gives a list of tables matching these fish).</p>
<p>The astute reader might notice that with <code>EcologyPage</code> etc we have access to a lot more (tabular) data in this approach, which could be extended beyond that page as well, following the same template shown here. Just a matter of time and the other limitations discussed above.</p>
<p>So anyway, if you need the getTrophicLevel, it’s available now on the <a href="https://github.com/ropensci/rfishbase/issues/12">Github version</a> and should be on CRAN in the near future. Meanwhile, I’d certainly welcome other requests (or better yet, forks and pull requests implementing additional data fields along these same lines).</p>
<section class="footnotes">
<hr />
<ol>
<li id="fn1"><p>taking the <code>fish.data</code> list as an argument – potentially a weird choice but oh well, that’s what I did when I first wrote the rfishbase API so may as well keep it. A more natural choice taking scientific names as arguments directly, instead of needing to start with a which_fish query, should be added one day to the entire package API.<a href="#fnref1">↩</a></p></li>
</ol>
</section>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/08/some-rfishbase-updates.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/07/RNeXML-milestone.html">Rnexml Milestone</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 07 Sep 2013</p>

<p style="font-style:italic"> pageviews: 3 </p>

<article>
<div class="excerpt">
<p>We have also recently reached the first <a href="https://github.com/ropensci/RNeXML/issues?milestone=1&amp;state=closed">milestone</a> in <a href="https://github.com/ropensci/RNeXML">RNeXML</a>. We can now read NeXML files into R and convert to the popular <code>phylo</code> format from the <code>ape</code> package, as well as writing the <code>ape::phylo</code> trees out into valid NeXML. We can also add basic metadata such as author, license, title and description to the file.</p>
<h3 id="development-notes">Development notes</h3>
<p>At the heart of this package is an internal S4 representation of the schema. While knowledgable users could interact with the S4 objects directly (easier and less error-prone than interacting with the XML parsing/serializing directly), we in general provide an end-user API which hides the internal complexity of S4 classes, as illustrated below. In principle this representation could be automatically generated from the schema using the <code>XMLSchema</code> package, but that must wait until that package is more mature. Meanwhile, we rely on hand-coded classes and hand-coded coercion functions.</p>
<p>The coercion functions take care of the translation into and out of the XML (parsing and serializing) without any need for xpath expressions. At the moment they are verbosely spelled out under the class definitions in classes.R file, though ideally they could be abstracted (e.g. as we do in <code>reml</code>). Classes are based on the <a href="https://github.com/ropensci/RNeXML/issues/1#issuecomment-22867535">inheritance structure</a> of the schema (unlike <code>reml</code>), which makes additional nodes much easier to add and hence it is easier to extend this library. At this time we have only certain core elements from the schema as map naturally to <code>ape</code> trees, along with the essential annotation elements (<code>meta</code> nodes). Support for additional classes / schema elements will be added as driven by use cases.</p>
<p>(See the <a href="https://github.com/ropensci/RNeXML">Github page</a> for the most recent tutorial, this is just added here for my record).</p>
<h3 id="tutorial">Tutorial</h3>
<p>The development version of RNeXML is <a href="https://github.com/ropensci/RNeXML">available on Github</a>. With the <code>devtools</code> package installed on your system, RNeXML can be installed using:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;RNeXML&quot;</span>, <span class="st">&quot;ropensci&quot;</span>)
<span class="kw">library</span>(RNeXML)</code></pre>
<p>Read in a <code>nexml</code> file into the <code>ape::phylo</code> format:</p>
<pre class="sourceCode r"><code class="sourceCode r">f &lt;- <span class="kw">system.file</span>(<span class="st">&quot;examples&quot;</span>, <span class="st">&quot;trees.xml&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;RNeXML&quot;</span>)
tr &lt;- <span class="kw">nexml_read</span>(f, <span class="st">&quot;phylo&quot;</span>)</code></pre>
<p>Write an <code>ape::phylo</code> tree into the <code>nexml</code> format, including optional metadata:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nexml_write</span>(bird.orders, <span class="st">&quot;birds.xml&quot;</span>, <span class="dt">creator =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>, 
    <span class="dt">title =</span> <span class="st">&quot;example NeXML file for bird orders&quot;</span>, <span class="dt">description =</span> <span class="st">&quot;Example phylogeny taken from the ape documentation showing the major bird orders&quot;</span>)</code></pre>
<h3 id="next-steps">Next Steps</h3>
<p>Developing use cases, soliciting input, and writing a manuscript… See issues log.</p>
<ul>
<li>Outreach <a href="https://github.com/ropensci/RNeXML/issues/22">#22</a></li>
<li>metadata use cases <a href="https://github.com/ropensci/RNeXML/issues/21">#21</a></li>
<li>serializing metadata (implementation) <a href="https://github.com/ropensci/RNeXML/issues/18">#18</a></li>
<li>Testing <a href="https://github.com/ropensci/RNeXML/issues/15">#15</a></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/07/RNeXML-milestone.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/06/NSF-math-bio-postdoc.html">Nsf Math Bio Postdoc</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 06 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>Guess it’s that time of year again. Just saw the reminder for <a href="http://www.nsf.gov/funding/pgm_summ.jsp?pims_id=503622&amp;WT.mc_id=USNSF_39&amp;WT.mc_ev=click">this year’s RFP for NSF Biology Post-docs</a>. As I was fortunate enough to receive one of these for the “Mathematical Biology” track through last year’s call, I have received a few requests for materials and advice. I’m not sure if I have any more insight for having received this funding, but as I appreciated the advice of others this time last year I’ll give it a shot. I found it particularly useful to see another proposal, particularly one that was also from a theoretical angle, so I am grateful to <a href="http://whitelab.weecology.org/">Ethan White</a> for sharing his successful proposal <a href="http://dx.doi.org/10.6084/m9.figshare.93938">on figshare</a> (though for a call under “Bioinformatics” rather than “Mathematical Biology”). So naturally, my own full proposal can be found <a href="http://dx.doi.org/10.6084/m9.figshare.652970">on figshare</a>. (This is also linked from my vita, which hasn’t proven very discoverable). While I do not feel it appropriate to post my reviews (after all, they are not written with the expectation of public release and I don’t own the copyright to them), you can drop me an email if you would like to see these privately.</p>
<p>In that spirit, here is a bit of background on my own proposal with some general reflections on my reviews, both as a chance for me to think about the feedback I received as I now embark on this proposed research, and for whatever value it may have to this year’s applicants.</p>
<hr />
<p>In July I officially started my NSF mathematical biology post-doc with mentors Marc Mangel and Steve Munch: <em>Management for an uncertain world: robust decision theory in face of regime shifts</em>. Over the next two years, this research will focus primarily in pursing three questions:</p>
<ol type="1">
<li><strong>Can early warning signals be incorporated into decision-theoretic approaches to decrease the chance of an undesirable critical transition?</strong><br /></li>
<li><strong>Can the use of non-parametric and machine learning approaches to ecosystem modeling and prediction decrease or increase the chance of an undesirable transition?</strong></li>
<li><strong>What learning strategies can be employed to reduce uncertainty without increasing the chance of accidentally crossing a tipping point?</strong></li>
</ol>
<p>The project brings me into rather new ground into at least two new methodologies (Optimal control, specifically through stochastic dynamic programming – something I have been recently introduced to by economists in a NIMBIOS working group) and a new ecological system (fisheries dynamics). I had to smile on seeing that some of my reviewers fretted that the project sounded too similar to my thesis work. While potentially understandable from a comparison of my thesis title and my NSF proposal title, I had feared the actual content had more the opposite problem.<sup><a href="#fn1" class="footnoteRef" id="fnref1">1</a></sup></p>
<p>Overall my reviews were somewhat more applied than I anticipated, focusing much more on the conceptual goals of the research than the details of the implementation. In retrospect this was probably fortuitous, as I did not have space to fully develop the technical challenges in some of these areas, which are at points formidable (several areas involve methodology that is quite new to me. Question 3 in particular draws on two classically hard problems and I worry myself if I will have the capacity to tackle it effectively). From others I understand that my proposal was potentially somewhat more focused on big picture issues than on implementation details, which may not always be advisable but worked well in this case. It’s unclear to me if this is typical of more theoretical proposals or just luck.</p>
<h3 id="broader-impacts-and-the-open-notebook">Broader Impacts and the Open Notebook</h3>
<p>NSF gives quite a wide range of potential areas for broader impacts. The 2013 changes only made this more vague, extending the listed examples and emphasizing that they were meant to be illustrative, not prescriptive. A particularly promising change comes in using the same <a href="http://www.nsf.gov/pubs/policydocs/pappguide/nsf13001/gpg_3.jsp#IIIA2">criteria as the merit review</a>; emphasizing that the effectiveness of the impact is assessed, and that the applicant is well-qualified for the task. Though I have volunteered much of my adult life with non-profit organizations working with under-served youth in my <a href="/community.html">community</a>, I did not think it appropriate to shoehorn those activities into my proposal. For these reasons I chose not to focus on broadening participation of under-represented groups in this proposal, and instead emphasize three of five (now eight) suggested areas that I felt I could effectively impact through my open lab notebook: education, dissemination, and infrastructure.</p>
<p>I think this was a bit of a gamble, focusing on something that was both more unique to me and that I could implement and assess well, rather than on something more traditionally recognized for broader impacts. It is difficult to know just what my panel thought of this idea – but my impressions is that “good enough” might be the general sentiment. Comments included:</p>
<blockquote>
<p>The PI operates a blog (open notebook) that already receives attention and is an interesting way to promote engagement and learning across a broad community of scientific peers and perhaps the public.</p>
</blockquote>
<blockquote>
<p>the notebook/blog could be developed into an online asynchronous classroom to mentor others in applying the results.</p>
</blockquote>
<blockquote>
<p>The blog is a good idea for dissemination of results.</p>
</blockquote>
<p>Whether or not this was a lukewarm endorsement of the open notebook, I’m excited that as part of my proposal the practice is officially part of my supported proposal and not merely a time-saving endeavor for myself. Though the more technical features seemed to be of little interest to my reviewers, I’ll nonetheless particularly interested in exploring the potential to better leverage such elements as <a href="http://carlboettiger.info/2013/04/04/notebook-parsing.html">semantic markup</a> and <a href="http://carlboettiger.info/2013/05/03/notebook-features-hashes-providing-an-immutable-and-verifiable-research-record.html">hashes</a> in the notebook as I move forward with my research.</p>
<h3 id="reflections-on-data-management">Reflections on Data Management</h3>
<p>The data management plan is supposed to be assessed under both ‘Intellectual Merit’ and ‘Broader Impact’ sections where applicable,</p>
<blockquote>
<p>“The Data Management Plan will be reviewed as an integral part of the proposal, coming under Intellectual Merit or Broader Impacts or both, as appropriate for the scientific community of relevance.” <a href="http://www.nsf.gov/pubs/policydocs/pappguide/nsf13001/gpg_2.jsp#IIC2j">NSF GPG</a></p>
</blockquote>
<p>which may have the effect that it is assessed under neither, unless a reviewer particularly cares about this section. I saw no mention of it in my own reviews. I have posted a draft of my <a href="http://carlboettiger.info/2012/10/09/data-management-plan.html">data management plan</a> previously. Having recently become familiar with the Ecological Metadata Standard I will probably amend the plan to reflect this.</p>
<h4 id="weaknesses">Weaknesses</h4>
<ol type="1">
<li>One reviewer expresses a desire for a case-study application</li>
<li>One suggests more on broader public and under-served groups</li>
<li>One feels the proposal has “some similarities to the dissertation”</li>
</ol>
<section class="footnotes">
<hr />
<ol>
<li id="fn1"><p>After moving from an undergraduate in physics, I did my best to stitch my graduate work for an evolution and ecology PhD under a unifying theme in my dissertation, but it could hardly have been more across the map. My dissertation had three papers: one in Adaptive Dynamics, one in Comparative Phylogenetics, and one in warning signals of ecological shifts – there was no overlap in authors and essentially no overlap in the topic or reference lists (Okay, Cox 1968, in the latter two). I didn’t even attempt to weave in my two ecoinformatics papers. My undergraduate was similarly varied. I spent summer 2004 in an astrophysics lab (lambda CDM models with Neta Bahcall), 2005 sequencing extremeophilic microbes in a geomicrobiology lab (with TC Onstott), 2006 doing computational methods for boundary value problems during a plasma physics fellowship (with Hersh Rabitz). Though a physics major I wrote my first JP on population dynamics under clonal interference (David Huse), my second on adaptive dynamics (with Joshua Weitz), and my thesis on forest canopy models (with Stephen Pacala). It’s all a logical path of closely connected ideas in my mind of course ;-)<a href="#fnref1">↩</a></p></li>
</ol>
</section>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/06/NSF-math-bio-postdoc.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/09/02/reml-milestone.html">Reml Milestone</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 02 Sep 2013</p>

<p style="font-style:italic"> pageviews: 11 </p>

<article>
<div class="excerpt">
<p>Reached a <a href="https://github.com/ropensci/reml/issues/milestones?state=closed&amp;with_issues=yes">milestone</a> in <a href="https://github.com/ropensci/reml">reml</a> development today with basic implementation of reading, writing, and publishing valid EML files and a working test suite.</p>
<h3 id="an-s4-based-structure">An S4-based structure</h3>
<p>This involved a complete re-write of the pre-ESA code and is now completely S4-class based, allowing a more flexible structure that is easier and faster to extend. For instance, rather than having to explicitly think about the commands to parse and serialize a new metadata element that needs to be added from the schema, we just define the class for the object. Then we can leverage automated methods that can read into this class from XML and write out of it.</p>
<p>For instance, we recently added the class <code>alternativeTimeScale</code> following it’s <a href="knb.ecoinformatics.org/software/eml/eml-2.1.1/eml-coverage.html#alternativeTimeScale">definition in the Schema documentation</a>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;alternativeTimeScale&quot;</span>,
         <span class="kw">representation</span>(<span class="dt">timeScaleName =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeEstimate =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeUncertainty =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeExplanation =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleCitation =</span> <span class="st">&quot;character&quot;</span>) 
                        <span class="co"># FIXME should be citation type</span>
        )
<span class="kw">setAs</span>(<span class="st">&quot;XMLInternalElementNode&quot;</span>, <span class="st">&quot;alternativeTimeScale&quot;</span>,   function(from) <span class="kw">emlToS4</span>(from))
<span class="kw">setAs</span>(<span class="st">&quot;alternativeTimeScale&quot;</span>, <span class="st">&quot;XMLInternalElementNode&quot;</span>,   function(from) <span class="kw">S4Toeml</span>(from))</code></pre>
<p>Rather than bother with the XML parsing and serializing, we simply point R to use our predefined S4 coercion methods, specified after the class definition.</p>
<p>Note this need a little more work as the FIXME note indicates, since we have not yet written the <code>citation</code> class. Luckily, using the S4 approach this will be easy to update once that class is defined. In this manner, we can easily add elements that we need.</p>
<p>Classes are defined in groups in files named after the class. Constructor functions are defined in files named after the functions themselves, which prefix <code>eml_</code> to the class name. This adds some clarity (e.g. calling a function <code>eml_dataset</code> vs <code>dataset</code>) can also reduce namespace collisions.</p>
<p><strong>Note</strong> that the S4 code is intended only to be developer-facing. Users interact with documented constructor functions, which have more end-user documentation (though still in development).</p>
<h2 id="examples-of-reading-writing-and-publishing-eml">Examples of reading, writing, and publishing EML</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;devtools&quot;</span>)
<span class="kw">install_github</span>(<span class="st">&quot;reml&quot;</span>, <span class="st">&quot;ropensci&quot;</span>)</code></pre>
<p>Load the package:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;reml&quot;</span>)</code></pre>
<h3 id="writing-r-data-into-eml">Writing R data into EML</h3>
<p>Consider some dataset as an R <code>data.frame</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">dat = <span class="kw">data.frame</span>(<span class="dt">river =</span> <span class="kw">c</span>(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>),
                 <span class="dt">spp   =</span> <span class="kw">c</span>(<span class="st">&quot;king&quot;</span>, <span class="st">&quot;king&quot;</span>, <span class="st">&quot;ccho&quot;</span>),
                 <span class="dt">stg   =</span> <span class="kw">c</span>(<span class="st">&quot;smolt&quot;</span>, <span class="st">&quot;parr&quot;</span>, <span class="st">&quot;smolt&quot;</span>),
                 <span class="dt">ct    =</span>  <span class="kw">c</span>(293L, 410L, 210L))</code></pre>
<p>Provide a list giving the <strong>column header</strong> used in the data.frame, followed by a plain text definitions for the column, followed by a character vector giving the definition for the units:</p>
<p><strong>column descriptions</strong>: These are usually just plain text definitions, though a URI to a semantic definition can be particularly powerful. See “Advanced Use” for details on adding richer information, such as the method used to collect the data or set the geographic, taxonomic, or temporal coverage of an individual column.</p>
<p><strong>column units</strong>: For factors, this is a definition of the levels involved. For numeric data, specify the units from <a href="http://knb.ecoinformatics.org/software/eml/eml-2.1.1/eml-unitTypeDefinitions.html#StandardUnitDictionary">this list</a>. For dates, specify the format, (e.g. YYYY or MM-DD-YY). For character strings, a definition of the kind of string can be given, (e.g. species scientific name), otherwise the column description will be used.</p>
<pre class="sourceCode r"><code class="sourceCode r">metadata &lt;- 
  <span class="kw">list</span>(<span class="st">&quot;river&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;river&quot;</span>,
                      <span class="st">&quot;River site used for collection&quot;</span>,
                      <span class="kw">c</span>(<span class="dt">SAC =</span> <span class="st">&quot;The Sacramento River&quot;</span>, 
                        <span class="dt">AM =</span> <span class="st">&quot;The American River&quot;</span>)),
       <span class="st">&quot;spp&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;spp&quot;</span>,
                    <span class="st">&quot;Species common name&quot;</span>, 
                    <span class="kw">c</span>(<span class="dt">king =</span> <span class="st">&quot;King Salmon&quot;</span>, 
                      <span class="dt">ccho =</span> <span class="st">&quot;Coho Salmon&quot;</span>)),
       <span class="st">&quot;stg&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;stg&quot;</span>,
                    <span class="st">&quot;Life Stage&quot;</span>, 
                    <span class="kw">c</span>(<span class="dt">parr =</span> <span class="st">&quot;third life stage&quot;</span>, 
                      <span class="dt">smolt =</span> <span class="st">&quot;fourth life stage&quot;</span>)),
       <span class="st">&quot;ct&quot;</span>  = <span class="kw">list</span>(<span class="st">&quot;ct&quot;</span>,
                    <span class="st">&quot;count&quot;</span>, 
                    <span class="st">&quot;number&quot;</span>))</code></pre>
<p><em>for convenience, this metadata list can instead be constructed with the help of the reml wizard</em>. Just use <code>metadata &lt;- metadata_wizard(dat)</code> to begin. While this may be helpful starting out, regular users will find it faster to define the columns and units directly in the format above.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">eml_write</span>(dat, metadata, <span class="dt">title =</span> <span class="st">&quot;reml example&quot;</span>,  
          <span class="dt">description =</span> <span class="st">&quot;An example, intended for</span>
<span class="st">                              illustrative purposes only.&quot;</span>,
          <span class="dt">creator =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@gmail.com&gt;&quot;</span>,
          <span class="dt">file =</span> <span class="st">&quot;reml_example.xml&quot;</span>)</code></pre>
<pre><code>[1] &quot;reml_example.xml&quot;</code></pre>
<p>See the <a href="https://github.com/ropensci/reml/tree/master/inst/doc/reml_example.xml">EML generated</a> by this example. Note that if the <code>metadata</code> argument providing the definitions for the <code>data.frame</code> is not specified, <code>reml</code> will launch the wizard to prompt the user.</p>
<h5 id="configuration">Configuration</h5>
<p>Configure general metadata you may want to frequently reuse, avoiding having to specify things like the name of the creator or contact.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">eml_config</span>(<span class="dt">creator =</span> <span class="kw">list</span>(<span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>, <span class="st">&quot;Karthik Ram&quot;</span>), <span class="dt">contact =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>)</code></pre>
<ul>
<li>See <a href="https://github.com/ropensci/reml/issues?labels=write+eml&amp;page=1&amp;state=open">Open issues in writing EML</a></li>
</ul>
<h3 id="publish-eml">Publish EML</h3>
<p>Once we have generated a basic EML file, we can publish it to a variety of academic repositories where it will be indexed, permenently archived, more easily citable, and publicly discoverable.</p>
<p>In this example, we will publish the above EML to <a href="http://figshare.com">figshare</a> using the <a href="https://github.com/ropensci/rfigshare">rfigshare</a> package. Like most academic repostories, figshare requires some basic metadata for any entry. Much of this we can automatically extract from our existing EML file. Additional metadata unique to figshare (e.g. matching it’s allowed “categories” can also be provided.</p>
<p><em>Note: this example assumes <code>rfigshare</code> is installed, even though it is not required to use reml and run the examples in the other sections</em></p>
<pre class="sourceCode r"><code class="sourceCode r">id = <span class="kw">eml_publish</span>(<span class="st">&quot;reml_example.xml&quot;</span>, <span class="dt">description=</span><span class="st">&quot;Example EML file from reml&quot;</span>, <span class="dt">categories =</span> <span class="st">&quot;Ecology&quot;</span>, <span class="dt">tags =</span> <span class="st">&quot;EML&quot;</span>, <span class="dt">destination=</span><span class="st">&quot;figshare&quot;</span>)</code></pre>
<p>This creates a draft file visible only to the user configured in <code>rfigshare</code>. The document can be made (permanently) public using either the figshare web interface, the <code>rfigshare</code> package (using <code>fs_make_public(id)</code>) or just by adding the argument <code>visibility = TRUE</code> to the above <code>eml_publish</code> command.</p>
<h3 id="reading-eml">Reading EML</h3>
<pre class="sourceCode r"><code class="sourceCode r">obj &lt;- <span class="kw">eml_read</span>(<span class="st">&quot;reml_example.xml&quot;</span>)</code></pre>
<p>We can use various accessor functions to return the data and metadata elements in native R formats.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dataTable</span>(obj)</code></pre>
<pre><code>  river  spp   stg  ct
1   SAC king smolt 293
2   SAC king  parr 410
3    AM ccho smolt 210</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">metadata &lt;- <span class="kw">attributeList</span>(obj)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">contact</span>(obj)</code></pre>
<pre><code>[1] &quot;Carl Boettiger &lt;cboettig@gmail.com&gt;&quot;</code></pre>
<p>Note that the contact has been coerced into R’s built-in ‘person’ object:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(<span class="kw">contact</span>(obj))</code></pre>
<pre><code>[1] &quot;person&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citationInfo</span>(obj)</code></pre>
<pre><code>Boettiger, Carl, 2013. reml example.</code></pre>
<p>Vignette + read and extract examples</p>
<h2 id="notes">Notes</h2>
<ul>
<li><p>Resolved the strange segfault errors that arise unpredictably in tests by avoiding use of <code>XMLInternalDocument</code> class. (Perhaps it is referencing the nodes after writing this and finding they have no address that causes this error?) Now segfault free.</p></li>
<li><p>Still need to fix external validator test (currently not run by <code>testthat::test_dir</code> by avoiding the <code>test_</code> prefix on the filename…) <strong>EDIT</strong> 2013-09-05 Duncan has <a href="">fixed</a> this, looking into having the form give more structured output.</p></li>
<li><p><strong>Edit</strong> 2013-09-04 some attribute definitions were not being written out correctly and have now been fixed. Still writing the <a href="">test cases</a>.</p></li>
</ul>
<h2 id="where-next">Where next?</h2>
<p>Lots more can be done on the package, as documented in our extensive <a href="https://github.com/ropensci/reml/issues">issues</a> list and the discussions there-in.</p>
<p>Further development will be driven around identifying and illustrating some particularly motivating use-cases beyond the basic creation and reading in of EML files, adding what additional functionality and debugging is required along the way.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/02/reml-milestone.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/29/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 29 Aug 2013</p>

<p style="font-style:italic"> pageviews: 11 </p>

<article>
<div class="excerpt">
<p>Digging through some old stuff in the course of correspondance on related issues, I came across a few outstanding puzzles I’ve been meaning to wrap up one day:</p>
<ul>
<li><a href="http://carlboettiger.info/2009/08/25/drift-effects.html">Drift effecs – some difficulties</a></li>
<li><a href="http://carlboettiger.info/2010/04/20/waiting-time.html">Waiting Time</a></li>
</ul>
<p>and some actual results that could deserve some renewed attention too.</p>
<ul>
<li><a href="http://carlboettiger.info/2009/07/28/branching-times.html">Phase 1</a></li>
<li><a href="http://carlboettiger.info/2011/01/06/two-routes-to-adaptive-branching.html">Routes to branching</a></li>
</ul>
<p>Also raises an interesting new question that could be addressed in this context: can we formulate a similar approach to the stochastic escape from an adaptive peak?</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/29/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/28/rnexml-coding.html">Rnexml Coding</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 28 Aug 2013</p>

<p style="font-style:italic"> pageviews: 4 </p>

<article>
<div class="excerpt">
<ul>
<li>test_ape.R unit tests now passing. Should still revise the coercion methods in nexml_write.R to something more clever. We aren’t coercing a tree to a nexml, we are building a nexml from a tree. Likewise handling the otu assignments of species names should be done directly <a href="https://github.com/ropensci/RNeXML/commit/8a182d4cc6fb75099067142327788be8d51eef6e">11:20 am 2013/08/28</a></li>
<li>start cleaning up old read/write methods obsoleted by toNeXML/fromNeXML methods <a href="https://github.com/ropensci/RNeXML/commit/ffb7342f1d28b351984685a83e11d7df488696e8">10:59 am 2013/08/28</a></li>
<li>inherited classes now replace all the original classes <a href="https://github.com/ropensci/RNeXML/commit/004d31fc2c87a4cf2dba45ab08f1b8a663ab5258">10:54 am 2013/08/28</a></li>
<li>All inheritance-based classes appearing in inst/examples/trees.xml now implemented <a href="https://github.com/ropensci/RNeXML/commit/d9e649ac64b697d4ea982a1bb78b403400984392">10:53 am 2013/08/28</a></li>
<li><p>added classes for otu, otus, tree, trees, nexml, edge, rootEdge <a href="https://github.com/ropensci/RNeXML/commit/b9543f64352759f4dbb1f77a170cced78280c9a7">12:04 am 2013/08/28</a></p></li>
<li><p>Working on #1 : What works and what doesn’t: The example in inst/tests/inheritance.R should run using the class definitions in R/inherited_classes.R. Other class definitions are obsoleted by these updates and should eventually be removed. The ape methods should be unaffected, since the S4 classes have the same structure as before, only defined through inheritance instead.</p></li>
</ul>
<p>ToDo: - Remove the obsoleted methods in classes.R, nexml_read.R and nexml_write.R - Check the ape.R unit tests, update other unit tests - Write remaining essential classes out by hand based on inheritance and schema definitions: e.g. edge, otu, otus, tree, trees, nexml <a href="https://github.com/ropensci/RNeXML/commit/fd2e6b0f8f31823c4c65916a20fe339fb2038646">04:26 pm 2013/08/27</a> - Begun replacing class definitions with inheritance-based classes taking a brute-force approach in defining conversion methods. Conversion methods between XML and S4 explicitly assign attributes and values to S4 slots, rather than using the clever XML package functions to guess what goes where (since that approach cannot include a call to callNextMethod, which allows us to recursively go down the inheritance classes. The up side of this approach is that you don’t have to think much to write methods, it’s merely verbose.) <a href="https://github.com/ropensci/RNeXML/commit/8b45dee59848702c5aea6bf23ea06802b462d55d">04:11 pm 2013/08/27</a></p>
<ul>
<li>fixed ape tests, extended setAs(‘nexml’, ‘phylo’) to handle the case of multiple trees <a href="https://github.com/ropensci/RNeXML/commit/4fa252e20c7ee44a93573d3eb2c5dd3e781103c3">04:27 pm 2013/08/28</a></li>
<li>Now use otu labels as species names in both directions, closes #8 See updated tests in inst/test/test_ape.R. Note that coercing phylo-&gt;tree is still permitted, since resulting S4 tree is valid (though really intended as internal), but coercing tree-&gt;phylo is no longer available as a separate routine, since this is lossy as the tip.labels are not stored in tree but in nexml@otu <a href="https://github.com/ropensci/RNeXML/commit/abf298614f587d7eb99c67a0acd15d731690df54">03:54 pm 2013/08/28</a></li>
<li>updated nexml-&gt;ape conversion to get tip.labels from otus instead of just using the otu attribute on the node, which simply corresponds to the id of the otu element in otu. the taxon (species name) is in the label of the matching element instead. Still needs some testing and a unit test to confirm the label is correct. This addresses 1/2 of #8.</li>
</ul>
<p>Still need to handle the case of coercing ape-&gt;nexml, in which a method generates the otus node from the tip.label list and matches the id to node’s otu attribute. This means that ape-&gt;tree doesn’t look at tip.label, but just generates locally unique keys for the node otu attribute <a href="https://github.com/ropensci/RNeXML/commit/b97df1abac843b33214109d95c5baaa740dbe83e">01:40 pm 2013/08/28</a></p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/28/rnexml-coding.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

