<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-07-31T22:00:31-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-07-31T22:00:31-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-07-31T22:00:31-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
    <div class="excerpt">
      <div class="scroll">
        <ul><li>cboettig pushed to master at cboettig/labnotebook: <em>cleanup new post</em> <a href="https://github.com/cboettig/labnotebook/compare/2440bdd284...75ec57927d">07:35 2013/09/08</a></li><li>cboettig commented on issue ropensci/RNeXML#15: <em>@SChamberlain Thanks for writing the extra test cases. Would you be interested in running some testing of trying to parse a bunch of nexml files (e…</em> <a href="https://github.com/ropensci/RNeXML/issues/15#issuecomment-24026708">06:48 2013/09/08</a></li><li>cboettig opened issue ropensci/RNeXML#22: <em>Reach out to R and NeXML lists</em> <a href="https://github.com/ropensci/RNeXML/issues/22">06:42 2013/09/08</a></li><li>cboettig pushed to master at cboettig/nonparametric-bayes: <em>Merge branch 'master' of github.com:cboettig/nonparametric-bayes update chunk options</em> <a href="https://github.com/cboettig/nonparametric-bayes/compare/e6008044f7...08cbafacc3">06:26 2013/09/08</a></li><li>cboettig created branch devel  at ropensci/RNeXML: <em></em> <a href="https://github.com/ropensci/RNeXML/compare/devel">06:24 2013/09/08</a></li></ul>
      </div>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
      <div class="scroll">
       <ul><li><p>RT @brembs: The nicest decline-to-review ever: <a href="http://t.co/8FkSpW3AlZ">http://t.co/8FkSpW3AlZ</a></p>
 <a href="http://twitter.com/cboettig/statuses/376774657963085824">11:30 2013/09/08</a> </li><li><p>RT @sharmanedit: Perspective in @PLOSBiology on release of phylogenetic data, via @catmacOA @RoliRoberts. Journals should police more. http…</p>
 <a href="http://twitter.com/cboettig/statuses/376350029817982977">07:23 2013/09/07</a> </li><li><p>@ucfagls @DistribEcology @recology_ sounds like you want something like uuid + <a href="http://t.co/DY6FJdTtsN">http://t.co/DY6FJdTtsN</a> maybe?</p>
 <a href="http://twitter.com/cboettig/statuses/375711796172034048">01:07 2013/09/05</a> </li><li><p>check out this survey from @dataoneorg&quot;A Study of the Motivations and Opinions of @figshare users&quot; <a href="http://t.co/0hN3K7wiCh">http://t.co/0hN3K7wiCh</a></p>
 <a href="http://twitter.com/cboettig/statuses/375697200296099840">12:09 2013/09/05</a> </li><li><p>Prose: looks like a very nice markdown editor for Github and Jekyll-based sites... <a href="http://t.co/zPi3sZeqVF">http://t.co/zPi3sZeqVF</a></p>
 <a href="http://twitter.com/cboettig/statuses/375697150761398272">12:09 2013/09/05</a> </li></ul>
      </div>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <div class="scroll">
      <ul>
<li>Stephan B Munch, A Kottas. 2009. A Bayesian modeling approach for determining productivity regimes and their characteristics <em>Ecological Applications</em> <strong>19</strong> 2 <a href="http://dx.doi.org/10.1890/07-2116.1">10.1890/07-2116.1</a> </li>
<li>Mark John Costello, Marta Coll, Roberto Danovaro et al. 2010. A Census of Marine Biodiversity Knowledge, Resources, and Future Challenges <em>PLoS ONE</em> <strong>5</strong> 8 <a href="http://dx.doi.org/10.1371/journal.pone.0012110">10.1371/journal.pone.0012110</a> </li>
<li>R. F. Hechinger, K. D. Lafferty, a. P. Dobson et al. 2011. A Common Scaling Rule for Abundance, Energetics, and Production of Parasitic and Free-Living Species <em>Science</em> <strong>333</strong> 6041 <a href="http://dx.doi.org/10.1126/science.1204337">10.1126/science.1204337</a> </li>
<li>Jonathan B. Shurin, Elizabeth T. Borer, Eric W. Seabloom et al. 2002. A cross-ecosystem comparison of the strength of trophic cascades <em>Ecology Letters</em> <strong>5</strong> 6 <a href="http://dx.doi.org/10.1046/j.1461-0248.2002.00381.x">10.1046/j.1461-0248.2002.00381.x</a> </li>
<li>Wayne M Getz, David Saltz. 2008. A framework for generating and analyzing movement paths on ecological landscapes. <em>Proceedings of the National Academy of Sciences of the United States of America</em> <strong>105</strong> 49 <a href="http://dx.doi.org/10.1073/pnas.0801732105">10.1073/pnas.0801732105</a> </li>

      </div>
    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/09/08/some-rfishbase-updates.html">Some Rfishbase Updates</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 08 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>Ever since writing the <code>rfishbase</code> package I’ve been frustrated by the lack of a proper API from Fishbase.org that would allow access to that wealth of information that is not exposed in the “Summary XML” files from which <code>rfishbase</code> extracts its content. Since writing the paper in Journal of Fish Biology, I continue to receive user requests asking if or how such-and-such a data element might be extracted. Usually this involves a disappointing boiler-plate reply about how the package is limited to the “Summary XML” data and that improved query potential must wait for a proper API from Fishbase.org at some unspecified time.</p>
<p>I have been in touch with many of the folks at Fishbase.org at various times looking for ways to address these issues. They have been in general keenly supportive of the idea, but I gather that they are largely overworked and under-resourced and as such, implementation of any of these steps has been very slow. Last October another rfishbase user contacted me, who works full time as a software developer. I put him in touch with the UBC team, and he is willing to volunteer the hours and expertise the FishBase development team lacks to help transform the database into a full API. I think this is a very promising development and a nice example of a community effort, but still faces the same fundamental challenges in getting things rolling. So, so far, no <code>rfishbase</code> version 2. (Actually in our semantic versioning it would be the less romantic version 0.3 or 0.4 or so, depending on how many CRAN submits we do before then and provided we don’t break any backwards compatibility)…</p>
<p>One of the more frequent requests involves accessing quantitative trophic level estimates provided for many species on the “Ecology” page, linked from the “Summary page”. The only way to access this data is by scraping the HTML.</p>
<p>Previously I’ve been reluctant to scrape the HTML directly for several reasons:</p>
<ul>
<li><p>It is generally unrealiable/unverifiable. Without any metadata (rdfa or microdata/microformats could have helped…) in the HTML to signpost that we have the correct value, we are generally counting arbitrary header, div and span elements, followed by some Regular Expression matching, to find the element we want.<br />This is error-prone and easily broken by any changes in the html pages either over time or between pages for different species.</p></li>
<li><p>Second, it puts a much greater load on the fishbase servers. Often we cannot even start by querying the page we want, (e.g. since URLs involve things like stock-ids that the end user will not know and I haven’t located a look-up table), so a single call may follow links through several pages before parsing the data it needs. Repeating such queries over many fish programmatically can create a rather substantial burst of traffic that could slow or crash the servers.</p></li>
<li><p>It’s slow. Making all the queries takes time, even for a computer. (Though caching, as we already do in rfishbase, could take care of much this).</p></li>
<li><p>It is potentially time consuming and annoying to code the scaping.</p></li>
</ul>
<p>Okay, the last is just me being lazy. Since this particular query about trophic level was a common email request, and as the data was available from the “Ecology” page of a species in the form of an HTML Table, (and hence somewhat less fragile to robustly identify…), I thought I’d give it a try.</p>
<p>Here’s my approach (or see <a href="">source code</a>). We would follow the same query structure as <code>getSize</code> did <sup><a href="#fn1" class="footnoteRef" id="fnref1">1</a></sup>, and execute several steps:</p>
<ol type="1">
<li>Visit the Summary page of the requested fish</li>
<li>Find it’s Ecology page</li>
<li>Find the Trophic level table</li>
<li>parse the table.</li>
</ol>
<p>In code, that looks like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">getTrophicLevel &lt;- function(<span class="dt">fish.data =</span> <span class="ot">NULL</span>,
                            <span class="dt">path =</span> <span class="ot">NULL</span>,
                            <span class="dt">as_table=</span><span class="ot">FALSE</span>, 
                            <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;diet composition&quot;</span>, <span class="st">&quot;individual food items&quot;</span>), 
                            <span class="dt">unfished =</span> <span class="ot">FALSE</span>,
                            <span class="dt">justSE =</span> <span class="ot">FALSE</span>,
                            <span class="dt">justReference =</span> <span class="ot">FALSE</span>){

  ids &lt;- <span class="kw">getIds</span>(<span class="dt">fish.data =</span> fish.data, <span class="dt">path=</span>path)
  out &lt;- <span class="kw">sapply</span>(ids, function(id){
    summaryPage &lt;- <span class="kw">getSummary</span>(id)
    ecologyPage &lt;- <span class="kw">getEcology</span>(summaryPage)
    tab &lt;- <span class="kw">readTrophicLevel</span>(ecologyPage)
    if(as_table)
      tab
    else
      <span class="kw">parseTrophicLevel</span>(tab, <span class="dt">from=</span>from, <span class="dt">unfished=</span>unfished,<span class="dt">justSE=</span>justSE, <span class="dt">justReference=</span>justReference)
    })
  out
}</code></pre>
<p>Which has a whole bunch of options just to let the user control the format of the output / target value, since there are different numbers.</p>
<p>Each of the steps corresponds to it’s own function:</p>
<pre class="sourceCode r"><code class="sourceCode r">getIds &lt;- function(<span class="dt">fish.data=</span><span class="ot">NULL</span>, <span class="dt">path=</span><span class="ot">NULL</span>){
  if(<span class="kw">is.null</span>(fish.data))
    fish.data &lt;- <span class="kw">loadCache</span>(<span class="dt">path=</span>path)
  ids &lt;- <span class="kw">sapply</span>(fish.data, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&#39;id&#39;</span>)
  species.names &lt;- <span class="kw">sapply</span>(fish.data, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&#39;ScientificName&#39;</span>)
  <span class="kw">names</span>(ids) &lt;- <span class="kw">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, species.names) <span class="co"># use underscores instead of spaces</span>
  ids
}


getSummary &lt;- function(id){ 
  <span class="co"># Grab and parse page matching id</span>
  url &lt;- <span class="kw">paste0</span>(<span class="st">&quot;http://www.fishbase.org/summary/speciessummary.php?id=&quot;</span>, id)
  summaryPage &lt;- <span class="kw">htmlParse</span>(url) 
}

getEcology &lt;- function(summaryPage){
  link &lt;- <span class="kw">xpathApply</span>(summaryPage, <span class="st">&quot;//*[contains(@href, &#39;/Ecology/FishEcologySummary.php&#39;)][1]&quot;</span>, xmlAttrs)[[<span class="dv">1</span>]][[<span class="st">&quot;href&quot;</span>]]
  ecologyPage &lt;- <span class="kw">htmlParse</span>(<span class="kw">paste0</span>(<span class="st">&quot;http://www.fishbase.org/&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">.</span><span class="ch">\\</span><span class="st">./&quot;</span>, <span class="st">&quot;&quot;</span>, link)))
}


readTrophicLevel &lt;- function(ecologyPage){ 
  tab &lt;- <span class="kw">readHTMLTable</span>(ecologyPage)[[<span class="dv">6</span>]]
}</code></pre>
<p>which are all nice and short and potentially self-explanatory. Extracting values from the table we isolate is done rather crudely at the moment and could no doubt be improved, but this is what I came up with:</p>
<pre class="sourceCode r"><code class="sourceCode r">parseTrophicLevel &lt;- function(tab, 
                              <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;diet composition&quot;</span>, <span class="st">&quot;individual food items&quot;</span>), 
                              <span class="dt">unfished =</span> <span class="ot">FALSE</span>,
                              <span class="dt">justSE =</span> <span class="ot">FALSE</span>,
                              <span class="dt">justReference =</span> <span class="ot">FALSE</span>){
  from &lt;- <span class="kw">match.arg</span>(from)
  if(justReference)
    out &lt;- <span class="kw">as.character</span>(tab[<span class="dv">3</span>,<span class="dv">2</span>])
  else{
    col &lt;- <span class="dv">2</span>
    adj &lt;- <span class="dv">0</span>
    if(unfished)
      col &lt;- <span class="dv">4</span>
    if(justSE)
      adj &lt;- <span class="dv">1</span>
    if(from == <span class="st">&quot;diet composition&quot;</span>)
      out &lt;- <span class="kw">as.numeric</span>(<span class="kw">as.character</span>(tab[<span class="dv">2</span>,col+adj]))
    else if(from == <span class="st">&quot;individual food items&quot;</span>)
      out &lt;- <span class="kw">as.numeric</span>(<span class="kw">as.character</span>(tab[<span class="dv">4</span>,col+adj]))
  }
  out 
}</code></pre>
<p>Most of these are internal functions of course, only the top <code>getTrophicLevel</code> is user-facing.<br />The use is straight forward:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(rfishbase)
<span class="kw">data</span>(fishbase)

<span class="kw">getTrophicLevel</span>(fish.data[<span class="dv">1</span>:<span class="dv">10</span>])</code></pre>
<p>or any of the various optional configurations, e.g.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getTrophicLevel</span>(fish.data[<span class="dv">1</span>:<span class="dv">10</span>], <span class="dt">as_table=</span><span class="ot">TRUE</span>)</code></pre>
<p>(which gives a list of tables matching these fish).</p>
<p>The astute reader might notice that with <code>EcologyPage</code> etc we have access to a lot more (tabular) data in this approach, which could be extended beyond that page as well, following the same template shown here. Just a matter of time and the other limitations discussed above.</p>
<p>So anyway, if you need the getTrophicLevel, it’s available now on the <a href="https://github.com/ropensci/rfishbase/issues/12">Github version</a> and should be on CRAN in the near future. Meanwhile, I’d certainly welcome other requests (or better yet, forks and pull requests implementing additional data fields along these same lines).</p>
<section class="footnotes">
<hr />
<ol>
<li id="fn1"><p>taking the <code>fish.data</code> list as an argument – potentially a weird choice but oh well, that’s what I did when I first wrote the rfishbase API so may as well keep it. A more natural choice taking scientific names as arguments directly, instead of needing to start with a which_fish query, should be added one day to the entire package API.<a href="#fnref1">↩</a></p></li>
</ol>
</section>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/08/some-rfishbase-updates.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/07/RNeXML-milestone.html">Rnexml Milestone</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 07 Sep 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>We have also recently reached the first <a href="https://github.com/ropensci/RNeXML/issues?milestone=1&amp;state=closed">milestone</a> in <a href="https://github.com/ropensci/RNeXML">RNeXML</a>. We can now read NeXML files into R and convert to the popular <code>phylo</code> format from the <code>ape</code> package, as well as writing the <code>ape::phylo</code> trees out into valid NeXML. We can also add basic metadata such as author, license, title and description to the file.</p>
<h3 id="development-notes">Development notes</h3>
<p>At the heart of this package is an internal S4 representation of the schema. While knowledgable users could interact with the S4 objects directly (easier and less error-prone than interacting with the XML parsing/serializing directly), we in general provide an end-user API which hides the internal complexity of S4 classes, as illustrated below. In principle this representation could be automatically generated from the schema using the <code>XMLSchema</code> package, but that must wait until that package is more mature. Meanwhile, we rely on hand-coded classes and hand-coded coercion functions.</p>
<p>The coercion functions take care of the translation into and out of the XML (parsing and serializing) without any need for xpath expressions. At the moment they are verbosely spelled out under the class definitions in classes.R file, though ideally they could be abstracted (e.g. as we do in <code>reml</code>). Classes are based on the <a href="https://github.com/ropensci/RNeXML/issues/1#issuecomment-22867535">inheritance structure</a> of the schema (unlike <code>reml</code>), which makes additional nodes much easier to add and hence it is easier to extend this library. At this time we have only certain core elements from the schema as map naturally to <code>ape</code> trees, along with the essential annotation elements (<code>meta</code> nodes). Support for additional classes / schema elements will be added as driven by use cases.</p>
<p>(See the <a href="https://github.com/ropensci/RNeXML">Github page</a> for the most recent tutorial, this is just added here for my record).</p>
<h3 id="tutorial">Tutorial</h3>
<p>The development version of RNeXML is <a href="https://github.com/ropensci/RNeXML">available on Github</a>. With the <code>devtools</code> package installed on your system, RNeXML can be installed using:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;RNeXML&quot;</span>, <span class="st">&quot;ropensci&quot;</span>)
<span class="kw">library</span>(RNeXML)</code></pre>
<p>Read in a <code>nexml</code> file into the <code>ape::phylo</code> format:</p>
<pre class="sourceCode r"><code class="sourceCode r">f &lt;- <span class="kw">system.file</span>(<span class="st">&quot;examples&quot;</span>, <span class="st">&quot;trees.xml&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;RNeXML&quot;</span>)
tr &lt;- <span class="kw">nexml_read</span>(f, <span class="st">&quot;phylo&quot;</span>)</code></pre>
<p>Write an <code>ape::phylo</code> tree into the <code>nexml</code> format, including optional metadata:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">nexml_write</span>(bird.orders, <span class="st">&quot;birds.xml&quot;</span>, <span class="dt">creator =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>, 
    <span class="dt">title =</span> <span class="st">&quot;example NeXML file for bird orders&quot;</span>, <span class="dt">description =</span> <span class="st">&quot;Example phylogeny taken from the ape documentation showing the major bird orders&quot;</span>)</code></pre>
<h3 id="next-steps">Next Steps</h3>
<p>Developing use cases, soliciting input, and writing a manuscript… See issues log.</p>
<ul>
<li>Outreach <a href="https://github.com/ropensci/RNeXML/issues/22">#22</a></li>
<li>metadata use cases <a href="https://github.com/ropensci/RNeXML/issues/21">#21</a></li>
<li>serializing metadata (implementation) <a href="https://github.com/ropensci/RNeXML/issues/18">#18</a></li>
<li>Testing <a href="https://github.com/ropensci/RNeXML/issues/15">#15</a></li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/07/RNeXML-milestone.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/09/02/reml-milestone.html">Reml Milestone</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 02 Sep 2013</p>

<p style="font-style:italic"> pageviews: 4 </p>

<article>
<div class="excerpt">
<p>Reached a <a href="https://github.com/ropensci/reml/issues/milestones?state=closed&amp;with_issues=yes">milestone</a> in <a href="https://github.com/ropensci/reml">reml</a> development today with basic implementation of reading, writing, and publishing valid EML files and a working test suite.</p>
<h3 id="an-s4-based-structure">An S4-based structure</h3>
<p>This involved a complete re-write of the pre-ESA code and is now completely S4-class based, allowing a more flexible structure that is easier and faster to extend. For instance, rather than having to explicitly think about the commands to parse and serialize a new metadata element that needs to be added from the schema, we just define the class for the object. Then we can leverage automated methods that can read into this class from XML and write out of it.</p>
<p>For instance, we recently added the class <code>alternativeTimeScale</code> following it’s <a href="knb.ecoinformatics.org/software/eml/eml-2.1.1/eml-coverage.html#alternativeTimeScale">definition in the Schema documentation</a>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;alternativeTimeScale&quot;</span>,
         <span class="kw">representation</span>(<span class="dt">timeScaleName =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeEstimate =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeUncertainty =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleAgeExplanation =</span> <span class="st">&quot;character&quot;</span>,
                        <span class="dt">timeScaleCitation =</span> <span class="st">&quot;character&quot;</span>) 
                        <span class="co"># FIXME should be citation type</span>
        )
<span class="kw">setAs</span>(<span class="st">&quot;XMLInternalElementNode&quot;</span>, <span class="st">&quot;alternativeTimeScale&quot;</span>,   function(from) <span class="kw">emlToS4</span>(from))
<span class="kw">setAs</span>(<span class="st">&quot;alternativeTimeScale&quot;</span>, <span class="st">&quot;XMLInternalElementNode&quot;</span>,   function(from) <span class="kw">S4Toeml</span>(from))</code></pre>
<p>Rather than bother with the XML parsing and serializing, we simply point R to use our predefined S4 coercion methods, specified after the class definition.</p>
<p>Note this need a little more work as the FIXME note indicates, since we have not yet written the <code>citation</code> class. Luckily, using the S4 approach this will be easy to update once that class is defined. In this manner, we can easily add elements that we need.</p>
<p>Classes are defined in groups in files named after the class. Constructor functions are defined in files named after the functions themselves, which prefix <code>eml_</code> to the class name. This adds some clarity (e.g. calling a function <code>eml_dataset</code> vs <code>dataset</code>) can also reduce namespace collisions.</p>
<p><strong>Note</strong> that the S4 code is intended only to be developer-facing. Users interact with documented constructor functions, which have more end-user documentation (though still in development).</p>
<h2 id="examples-of-reading-writing-and-publishing-eml">Examples of reading, writing, and publishing EML</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;devtools&quot;</span>)
<span class="kw">install_github</span>(<span class="st">&quot;reml&quot;</span>, <span class="st">&quot;ropensci&quot;</span>)</code></pre>
<p>Load the package:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;reml&quot;</span>)</code></pre>
<h3 id="writing-r-data-into-eml">Writing R data into EML</h3>
<p>Consider some dataset as an R <code>data.frame</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">dat = <span class="kw">data.frame</span>(<span class="dt">river =</span> <span class="kw">c</span>(<span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;SAC&quot;</span>, <span class="st">&quot;AM&quot;</span>),
                 <span class="dt">spp   =</span> <span class="kw">c</span>(<span class="st">&quot;king&quot;</span>, <span class="st">&quot;king&quot;</span>, <span class="st">&quot;ccho&quot;</span>),
                 <span class="dt">stg   =</span> <span class="kw">c</span>(<span class="st">&quot;smolt&quot;</span>, <span class="st">&quot;parr&quot;</span>, <span class="st">&quot;smolt&quot;</span>),
                 <span class="dt">ct    =</span>  <span class="kw">c</span>(293L, 410L, 210L))</code></pre>
<p>Provide a list giving the <strong>column header</strong> used in the data.frame, followed by a plain text definitions for the column, followed by a character vector giving the definition for the units:</p>
<p><strong>column descriptions</strong>: These are usually just plain text definitions, though a URI to a semantic definition can be particularly powerful. See “Advanced Use” for details on adding richer information, such as the method used to collect the data or set the geographic, taxonomic, or temporal coverage of an individual column.</p>
<p><strong>column units</strong>: For factors, this is a definition of the levels involved. For numeric data, specify the units from <a href="http://knb.ecoinformatics.org/software/eml/eml-2.1.1/eml-unitTypeDefinitions.html#StandardUnitDictionary">this list</a>. For dates, specify the format, (e.g. YYYY or MM-DD-YY). For character strings, a definition of the kind of string can be given, (e.g. species scientific name), otherwise the column description will be used.</p>
<pre class="sourceCode r"><code class="sourceCode r">metadata &lt;- 
  <span class="kw">list</span>(<span class="st">&quot;river&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;river&quot;</span>,
                      <span class="st">&quot;River site used for collection&quot;</span>,
                      <span class="kw">c</span>(<span class="dt">SAC =</span> <span class="st">&quot;The Sacramento River&quot;</span>, 
                        <span class="dt">AM =</span> <span class="st">&quot;The American River&quot;</span>)),
       <span class="st">&quot;spp&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;spp&quot;</span>,
                    <span class="st">&quot;Species common name&quot;</span>, 
                    <span class="kw">c</span>(<span class="dt">king =</span> <span class="st">&quot;King Salmon&quot;</span>, 
                      <span class="dt">ccho =</span> <span class="st">&quot;Coho Salmon&quot;</span>)),
       <span class="st">&quot;stg&quot;</span> = <span class="kw">list</span>(<span class="st">&quot;stg&quot;</span>,
                    <span class="st">&quot;Life Stage&quot;</span>, 
                    <span class="kw">c</span>(<span class="dt">parr =</span> <span class="st">&quot;third life stage&quot;</span>, 
                      <span class="dt">smolt =</span> <span class="st">&quot;fourth life stage&quot;</span>)),
       <span class="st">&quot;ct&quot;</span>  = <span class="kw">list</span>(<span class="st">&quot;ct&quot;</span>,
                    <span class="st">&quot;count&quot;</span>, 
                    <span class="st">&quot;number&quot;</span>))</code></pre>
<p><em>for convenience, this metadata list can instead be constructed with the help of the reml wizard</em>. Just use <code>metadata &lt;- metadata_wizard(dat)</code> to begin. While this may be helpful starting out, regular users will find it faster to define the columns and units directly in the format above.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">eml_write</span>(dat, metadata, <span class="dt">title =</span> <span class="st">&quot;reml example&quot;</span>,  
          <span class="dt">description =</span> <span class="st">&quot;An example, intended for</span>
<span class="st">                              illustrative purposes only.&quot;</span>,
          <span class="dt">creator =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@gmail.com&gt;&quot;</span>,
          <span class="dt">file =</span> <span class="st">&quot;reml_example.xml&quot;</span>)</code></pre>
<pre><code>[1] &quot;reml_example.xml&quot;</code></pre>
<p>See the <a href="https://github.com/ropensci/reml/tree/master/inst/doc/reml_example.xml">EML generated</a> by this example. Note that if the <code>metadata</code> argument providing the definitions for the <code>data.frame</code> is not specified, <code>reml</code> will launch the wizard to prompt the user.</p>
<h5 id="configuration">Configuration</h5>
<p>Configure general metadata you may want to frequently reuse, avoiding having to specify things like the name of the creator or contact.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">eml_config</span>(<span class="dt">creator =</span> <span class="kw">list</span>(<span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>, <span class="st">&quot;Karthik Ram&quot;</span>), <span class="dt">contact =</span> <span class="st">&quot;Carl Boettiger &lt;cboettig@ropensci.org&gt;&quot;</span>)</code></pre>
<ul>
<li>See <a href="https://github.com/ropensci/reml/issues?labels=write+eml&amp;page=1&amp;state=open">Open issues in writing EML</a></li>
</ul>
<h3 id="publish-eml">Publish EML</h3>
<p>Once we have generated a basic EML file, we can publish it to a variety of academic repositories where it will be indexed, permenently archived, more easily citable, and publicly discoverable.</p>
<p>In this example, we will publish the above EML to <a href="http://figshare.com">figshare</a> using the <a href="https://github.com/ropensci/rfigshare">rfigshare</a> package. Like most academic repostories, figshare requires some basic metadata for any entry. Much of this we can automatically extract from our existing EML file. Additional metadata unique to figshare (e.g. matching it’s allowed “categories” can also be provided.</p>
<p><em>Note: this example assumes <code>rfigshare</code> is installed, even though it is not required to use reml and run the examples in the other sections</em></p>
<pre class="sourceCode r"><code class="sourceCode r">id = <span class="kw">eml_publish</span>(<span class="st">&quot;reml_example.xml&quot;</span>, <span class="dt">description=</span><span class="st">&quot;Example EML file from reml&quot;</span>, <span class="dt">categories =</span> <span class="st">&quot;Ecology&quot;</span>, <span class="dt">tags =</span> <span class="st">&quot;EML&quot;</span>, <span class="dt">destination=</span><span class="st">&quot;figshare&quot;</span>)</code></pre>
<p>This creates a draft file visible only to the user configured in <code>rfigshare</code>. The document can be made (permanently) public using either the figshare web interface, the <code>rfigshare</code> package (using <code>fs_make_public(id)</code>) or just by adding the argument <code>visibility = TRUE</code> to the above <code>eml_publish</code> command.</p>
<h3 id="reading-eml">Reading EML</h3>
<pre class="sourceCode r"><code class="sourceCode r">obj &lt;- <span class="kw">eml_read</span>(<span class="st">&quot;reml_example.xml&quot;</span>)</code></pre>
<p>We can use various accessor functions to return the data and metadata elements in native R formats.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dataTable</span>(obj)</code></pre>
<pre><code>  river  spp   stg  ct
1   SAC king smolt 293
2   SAC king  parr 410
3    AM ccho smolt 210</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">metadata &lt;- <span class="kw">attributeList</span>(obj)</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">contact</span>(obj)</code></pre>
<pre><code>[1] &quot;Carl Boettiger &lt;cboettig@gmail.com&gt;&quot;</code></pre>
<p>Note that the contact has been coerced into R’s built-in ‘person’ object:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(<span class="kw">contact</span>(obj))</code></pre>
<pre><code>[1] &quot;person&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">citationInfo</span>(obj)</code></pre>
<pre><code>Boettiger, Carl, 2013. reml example.</code></pre>
<p>Vignette + read and extract examples</p>
<h2 id="notes">Notes</h2>
<ul>
<li><p>Resolved the strange segfault errors that arise unpredictably in tests by avoiding use of <code>XMLInternalDocument</code> class. (Perhaps it is referencing the nodes after writing this and finding they have no address that causes this error?) Now segfault free.</p></li>
<li><p>Still need to fix external validator test (currently not run by <code>testthat::test_dir</code> by avoiding the <code>test_</code> prefix on the filename…) <strong>EDIT</strong> 2013-09-05 Duncan has <a href="">fixed</a> this, looking into having the form give more structured output.</p></li>
<li><p><strong>Edit</strong> 2013-09-04 some attribute definitions were not being written out correctly and have now been fixed. Still writing the <a href="">test cases</a>.</p></li>
</ul>
<h2 id="where-next">Where next?</h2>
<p>Lots more can be done on the package, as documented in our extensive <a href="https://github.com/ropensci/reml/issues">issues</a> list and the discussions there-in.</p>
<p>Further development will be driven around identifying and illustrating some particularly motivating use-cases beyond the basic creation and reading in of EML files, adding what additional functionality and debugging is required along the way.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/09/02/reml-milestone.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/08/29/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 29 Aug 2013</p>

<p style="font-style:italic"> pageviews: 6 </p>

<article>
<div class="excerpt">
<p>Digging through some old stuff in the course of correspondance on related issues, I came across a few outstanding puzzles I’ve been meaning to wrap up one day:</p>
<ul>
<li><a href="http://carlboettiger.info/2009/08/25/drift-effects.html">Drift effecs – some difficulties</a></li>
<li><a href="http://carlboettiger.info/2010/04/20/waiting-time.html">Waiting Time</a></li>
</ul>
<p>and some actual results that could deserve some renewed attention too.</p>
<ul>
<li><a href="http://carlboettiger.info/2009/07/28/branching-times.html">Phase 1</a></li>
<li><a href="http://carlboettiger.info/2011/01/06/two-routes-to-adaptive-branching.html">Routes to branching</a></li>
</ul>
<p>Also raises an interesting new question that could be addressed in this context: can we formulate a similar approach to the stochastic escape from an adaptive peak?</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/29/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/28/rnexml-coding.html">Rnexml Coding</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 28 Aug 2013</p>

<p style="font-style:italic"> pageviews: 2 </p>

<article>
<div class="excerpt">
<ul>
<li>test_ape.R unit tests now passing. Should still revise the coercion methods in nexml_write.R to something more clever. We aren’t coercing a tree to a nexml, we are building a nexml from a tree. Likewise handling the otu assignments of species names should be done directly <a href="https://github.com/ropensci/RNeXML/commit/8a182d4cc6fb75099067142327788be8d51eef6e">11:20 am 2013/08/28</a></li>
<li>start cleaning up old read/write methods obsoleted by toNeXML/fromNeXML methods <a href="https://github.com/ropensci/RNeXML/commit/ffb7342f1d28b351984685a83e11d7df488696e8">10:59 am 2013/08/28</a></li>
<li>inherited classes now replace all the original classes <a href="https://github.com/ropensci/RNeXML/commit/004d31fc2c87a4cf2dba45ab08f1b8a663ab5258">10:54 am 2013/08/28</a></li>
<li>All inheritance-based classes appearing in inst/examples/trees.xml now implemented <a href="https://github.com/ropensci/RNeXML/commit/d9e649ac64b697d4ea982a1bb78b403400984392">10:53 am 2013/08/28</a></li>
<li><p>added classes for otu, otus, tree, trees, nexml, edge, rootEdge <a href="https://github.com/ropensci/RNeXML/commit/b9543f64352759f4dbb1f77a170cced78280c9a7">12:04 am 2013/08/28</a></p></li>
<li><p>Working on #1 : What works and what doesn’t: The example in inst/tests/inheritance.R should run using the class definitions in R/inherited_classes.R. Other class definitions are obsoleted by these updates and should eventually be removed. The ape methods should be unaffected, since the S4 classes have the same structure as before, only defined through inheritance instead.</p></li>
</ul>
<p>ToDo: - Remove the obsoleted methods in classes.R, nexml_read.R and nexml_write.R - Check the ape.R unit tests, update other unit tests - Write remaining essential classes out by hand based on inheritance and schema definitions: e.g. edge, otu, otus, tree, trees, nexml <a href="https://github.com/ropensci/RNeXML/commit/fd2e6b0f8f31823c4c65916a20fe339fb2038646">04:26 pm 2013/08/27</a> - Begun replacing class definitions with inheritance-based classes taking a brute-force approach in defining conversion methods. Conversion methods between XML and S4 explicitly assign attributes and values to S4 slots, rather than using the clever XML package functions to guess what goes where (since that approach cannot include a call to callNextMethod, which allows us to recursively go down the inheritance classes. The up side of this approach is that you don’t have to think much to write methods, it’s merely verbose.) <a href="https://github.com/ropensci/RNeXML/commit/8b45dee59848702c5aea6bf23ea06802b462d55d">04:11 pm 2013/08/27</a></p>
<ul>
<li>fixed ape tests, extended setAs(‘nexml’, ‘phylo’) to handle the case of multiple trees <a href="https://github.com/ropensci/RNeXML/commit/4fa252e20c7ee44a93573d3eb2c5dd3e781103c3">04:27 pm 2013/08/28</a></li>
<li>Now use otu labels as species names in both directions, closes #8 See updated tests in inst/test/test_ape.R. Note that coercing phylo-&gt;tree is still permitted, since resulting S4 tree is valid (though really intended as internal), but coercing tree-&gt;phylo is no longer available as a separate routine, since this is lossy as the tip.labels are not stored in tree but in nexml@otu <a href="https://github.com/ropensci/RNeXML/commit/abf298614f587d7eb99c67a0acd15d731690df54">03:54 pm 2013/08/28</a></li>
<li>updated nexml-&gt;ape conversion to get tip.labels from otus instead of just using the otu attribute on the node, which simply corresponds to the id of the otu element in otu. the taxon (species name) is in the label of the matching element instead. Still needs some testing and a unit test to confirm the label is correct. This addresses 1/2 of #8.</li>
</ul>
<p>Still need to handle the case of coercing ape-&gt;nexml, in which a method generates the otus node from the tip.label list and matches the id to node’s otu attribute. This means that ape-&gt;tree doesn’t look at tip.label, but just generates locally unique keys for the node otu attribute <a href="https://github.com/ropensci/RNeXML/commit/b97df1abac843b33214109d95c5baaa740dbe83e">01:40 pm 2013/08/28</a></p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/28/rnexml-coding.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/26/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 26 Aug 2013</p>

<p style="font-style:italic"> pageviews: 14 </p>

<article>
<div class="excerpt">
<h3 id="misc-tricks">misc tricks</h3>
<p>Always forget how to see the function definition of an S4 method:</p>
<pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">showMethods</span>(<span class="st">&quot;xmlToS4&quot;</span>, <span class="dt">includeDefs=</span><span class="ot">TRUE</span>)</code></pre>
<h3 id="ropensci">ropensci</h3>
<ul>
<li>Closed rfigshare issue <a href="https://github.com/ropensci/rfigshare/issues/50">#50</a></li>
<li>Still retooling reml to S4 and ReXML to inheritance-based S4 classes</li>
<li>building classes on the schema’s inheritance model <a href="https://github.com/ropensci/RNeXML/commit/58f45e087f8a973674e6170df7309f8c6a6dbc7f">10:49 am 2013/08/26</a></li>
</ul>
<hr />
<p>In behavioral ecology we usually equate optimality with expected behavior arising from natural selection – i.e. the extent to which the observed behavior deviates from the optimal behavior is assumed to be due to modeling inaccuracies (optimizing the wrong thing).</p>
<p>In the operations research / engineering context, it seems optimization is more of a direct decision making tool to decide a strategy / design a controller, rather than to understand behavior that arises from the interplay of other forces.</p>
<p>I’m not quite clear what the standard interpretation is in the environmental economics / natural resource management context – do we also ascribe these deviations to modeling inaccuracy or rather to naive decision-makers (or a combination?)</p>
<p>I’m guessing we are closer to the behavioral ecologist’s viewpoint – that, for instance, the deviations observed in Figure 1 (what is “optimal” vs what policymakers do) can be explained primarily in terms of optimizing the wrong thing (ignoring the policy costs). I’m wondering if it will appear crazy or standard to assert that the observed behaviour truly is optimal relative to the true costs and constraints, and therefore, just as the ecologist would do, we might use the data to guide us towards understanding the nature of the correct constraints and costs.</p>
<p>A engineering minded critic might prefer a softer assumption – that the observed pattern reflects some design constraints perhaps (policy costs) but is not already “optimal” in reference to them, but rather some crude rule of thumb for lack of knowledge of how to determine the precise optimal tradeoff between adjustment costs and benefits. The difficulty with this latter approach is that it invites a third model for the ‘rule-of-thumb’ in addition to the optimal models with and without policy costs included.</p>
<hr />
<p>ht @noamross for pointing out economists frequently assume deviations between predicted optimal and observed economic behavior (perhaps excluding policy..) are due to naiive models while we ecologists tend to assume they are the result of naiive policy.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/26/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/08/23/notes.html">Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 23 Aug 2013</p>

<p style="font-style:italic"> pageviews: 11 </p>

<article>
<div class="excerpt">
<h2 id="work-work">Work Work</h2>
<ul>
<li>Finished review for R Journal (Wednesday)</li>
<li>Finished review for TE.</li>
<li>Began review for NSF</li>
<li>updated service list (vita)</li>
<li>Handling reimbursement of expenses</li>
</ul>
<h2 id="misc">Misc</h2>
<ul>
<li>Whoa, we can/should now use http://doi.org instead of http://dx.doi.org</li>
</ul>
<h2 id="software-sustainability-materials">Software Sustainability materials</h2>
<p>Seems like a popular new item, listing a couple recent articles in one place here:</p>
<p>- - <a href="http://dx.doi.org/10.1126/science.1231535">Perspective in <em>Science Magazine</em></a>; also <a href="http://dx.doi.org/10.1126/science.341.6143.236-b">a critique</a> and <a href="http://dx.doi.org/10.1126/science.341.6143.237-a">author’s rebuttal</a> in a later edition. (Focuses on ecological code) - <a href="http://kaythaney.com/2013/08/08/experiment-exploring-code-review-for-science/">Mozilla Science Lab Code Review</a> - <a href="http://www.timeshighereducation.co.uk/news/save-your-work-give-software-engineers-a-career-track/2006431.article">Times Higher Ed piece</a> - Very good <a href="http://dirkgorissen.com/2012/09/13/the-research-software-engineer/">conference proceedings publication</a> from the UK’s Software Sustainability Institute and other folks. - <a href="http://isees.nceas.ucsb.edu/">ISEES workshops</a></p>
<h2 id="some-online-writing">Some Online writing</h2>
<p>Couldn’t resist weighing in on <a href="http://ideas4sustainability.wordpress.com/2013/07/13/a-critical-appraisal-of-a-new-paper-on-big-data-and-the-future-of-ecology">this “don’t require data sharing”</a> post.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/23/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/14/ISEES-day-2.html">Isees Day 2</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 14 Aug 2013</p>

<p style="font-style:italic"> pageviews: 6 </p>

<article>
<div class="excerpt">
<h2 id="other-services-tangible-infrastructure-support-services">Other services: Tangible infrastructure support services</h2>
<ul>
<li>Discovery registration system. Includes:</li>
<li>Software review and certification</li>
<li>Use and quality metrics</li>
<li>Ticket tracking</li>
<li>Seed funding to join the ecosystem</li>
</ul>
<h2 id="community-building">Community Building</h2>
<ul>
<li>user defined registry</li>
</ul>
<p>// Hackathon?</p>
<h2 id="consulting-services">Consulting services</h2>
<ul>
<li><p>Support software lifecycle</p></li>
<li>Function: Assist in full lifecycle software development: e.g. (and order notwithstanding) initiation, licensing, maintenance, hardening, extensible, release, decommissioning</li>
<li>Mechanisms:</li>
<li>Call for Proposals - to solicit consulting help (or to nominate a project to receive help) (e.g. Developer needing help supporting software. Community rescue of software. Developer wanting to archive obsolete software.)</li>
<li>Selection mechanisms for projects to assist - community-based steering group</li>
<li>Dissemination of best practices and training materials</li>
<li>ISEES: review successful proposals, identification of likely problems, selection of appropriate consulting expertise</li>
<li>ISEES: factiliation of execution consulting service - on-site, reverse-site, web-based, …</li>
<li><p>Consultant “hotline” - quick answers to funnel questions to appropriate commmunity-based solutions</p></li>
<li>Resources required:</li>
<li>Network/Registry of consultants / developers / analysts / etc. The experts may be full-time staff, part-time staff and ad hoc groups of experts to help with a well defined problem.</li>
<li><span class="math">\($\)</span> for consultants</li>
<li>Website for submissions</li>
<li>People to review submissions and identify resources</li>
<li>Facilities for teleconferencing, online working space (e.g. wikis, repositories for data, software, cook books)</li>
<li><p>Physical meeting/collaboration spaces</p></li>
<li><p>Reproduciblity Editor service, a la Journal of Biostats</p></li>
</ul>
<h2 id="advocacy">Advocacy</h2>
<ul>
<li>Take minimal advocacy approach. Roadmap best practices.</li>
<li>Advocate NSF adopt software management plan: not gonna fly. Outline best practices</li>
<li>Build brand, swag</li>
<li>managing sponsors for bootcamps?</li>
</ul>
<h2 id="crisis">Crisis</h2>
<p>The challenge with registries: Quickly become useless. Successful ones need automated updating. Does a few things things well.</p>
<ol type="1">
<li>How would ISEES uniquely provide value to enable science</li>
</ol>
<p>Intergrative science through software collaboration.</p>
<ol start="2" type="1">
<li>What 1 thing is most critical to pursue?</li>
</ol>
<p>via Hackathon / working group collaboration of researchers and expert developers. Building a sustainable software model emboding modularity and best practices.</p>
<p>Registry model as incentive problem.<br />Education and Code review Value is science. Pursue illustrating what can be achieved through open science.</p>
<p>Right Scale. Why 1 institute, not 6 regional institutes?</p>
<p>Efficient, Scalable, Reproducible, Impactful</p>
<ul>
<li>Ian Truslove</li>
<li>Christine White</li>
<li>Annie Bryant</li>
<li>Patricia Cruse</li>
<li>Chris Mattmann</li>
<li>Bryan Heidorn</li>
<li>Greg Ederer</li>
<li>Karthik Ram</li>
<li>Bill Michener</li>
<li>Bruce Caron</li>
<li>Matt Jones</li>
</ul>
<p>Science via Integrated Software.</p>
<p>Logic Model:</p>
<h3 id="functionserviceactivity">Function/Service/Activity</h3>
<p>Working group / hackathon on Software Lifecycle challenges that are impeding grand challenge science.</p>
<h3 id="resources">Resources</h3>
<p>2 - 4 meetings over 1-2 years. 8-10 people.</p>
<h3 id="outputs">Outputs</h3>
<p>Improvements to software lifecycle. (documentation, hardening, maintance, interface, usability, efficency, scalability)</p>
<h3 id="impact">Impact</h3>
<p>E,S,R,I (efficency, scalability, reproducibility, impact) science advances. New science that can’t be done now. Overall software quality advances.</p>
<h3 id="assessment">Assessment</h3>
<p>Number, quality, use of software products. Publications. Citations. Co-author networks (proxy of community)</p>
<h3 id="functionserviceactivity-1">Function/Service/Activity</h3>
<p>Workshop model. Address the other generic challenges (code review/registry, etc, see ‘other services’ at top)</p>
<hr />
<p>What is the metadata for scientific software? The DataCite Schema for software? Avoid duplicating listing efforts of other repositories (CPAN, CRAN, etc), while providing academically relevant elements like citations (and potentially other use metrics, a la Github), code review scoring, link to permanent repository and bug tracking.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/14/ISEES-day-2.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/13/ISEES-Day-1.html">ISEES software lifecycle and components workshop, Day 1</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 13 Aug 2013</p>

<p style="font-style:italic"> pageviews: 42 </p>

<article>
<div class="excerpt">
<p>Range of large project to the long tail.</p>
<ul>
<li>statistical analyses</li>
<li>One-off models</li>
<li>community model</li>
<li>Services (Blast)</li>
</ul>
<h3 id="science-challenges">Science Challenges</h3>
<p>Outcomes of previous community workshops</p>
<h4 id="grand-challenge-areas.">18 Grand Challenge areas.</h4>
<p>Highlights: Supermodels. human/biophysical systems shape and be shaped by water availability. (A vision similar to Microsoft’s “Model all Earth” <a href="http://dx.doi.org/10.1038/493295a">Purves et al</a>)</p>
<h4 id="functions-and-services-areas">Functions and services areas:</h4>
<p>(priority ordering)</p>
<ul>
<li>Computational training (early career - all career)</li>
<li>Assimilation and QA/QC tools</li>
<li>Collaborative environment</li>
<li>decisions and workflows</li>
<li>software discovery</li>
<li>consultants / collaborator</li>
<li>community hub to converge on standards</li>
<li>merging disparate tools</li>
<li>user-friendly interfaces</li>
<li>multiscale coupled modeling framework</li>
<li>Software vetting (how about paper vetting?)</li>
</ul>
<h3 id="software-sysems-and-lifecycle">Software Sysems and Lifecycle</h3>
<p>Innovation -&gt; incubation/integration -&gt; prduct development/hardening -&gt; maintenance and support</p>
<hr />
<ul>
<li>What is and is not out of scope?<br /></li>
<li>Models of success (ESIP grassroots conception of problems). NCEAS working groups. NeSCENT Hackathons.<br /></li>
<li>Models of failure</li>
</ul>
<blockquote>
<p>NSF has funded 16 Visions. Many are computer science efforts (compilers, language integration etc). 3 proposals are domain focused with close overlap: ISEES, Long Tail, and Water Science. These need to be aligned for us to succeed.</p>
</blockquote>
<blockquote>
<p>How about existing stuff:</p>
</blockquote>
<ul>
<li>ESIP? Lots of overlap.</li>
<li>EarthCube Large top-down effort. Geo directorate, not Cyberinfrastructure (OCI) anyway, so overlap matters less. However, better funding chances when draws across directorates. (tap BIO).</li>
</ul>
<p>Those are integration in general, not software sustainability issues in science.</p>
<hr />
<p>Software huge range! Goal is different for different products! Incremental improvement from paper-only model.</p>
<p>Sustainability and Adoption</p>
<p>Town Hall ESIP (Environmental Science Information Planning)</p>
<p>NCEAS for software. Instead of showing up with spreadsheets and trying to combine them, we show up with software. We add nubs.</p>
<ul>
<li>Science as Lego. Papers: blocks. without the nubs. Easy to build tall towers. Fragile.<br /></li>
<li>Data as Lego. Metadata as nubs.<br /></li>
<li>Software as Lego. What are the nubs?</li>
</ul>
<p>small. unit tests. versioned. dependencies. accessible language and style. Licenses.</p>
<h2 id="software-lifecyle-session">Software Lifecyle Session</h2>
<p>Comparison: Commercial, Open Source, Academic Lab.</p>
<p>Note this isn’t software <em>development</em> lifecycle, which would occur within any of these blocks.</p>
<p>Goals:</p>
<ul>
<li>Issues that occur (strengths, weaknesses) in the lifecycle</li>
</ul>
<hr />
<h3 id="services-activities-and-functions">Services, Activities and Functions</h3>
<ul>
<li>Software quality assessment</li>
<li>Training</li>
<li>Hackathons</li>
<li>Archiving</li>
</ul>
<p>Add nubs. Good development makes for faster development.</p>
<p>User/developer access</p>
<p>Documentation. Versioning. Dependencies</p>
<hr />
<h2 id="afternoon-breakout">Afternoon breakout</h2>
<p>Brainstorm Services, Activities, Functions.</p>
<h4 id="assist-groups-in-disseminating-and-hardening-software.">Assist groups in disseminating and hardening software.</h4>
<p>Successsful academic software is frequently a victim of it’s own success. They cannot keep up with support, bug requests.</p>
<ul>
<li><strong>Strategies for dissemination and maintenance.</strong></li>
<li>Commercialization
<ul>
<li>Connections</li>
<li>test data and use cases. (Because open source and commercial often don’t meet academic goals because they don’t have access to the data or questions!)<br /></li>
</ul></li>
<li>Open Source Release
<ul>
<li>Strategy template: Host repository, contribution guidelines</li>
<li>Mentorship for managing development<br /></li>
<li>Language and style accessible to community</li>
<li>Hackathon for engaging community for future development</li>
</ul></li>
<li><strong>Hardening of software</strong></li>
<li>Written product: Objectives checklist: unit tests, RESTful APIs</li>
<li>Meeting space: hackathon for hardening</li>
<li>Funding: e.g. for a student Summer of Code style for hardening</li>
</ul>
<h4 id="software-review-and-certification">Software review and certification</h4>
<ul>
<li>tied to software lifecycle</li>
<li>Tiered system. e.g. level 3 is external hands-on testing by experts.</li>
<li>Couple to publication of software papers or separate model?</li>
<li>Scoring system analgous to repository scoring.</li>
</ul>
<p>Computational career path. // (me): Doesn’t have to be - phylogenetics research driven by computational side.</p>
<p>Leveraging social media, hashtags, stackoverflow for community</p>
<h3 id="synthesizing-functions">Synthesizing Functions</h3>
<ol type="1">
<li>Assist Groups in Maintenance (CS)</li>
<li>Assist Groups in Hardening (CS)</li>
<li>Software Review and Certification Program (OS) (ASSESSMENT)</li>
<li>Mechanisms for sharing data / use cases (OS)</li>
<li>Operation community based support services (CB) // no, these discussions happen in native environments.<br /></li>
<li>Provide sofware use and quality metrics (OS) (ASSESSMENT)</li>
<li>Provide best practices boot camp (T) (BEST PRACTICES)</li>
<li>On site and remote course training (T) (BEST PRACTICES)</li>
<li>The Dating Game - match researchers with engineers (CB)</li>
<li>Support collaborative groups focused on software projects (CB)</li>
<li>Advocate for open source approach (A)<br /></li>
<li>Develop policies to incentivise sharing and licenses (A)</li>
<li>Promote software management plans to NSF etc (A)<br /></li>
<li>Consulation and mentoring services (CS)</li>
<li>Directory for discovery (citation, evaluation) (OS)</li>
<li>Kickstarter funding for software (OS)</li>
<li>Connect industry to academia (CB)</li>
<li>Certify papers as reproducible (RunMyCode service) (OS)</li>
<li>Assist groups in making extensible (CS)</li>
<li>ISEES swag (CB)</li>
<li>Actively survey the community about practices. (OS)</li>
<li>Assist groups in archiving and decomissioning. (CS)</li>
</ol>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/13/ISEES-Day-1.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

