<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
  <head prefix="dc: http://purl.org/dc/terms/ og: http://ogp.me/ns#"> <!-- namespaces used in metadata.html -->
  <meta http-equiv='Content-Type' content='text/html; charset=utf-8'/>
  <title>Lab Notebook</title>
  <meta name="author" content="Carl Boettiger" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  

<!-- Get date last modified from git log. (Uses current time if file entry not found, e.g. projects/)  -->



<!-- For posts, page.date is the date they are published under, which we use as their 'canonical' dc:date -->
 <!-- If we don't have a page.date, then use modified time (pages) -->
   


<!-- Posts declare modified timestamps in the sidebar, so would be redundant to put here. But then 
     pages don't have a dc:modified... unless we give them their own (modified) sidebar?  
-->
<!-- Ideally we would want date originally created from the _oldest_ git commit too...-->

<!-- HTML5 metadata -->
<meta name="keywords" content="" />
<meta name="description" content="" />
<meta name="resource_type" content="website"/> 
<!-- RDFa Metadata (in DublinCore) -->
<meta property="dc:title" content="Lab Notebook" />
<meta property="dc:creator" content="Carl Boettiger" />
<meta property="dc:date" content="2013-07-31T22:00:31-07:00" />
<meta property="dc:format" content="text/html" />
<meta property="dc:language" content="en" />
<meta property="dc:identifier" content="/lab-notebook.html" />
<meta property="dc:rights" content="CC0" />
<meta property="dc:source" content="Lab Notebook" />
<meta property="dc:subject" content="Ecology" /> 
<meta property="dc:type" content="website" /> 
<!-- RDFa Metadata (in OpenGraph) -->
<meta property="og:title" content="Lab Notebook" />
<meta property="og:author" content="http://carlboettiger.info/index.html#me" />  <!-- Should be Liquid? URI? -->
<meta property="http://ogp.me/ns/profile#first_name" content="Carl"/>
<meta property="http://ogp.me/ns/profile#last_name" content="Boettiger"/>
<meta property="http://ogp.me/ns/article#published_time" content="2013-07-31T22:00:31-07:00" />
<meta property="og:site_name" content="Lab Notebook" /> <!-- Same as dc:source? -->
<meta property="og:url" content="http://carlboettiger.info/lab-notebook.html" />
<meta property="og:type" content="website" /> 
<!-- Google Scholar Metadata -->
<meta name="citation_author" content="Carl Boettiger"/>
<meta name="citation_date" content="2013-07-31T22:00:31-07:00"/>
<meta name="citation_title" content="Lab Notebook"/>
<meta name="citation_journal_title" content="Lab Notebook"/>

<!--NOTE: see also the COinS Metadata in span element in footer -->




  <!-- CSS Stylesheets (alternate sheets toggled with javascript or browsers supporting rel='alternative stylesheet') -->
  <link href="/assets/css/bootstrap.css" rel="stylesheet" 
        type="text/css" title="white" />
  <link href="/assets/css/light.css" rel="alternate stylesheet"
        type="text/css" id="stl" title="light" />
  <link href="/assets/css/dark.css" rel="alternate stylesheet" 
        type="text/css" title="dark" />
  <link href="/assets/css/bootstrap-responsive.css" rel="stylesheet" 
  type="text/css"/>
  <!-- Help the browser identify the RSS feed automatically -->
  <link rel="alternate" type="application/rss+xml" title="Carl Boettiger's Lab Notebook" href="/blog.xml" />
  <!-- Javascript needed for theme toggle, load immediately -->
  <script type="text/javascript" src="/assets/js/switch-css.js"></script>
  <script type="text/javascript">set_style_from_cookie(); </script>
</head>


  <body prefix="dc: http://purl.org/dc/terms/ foaf: http://xmlns.com/foaf/0.1/"> 
    <!-- Navbar  ================================================== -->
<div class="navbar navbar-fixed-top">
  <div class="navbar-inner">
    <div class="container">
      <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </a>
      <a class="brand" href="/README.html"><i style="float:right;" class="icon-info-sign" alt="info"></i></span></a>
      <div class="nav-collapse">
        <ul class="nav">
          <li>
          <a href="/index.html">Home</a></li>

          <li>
          <a href="/vita.html">Vita</a></li>

          <li>
          <a href="/research.html">Research</a></li>

          <li>
          <a href="/teaching.html">Teaching</a></li>

          <li>
          <a href="/community.html">Community</a></li>

          <li class="active">
          <a href="/lab-notebook.html">Lab Notebook</a></li>

        </ul>
      </div><!--/.nav-collapse -->

      <!-- Search site using Google's index -->
      <form class="navbar-search pull-right" method="get" action="http://google.com/search">
        <p>
          <input type="hidden" name="q" value="site:carlboettiger.info" />
          <input type="text" class="search-query" name="q" />
          <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      <!--
      <div id="search">
      <form method="get" class="navbar-search pull-right form-search">
        <p>
        <input type="text" name="search-text" id="search-text">
         <button class="btn btn-mini" type="submit"><i class="icon-search"></i></button> 
        </p>
      </form>
      </div>
      -->
     </div> <!-- /container -->
   </div> <!-- /navbar-inner -->
 </div> <!-- /navbar -->



    <div class="container"> <!-- Responsive grid layout --> 

      <header class="jumbotron">
  <h1 class="entry-title">Lab Notebook</h1>
  <h3>(<a href="http://www.carlboettiger.info/2012/09/28/Welcome-to-my-lab-notebook.html">Introduction</a>)</h3>
</header>



<div class="row feed">
  <div class="span3 offset1">
    <h4>  <a property="account" href="https://github.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); return false;"><i class="icon-github" alt="github"></i> Coding </a></h4> 
    <div class="excerpt">
      <div class="scroll">
        <ul><li>cboettig pushed to master at ropensci/reml: <em>passing unit tests in S4.R Working on S4 methods</em> <a href="https://github.com/ropensci/reml/compare/66e284ffe0...9e52295536">06:21 2013/08/14</a></li><li>cboettig pushed to master at ropensci/RNeXML: <em>minor changes:</em> <a href="https://github.com/ropensci/RNeXML/compare/9592588289...6f96e1f4b9">07:27 2013/08/12</a></li><li>cboettig opened issue ropensci/RNeXML#14: <em>All elements need id(?), strategy for generating ids</em> <a href="https://github.com/ropensci/RNeXML/issues/14">07:25 2013/08/12</a></li><li>cboettig pushed to master at ropensci/RNeXML: <em>generate otus node #8</em> <a href="https://github.com/ropensci/RNeXML/compare/9b093e531b...9592588289">07:02 2013/08/12</a></li><li>cboettig opened issue ropensci/RNeXML#13: <em>Tackling "special cases"</em> <a href="https://github.com/ropensci/RNeXML/issues/13">06:18 2013/08/12</a></li></ul>
      </div>
    </div>
  </div>
  <div class="span3">
    <h4> <a property="account" href="https://twitter.com/cboettig" onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); return false;"><i class="icon-twitter"></i> Discussing </a></h4> 
     <div class="excerpt">
      <div class="scroll">
       <ul><li><p>Recent Perspective published in Science discussing some of the problems with academic software that keep coming up… <a href="http://t.co/SmbN1VAIal">http://t.co/SmbN1VAIal</a></p>
 <a href="http://twitter.com/cboettig/statuses/367413731224846336">03:33 2013/08/13</a> </li><li><p>I think much of the work of the Sloan funded Mozilla Science Lab / Software Carpentry project is quite relevant to… <a href="http://t.co/MiATLIOjIl">http://t.co/MiATLIOjIl</a></p>
 <a href="http://twitter.com/cboettig/statuses/367386404264738816">01:45 2013/08/13</a> </li><li><p>back from #ESA2013 and off to @NCEAS to discuss #SoftwareSustainability. Thanks  to @ethanwhite, ben bolker, @metamattj et al for ideas</p>
 <a href="http://twitter.com/cboettig/statuses/367021537620017152">01:35 2013/08/12</a> </li><li><p>@jagephart nice meeting you at the @sesync mixer, hope rest of your #ESA2013 went well!</p>
 <a href="http://twitter.com/cboettig/statuses/366761776714498048">08:23 2013/08/11</a> </li><li><p>RT @polesasunder: Consider signing this petition to form an ESA Open Science Section! <a href="https://t.co/xsC1brMLpS">https://t.co/xsC1brMLpS</a></p>
 <a href="http://twitter.com/cboettig/statuses/366619124631220225">10:56 2013/08/11</a> </li></ul>
      </div>
    </div> 
  </div> 
  <div class="span3">
    <h4> <a href="http://www.mendeley.com/groups/634301/theoretical-ecology/papers/" onClick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); return false;"><i class="icon-book"></i> Reading </a></h4> 
    <div class="excerpt">
      <div class="scroll">
      <ul>
<li>Stephan B Munch, A Kottas. 2009. A Bayesian modeling approach for determining productivity regimes and their characteristics <em>Ecological Applications</em> <strong>19</strong> 2 <a href="http://dx.doi.org/10.1890/07-2116.1">10.1890/07-2116.1</a> </li>
<li>Mark John Costello, Marta Coll, Roberto Danovaro et al. 2010. A Census of Marine Biodiversity Knowledge, Resources, and Future Challenges <em>PLoS ONE</em> <strong>5</strong> 8 <a href="http://dx.doi.org/10.1371/journal.pone.0012110">10.1371/journal.pone.0012110</a> </li>
<li>R. F. Hechinger, K. D. Lafferty, a. P. Dobson et al. 2011. A Common Scaling Rule for Abundance, Energetics, and Production of Parasitic and Free-Living Species <em>Science</em> <strong>333</strong> 6041 <a href="http://dx.doi.org/10.1126/science.1204337">10.1126/science.1204337</a> </li>
<li>Jonathan B. Shurin, Elizabeth T. Borer, Eric W. Seabloom et al. 2002. A cross-ecosystem comparison of the strength of trophic cascades <em>Ecology Letters</em> <strong>5</strong> 6 <a href="http://dx.doi.org/10.1046/j.1461-0248.2002.00381.x">10.1046/j.1461-0248.2002.00381.x</a> </li>
<li>Wayne M Getz, David Saltz. 2008. A framework for generating and analyzing movement paths on ecological landscapes. <em>Proceedings of the National Academy of Sciences of the United States of America</em> <strong>105</strong> 49 <a href="http://dx.doi.org/10.1073/pnas.0801732105">10.1073/pnas.0801732105</a> </li>

      </div>
    </div>
  </div> 
</div>

<hr>
<div class="row postpreview">
  <div class="span11 offset1">
    <div class="row">
      <h4> <a href="http://carlboettiger.info/atom.xml"
              onClick="recordOutboundLink(this,
              'Outbound Links', 'RSS'); return false;"
              style="color: inherit;"
              ><i class="icon-rss" ></i> Entries</a></h4>
      
        <div class="span3">
          <header><h4><a href="/2013/08/13/ISEES-Day-1.html">ISEES software lifecycle and components workshop, Day 1</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 13 Aug 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>Range of large project to the long tail.</p>
<ul>
<li>statistical analyses</li>
<li>One-off models</li>
<li>community model</li>
<li>Services (Blast)</li>
</ul>
<h3 id="science-challenges">Science Challenges</h3>
<p>Outcomes of previous community workshops</p>
<h4 id="grand-challenge-areas.">18 Grand Challenge areas.</h4>
<p>Highlights: Supermodels. human/biophysical systems shape and be shaped by water availability. (A vision similar to Microsoft’s “Model all Earth” <a href="http://dx.doi.org/10.1038/493295a">Purves et al</a>)</p>
<h4 id="functions-and-services-areas">Functions and services areas:</h4>
<p>(priority ordering)</p>
<ul>
<li>Computational training (early career - all career)</li>
<li>Assimilation and QA/QC tools</li>
<li>Collaborative environment</li>
<li>decisions and workflows</li>
<li>software discovery</li>
<li>consultants / collaborator</li>
<li>community hub to converge on standards</li>
<li>merging disparate tools</li>
<li>user-friendly interfaces</li>
<li>multiscale coupled modeling framework</li>
<li>Software vetting (how about paper vetting?)</li>
</ul>
<h3 id="software-sysems-and-lifecycle">Software Sysems and Lifecycle</h3>
<p>Innovation -&gt; incubation/integration -&gt; prduct development/hardening -&gt; maintenance and support</p>
<hr />
<ul>
<li>What is and is not out of scope?<br /></li>
<li>Models of success (ESIP grassroots conception of problems). NCEAS working groups. NeSCENT Hackathons.<br /></li>
<li>Models of failure</li>
</ul>
<blockquote>
<p>NSF has funded 16 Visions. Many are computer science efforts (compilers, language integration etc). 3 proposals are domain focused with close overlap: ISEES, Long Tail, and Water Science. These need to be aligned for us to succeed.</p>
</blockquote>
<blockquote>
<p>How about existing stuff:</p>
</blockquote>
<ul>
<li>ESIP? Lots of overlap.</li>
<li>EarthCube Large top-down effort. Geo directorate, not Cyberinfrastructure (OCI) anyway, so overlap matters less. However, better funding chances when draws across directorates. (tap BIO).</li>
</ul>
<p>Those are integration in general, not software sustainability issues in science.</p>
<hr />
<p>Software huge range! Goal is different for different products! Incremental improvement from paper-only model.</p>
<p>Sustainability and Adoption</p>
<p>Town Hall ESIP (Environmental Science Information Planning)</p>
<p>NCEAS for software. Instead of showing up with spreadsheets and trying to combine them, we show up with software. We add nubs.</p>
<ul>
<li>Science as Lego. Papers: blocks. without the nubs. Easy to build tall towers. Fragile.<br /></li>
<li>Data as Lego. Metadata as nubs.<br /></li>
<li>Software as Lego. What are the nubs?</li>
</ul>
<p>small. unit tests. versioned. dependencies. accessible language and style. Licenses.</p>
<h2 id="software-lifecyle-session">Software Lifecyle Session</h2>
<p>Comparison: Commercial, Open Source, Academic Lab.</p>
<p>Note this isn’t software <em>development</em> lifecycle, which would occur within any of these blocks.</p>
<p>Goals:</p>
<ul>
<li>Issues that occur (strengths, weaknesses) in the lifecycle</li>
</ul>
<hr />
<h3 id="services-activities-and-functions">Services, Activities and Functions</h3>
<ul>
<li>Software quality assessment</li>
<li>Training</li>
<li>Hackathons</li>
<li>Archiving</li>
</ul>
<p>Add nubs. Good development makes for faster development.</p>
<p>User/developer access</p>
<p>Documentation. Versioning. Dependencies</p>
<hr />
<h2 id="afternoon-breakout">Afternoon breakout</h2>
<p>Brainstorm Services, Activities, Functions.</p>
<h4 id="assist-groups-in-disseminating-and-hardening-software.">Assist groups in disseminating and hardening software.</h4>
<p>Successsful academic software is frequently a victim of it’s own success. They cannot keep up with support, bug requests.</p>
<ul>
<li><strong>Strategies for dissemination and maintenance.</strong></li>
<li>Commercialization
<ul>
<li>Connections</li>
<li>test data and use cases. (Because open source and commercial often don’t meet academic goals because they don’t have access to the data or questions!)<br /></li>
</ul></li>
<li>Open Source Release
<ul>
<li>Strategy template: Host repository, contribution guidelines</li>
<li>Mentorship for managing development<br /></li>
<li>Language and style accessible to community</li>
<li>Hackathon for engaging community for future development</li>
</ul></li>
<li><strong>Hardening of software</strong></li>
<li>Written product: Objectives checklist: unit tests, RESTful APIs</li>
<li>Meeting space: hackathon for hardening</li>
<li>Funding: e.g. for a student Summer of Code style for hardening</li>
</ul>
<h4 id="software-review-and-certification">Software review and certification</h4>
<ul>
<li>tied to software lifecycle</li>
<li>Tiered system. e.g. level 3 is external hands-on testing by experts.</li>
<li>Couple to publication of software papers or separate model?</li>
<li>Scoring system analgous to repository scoring.</li>
</ul>
<p>Computational career path. // (me): Doesn’t have to be - phylogenetics research driven by computational side.</p>
<p>Leveraging social media, hashtags, stackoverflow for community</p>
<h3 id="synthesizing-functions">Synthesizing Functions</h3>
<ol type="1">
<li>Assist Groups in Maintenance (CS)</li>
<li>Assist Groups in Hardening (CS)</li>
<li>Software Review and Certification Program (OS) (ASSESSMENT)</li>
<li>Mechanisms for sharing data / use cases (OS)</li>
<li>Operation community based support services (CB) // no, these discussions happen in native environments.<br /></li>
<li>Provide sofware use and quality metrics (OS) (ASSESSMENT)</li>
<li>Provide best practices boot camp (T) (BEST PRACTICES)</li>
<li>On site and remote course training (T) (BEST PRACTICES)</li>
<li>The Dating Game - match researchers with engineers (CB)</li>
<li>Support collaborative groups focused on software projects (CB)</li>
<li>Advocate for open source approach (A)<br /></li>
<li>Develop policies to incentivise sharing and licenses (A)</li>
<li>Promote software management plans to NSF etc (A)<br /></li>
<li>Consulation and mentoring services (CS)</li>
<li>Directory for discovery (citation, evaluation) (OS)</li>
<li>Kickstarter funding for software (OS)</li>
<li>Connect industry to academia (CB)</li>
<li>Certify papers as reproducible (RunMyCode service) (OS)</li>
<li>Assist groups in making extensible (CS)</li>
<li>ISEES swag (CB)</li>
<li>Actively survey the community about practices. (OS)</li>
<li>Assist groups in archiving and decomissioning. (CS)</li>
</ol>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/13/ISEES-Day-1.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/08/ESA-2013-notes-for-week.html">Esa 2013 Notes For Week</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 08 Aug 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<table>
<col style="width: 2%" /><tbody>
<tr class="odd">
<td style="text-align: left;">layout: post category: ecology tags: - conference - open-science</td>
</tr>
</tbody>
</table>
<h2 id="sun">Sun</h2>
<p>Co-Leading the Workshop on R, DataONE and open science.</p>
<p>All material on <a href="github.com/mbjones/opensci_r_esa_2013">github</a></p>
<h2 id="mon">Mon</h2>
<ul>
<li>Ives Plenary</li>
</ul>
<p>Talking to Dave, Vasilis, Tye,</p>
<ul>
<li><p>Lunch Vasilis, Ryan Batt</p></li>
<li>2:30 Chesson L100H / Ryan Batt L100J</li>
<li>3:20 O’Regan L100H</li>
<li>3:40 Adler L100H</li>
<li><p>4 Yaekel M100HC</p></li>
<li>ropensci meetup, funded by Github. Great turnout, great success. (And rare <a href="http://t.co/AtmzFbJCCg">shot of dev team</a> together in person).<br /></li>
<li>Theoretical ecology mixer</li>
<li><p>Hastings Lab dinner</p></li>
</ul>
<p>Great discussions with Trevor Hefley on some very nice work dealing with biases in messy data. See <a href="http://dx.doi.org/10.1016/j.ecolmodel.2013.05.003">Hefley 2013</a> and <a href="http://dx.doi.org/10.1007/s12080-013-0195-3">2013b</a>, and more in the works.</p>
<h2 id="tues">Tues</h2>
<ul>
<li>8am Carpenter L100B</li>
<li>M100IB 8am - 10am</li>
<li>10am meeting Carpenter, see seperate nodes</li>
<li>12am Lunch with Lei Dai</li>
<li>2:50pm Drake L100G</li>
</ul>
<p>Afternoon inteview with Virginia Gewin on Yeakel’s talk. Remarkable challenge to explain generalized modeling for ecological networks in plain english. Virginia gets all the credit for condensing my lengthy verbage into a tweet size summary. Now out in <a href="http://www.nature.com/news/ancient-art-fills-in-egypt-s-ecological-history-1.13528">Nature News</a></p>
<ul>
<li>Open Science session in evening</li>
</ul>
<p>Amazing talk and later discussion with Russell Niches.</p>
<ul>
<li>open hardware and 3D printers for scientific research. 3 examples:</li>
<li>cheaper than alternatives (water quality monitor on rduino chip, tweets data)</li>
<li>More flexible than alternatives: variable depth 96 well plates to automate concentration of sample. (also cheap gel combs)</li>
<li>When no alternative exists: measuring gape without harming fish.<br /></li>
<li>Avoid Nature’s $9000 APC open access option by including a $1000 new genome. (Genome papers are OA by agreement…)</li>
<li>Choose PhD advisor by Bayesian Classifier…</li>
</ul>
<h2 id="weds">Weds</h2>
<ul>
<li>3:20 David Hoover (missed)</li>
<li>3:40 Ingrid</li>
</ul>
<p>spatial waves maxwell</p>
<ul>
<li><p>4pm L Dai 101D</p></li>
<li><p>4:20 Anna Gardmark http://eco.confex.com/eco/2013/webprogram/Paper40488.html</p></li>
</ul>
<p>Explaining lack of recovery / hysteresis</p>
<ul>
<li>Prey competing with (younger life stages of) predators (Bundy Finning 2005 CJFAS, Walters .. , …)</li>
<li>Prey predating upon younger life stages of codes (Swan and Sinclair 2000 CJFAS)</li>
<li><p>Prey growth changes due to predation (poor condition, or prey size structure) (vulnerability - deRoos Peerson 2002 PNAS)</p></li>
<li><p>4:40 Mark Novak http://eco.confex.com/eco/2013/webprogram/Paper42117.html</p></li>
</ul>
<p>Kelp-urchin transition. Otters arrival leads to transitions in Vancouver Island set, resulting in shifts (used as training). San Nicolas Island sets transitions on its own, despite otter introduction in process. Spatial transects and 31 years of data, 5 replicate transects for all sites. Bayesian change-point analysis to identify shifts (e.g. see Erdman &amp; Emerson 2012).</p>
<!-- Prior probability that change-point occurs, and signal to noise ratio priors.  Get posterior prob distribution of tipping pt occurs - continuous.  -->

<p>Chatting with Mark after talk.</p>
<ul>
<li>SySnc / NCEAS Mixer</li>
<li>TweetUp</li>
</ul>
<p>Discussion of semantics and NeXML with Ben Morris (wrote python parser). Sounds like I need to read <em>semantic web for working ontologist</em></p>
<p>Excellent discussion with Ethan White on what a Institute for Software Sustainability would do.</p>
<ul>
<li>Training training training. Materials, workshops.</li>
<li>Much would be improved with a few best practices: unit tests, version control</li>
<li>Google Summer of Code style funding</li>
</ul>
<h2 id="thurs">Thurs</h2>
<ul>
<li>8am Thurs Dakos</li>
<li>Discussion with Ryan Batt</li>
<li>Discussions with Lei Dai and Vasilis. Lei’s Science paper data on Dryad (temporal indicators). Get back to Lei on format for this data.<br /></li>
<li>Exchange with Bestlemyer - lack of time resolution for warining indicators of spatial study. Data from Ecosphere paper on Harvard’s LTER site.</li>
</ul>
<p>Excellent discussion with Ben Bolker and Dave Harris on what Software Sustainability Institute would accomplish:</p>
<ul>
<li>Tackling archiving / backports of code. CRAN offers only “current” and “devel”, you’re on you’re own getting older stuff running.<br /></li>
<li>Training! Not just for grad students - Ben would attend. More experienced, but self-taught, developers like Ben would come with their own questions and could guide what training they needed.<br /></li>
<li>Bringing academics together with experts in hackathon style of improving existing codes would be great.<br /></li>
<li>One challenge is maintaining accessibility of code styles so that academics can continue to contribute. e.g. most academic R programmers don’t understand S4 classes, would be a potential barrier.</li>
</ul>
<p>Van Noorden’s <a href="http://t.co/KqFfGnpOAa">data sharing article</a> out, with quotes by Elizabeth, Wolkovich, Heather, Bill Mitchner, yours truly.</p>
<ul>
<li>Flight out</li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/08/ESA-2013-notes-for-week.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/08/07/F1000-Research.html">F1000-Research publishing for ecologists</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 07 Aug 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<p>F1000-Research is waving article processing charges (APCs) for ecologists looking for an open access, open peer review platform to publish original research, null results, software papers, data papers, and an intriguing new category called “Observations”,</p>
<blockquote>
<p>to report serendipitous observations that they have not been able to study systematically, but that offer a starting point for further exploration. This is especially relevant for ecologists, as it provides a platform to report unexpected observations from the field, unrelated to the main goal of the field work.</p>
</blockquote>
<p>This sounds similar to the Herp Notes of the <a href="http://www.ssarherps.org/pages/HRinfo.php">Herpetological Review</a>, though obviously without a species bias. Hopefully F1000’s version will support intelligent metadata on place, time, and taxonomy involved.</p>
<p>While APC waivers are a reasonable way for the journal to say thanks while establishing themselves, I would prefer to see it built upon proper market competition that involves a real market signal, rather than on fee waivers. Authors should choose if the added value of the journal justifies the price, and publishers should experiment with ways to drive down prices or improve added value through further publishing innovation. They don’t post prices as promently as PeerJ, but it does look like their <a href="http://f1000research.com/article-processing-charges">prices</a> are a step in this direction. At $500 for short-format papers and $1000 for full papers, they have at least broken the pattern of copying PLoS ONE pricing (Nature Scientific Reports, ESA Ecosphere, looking at you) in the process. At this rate, costs per paper may be less than PeerJ’s for papers in the 5-10 author range…</p>
<p>F1000Research also has an interesting review model in place that tries to be both pre-print and publication platform. Reviews are solicited while a preprint is posted, and the article becomes approved when three positive reviews are recieved. As such, it has always struck me as being very close to that of <a href="http://www.biologydirect.com/about">Biology Direct</a>, with some key differences. Biology Direct, now six or seven years old, has become a place for high impact potentially controversial work, as well as more run-of-the-mill results. Reviews tend to be longer and more detailed than I see on F100 Research, though I hardly have statistically relevant sample. I suspect that the key differences in policy might help this: the signed reviews are published with the paper, regardless of whether the reviewers agree with the results (editors can reject inappropriate papers just by declining to have them reviewed).</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/07/F1000-Research.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/08/06/summarizing-earlier-warning-signal-under-control-notes.html">Summarizing Earlier Warning Signal Under Control Notes</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 06 Aug 2013</p>

<p style="font-style:italic"> pageviews: (not calculated) </p>

<article>
<div class="excerpt">
<ul>
<li>NSF goals:</li>
<li>combine control methods and warning signals</li>
<li>do non-parametric or machine learning increase or decrease chance of transition</li>
<li>learning</li>
<li>Further areas (from comment piece)</li>
<li>baselines (statistical; emperical)</li>
<li>context-specific (additional data sources, reflecting context in model)</li>
</ul>
<h2 id="warning-signals-in-managed-systems">Warning signals in managed systems</h2>
<p>Management decreases the signal: compare likelihood statistic in <a href="https://github.com/cboettig/pdg_control/blob/master/inst/examples/unmanaged_warning.md">unmanaged_warning</a> <a href="https://github.com/cboettig/pdg_control/blob/master/inst/examples/managed_warning.md">managed_warning</a>:</p>
<figure>
<img src="http://farm8.staticflickr.com/7108/7853909166_45ecb1af08_o.png" alt="unmanaged" /><figcaption>unmanaged</figcaption>
</figure>
<figure>
<img src="http://farm9.staticflickr.com/8445/7889665150_d2f03f6f86_o.png" alt="managed" /><figcaption>managed</figcaption>
</figure>
<h2 id="estimating-stability-in-managed-systems">Estimating Stability in managed systems</h2>
<p>Recovery rate in optimally fished vs unfished system (based on the OU model): <a href="https://github.com/cboettig/pdg_control/blob/e6aaf0ba27e4280cbbb04d5ecfeb9b18f158ca77/inst/examples/stability.md">stability.md</a></p>
<figure>
<img src="http://farm8.staticflickr.com/7041/6960365083_4b41dd1f0a_z.jpg" alt="OU parameters over replicates" /><figcaption>OU parameters over replicates</figcaption>
</figure>
<ul>
<li>How would a use of a signal feed back into a control algorithm? Need probabilities of possible models for driving a belief SDP?</li>
</ul>
<h2 id="learning">Learning</h2>
<p>Learning about a tipping point: models are the correct allee-threshold (Myers) model, and the best-fit Beverton Holt model. Inability to learn fast enough leads to crashes <a href="https://github.com/cboettig/pdg_control/blob/ff049e1f623413e1438555aec63ddbe65dec6d8f/inst/examples/model_uncertainty.md">model_uncertainty.md</a></p>
<figure>
<img src="http://farm8.staticflickr.com/7231/7019134197_cb196dbdb1_o.png" />
</figure>
<ul>
<li>Active learning about parameters can be worse than not learning?</li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/08/06/summarizing-earlier-warning-signal-under-control-notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/28/nonparametric-bayes-wrapping-up-sensitivity.html">Nonparametric Bayes Wrapping Up Sensitivity</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 28 Jul 2013</p>

<p style="font-style:italic"> pageviews: 20 </p>

<article>
<div class="excerpt">
<p>Remote RStudio configuration (keep forgetting options and ports).</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">ssh</span> -o <span class="st">&quot;ExitOnForwardFailure yes&quot;</span> -f -N -L 1234:localhost:8787 one</code></pre>
<p>Initialize runs for sensitivity, looping over parameters and data. See:</p>
<ul>
<li><a href="https://github.com/cboettig/nonparametric-bayes/blob/fd3585230a01afa525f456d13f5f958df4606006/inst/examples/sensitivity.md">sensitivity.md</a> (Myers)</li>
<li><a href="https://github.com/cboettig/nonparametric-bayes/blob/fd3585230a01afa525f456d13f5f958df4606006/inst/examples/sensitivity-allen.md">sensitivity-allen.md</a></li>
</ul>
<p>Overall:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(yields_dat) + <span class="kw">geom_density</span>(<span class="kw">aes</span>(value)) </code></pre>
<figure>
<img src="http://farm6.staticflickr.com/5455/9387766618_9772fca7e1_o.png" alt="plot of chunk unnamed-chunk-5" /><figcaption>plot of chunk unnamed-chunk-5</figcaption>
</figure>
<p>Compare over 12 simulation configurations with different parameters and different training data, the GP does consistently well. Only one case has relatively poor outcomes.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(yields_dat) + <span class="kw">geom_density</span>(<span class="kw">aes</span>(value, <span class="dt">color=</span><span class="kw">as.factor</span>(simulation), <span class="dt">fill=</span><span class="kw">as.factor</span>(simulation)), <span class="dt">alpha=</span>.<span class="dv">5</span>)</code></pre>
<figure>
<img src="http://farm4.staticflickr.com/3834/9385132341_ea3cffbb8d_o.png" alt="plot of chunk unnamed-chunk-6" /><figcaption>plot of chunk unnamed-chunk-6</figcaption>
</figure>
<ul>
<li>Run data backup to local disk</li>
<li>NSF expense accounting</li>
<li>calendar updates</li>
<li>software paper policies email</li>
<li>Luke Arbor reply</li>
<li>looking at quick way to check out solutions to sums of Fisher wave equation over different diffusion coefficients (heterogenous dispersal rates), e.g. <a href="http://www.os-scientific.org/physics/react/index.html#fisher_kolgomorov_doc">in C</a></li>
<li>Reply to Alex about wave + diffusion type solution pattern.</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r">f1 &lt;- function(t, <span class="dt">y =</span> <span class="dv">4</span>, <span class="dt">r =</span> <span class="dv">2</span>, <span class="dt">s =</span> <span class="dv">1</span>) (r * <span class="kw">dnorm</span>(y, <span class="dv">0</span>, <span class="dv">2</span> * r * <span class="kw">sqrt</span>(t)) + s * 
    <span class="kw">dnorm</span>(y, <span class="dv">0</span>, <span class="dv">2</span> * s * <span class="kw">sqrt</span>(t)))/(<span class="kw">dnorm</span>(y, <span class="dv">0</span>, <span class="dv">2</span> * r * <span class="kw">sqrt</span>(t)) + <span class="kw">dnorm</span>(y, <span class="dv">0</span>, 
    <span class="dv">2</span> * s * <span class="kw">sqrt</span>(t)))
<span class="kw">curve</span>(f1, <span class="fl">0.01</span>, <span class="dv">8</span>, <span class="dt">xlab =</span> <span class="st">&quot;time&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;mean dispersal&quot;</span>)</code></pre>
<figure>
<img src="http://farm4.staticflickr.com/3704/9387637650_939dd70f9c_o.png" alt="mean dispersal rate declines as slow population arrives" /><figcaption>mean dispersal rate declines as slow population arrives</figcaption>
</figure>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/28/nonparametric-bayes-wrapping-up-sensitivity.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/25/formal-semantics-with-reml.html">Formal Semantics With Reml</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 25 Jul 2013</p>

<p style="font-style:italic"> pageviews: 20 </p>

<article>
<div class="excerpt">
<p>A lot has been happening in the <code>reml</code> development over on our package Github page, particularly through discussion on the <a href="http://ropensci.org/reml">issues tracker</a> we have been working through several of the major conceptual challenges in designing the package.</p>
<h3 id="semantics">Semantics</h3>
<p><strong>Writing Semantics</strong>. Lots of elements of the EML schema have semantic meaning, but we can’t start leveraging ontological tools while these terms are defined only as part of the Schema vocabulary. In their current state, they are just “meaningless bits of syntax”, e.g. the schema only defines <code>gram</code> as a valid text string that lives at <code>//attribute/measurementScale/ratio/unit/standardUnit</code>. A computer reading the schema could never formally decide that this is the same gram we use to measure mass. We use formal semantics to say this: that is, that text-string <code>gram</code> means <a href="http://ecoinformatics.org/oboe/oboe.1.0/oboe-standards.owl#Gram">http://ecoinformatics.org/oboe/oboe.1.0/oboe-standards.owl#Gram</a>, which is embued with meaning.</p>
<p>We’ve made some definite progress in how to go about making this leap from structure to meaning after considering a variety of potential avenues. A few <a href="https://github.com/ropensci/reml/issues/5#issuecomment-21573824">open questions</a> (and a lot of implementation) remain. A useful observation comes from Ben in pointing out how we can add RDF directly to the EML, rather than relying on a second external schema file:</p>
<pre class="sourceCode xml"><code class="sourceCode xml"><span class="kw">&lt;eml&gt;</span>
…
<span class="kw">&lt;dataTable</span><span class="ot"> id=</span><span class="st">&quot;http://some.namespace#myUniqueEntityId1&quot;</span><span class="kw">&gt;</span>
        <span class="kw">&lt;attribute</span><span class="ot"> id=</span><span class="st">&quot;http://some.namespace#myUniqueAttributeId1&quot;</span><span class="kw">/&gt;</span>
        <span class="kw">&lt;attribute</span><span class="ot"> id=</span><span class="st">&quot;http://some.namespace#myUniqueAttributeId2&quot;</span><span class="kw">/&gt;</span>
<span class="kw">&lt;/dataTable&gt;</span>
…
<span class="kw">&lt;additionalMetadata&gt;</span>
        <span class="kw">&lt;describes&gt;</span>http://some.namespace#myUniqueAttributeId1<span class="kw">&lt;/describes&gt;</span>
        <span class="kw">&lt;metadata&gt;</span>
                <span class="co">&lt;!-- RDF stuff here that annotates http://some.namespace#myUniqueAttributeId1 --&gt;</span>
                <span class="kw">&lt;rdf:RDF</span>
<span class="ot">                        xmlns:rdf=</span><span class="st">&quot;http://www.w3.org/1999/02/22-rdf-syntax-ns#&quot;</span>
<span class="ot">                        xmlns:o=</span><span class="st">&quot;http:/oboe-core#&quot;</span><span class="kw">&gt;</span>
                        <span class="kw">&lt;rdf:Description</span><span class="ot"> rdf:about=</span><span class="st">&quot;http://some.namespace#myUniqueAttributeId1&quot;</span><span class="kw">&gt;</span>
                                <span class="kw">&lt;o:entity&gt;</span>Air<span class="kw">&lt;/o:entity&gt;</span>
                                <span class="kw">&lt;o:characteristic&gt;</span>Temperature<span class="kw">&lt;/o:characteristic&gt;</span>
                                <span class="kw">&lt;cd:unit&gt;</span>Celsius<span class="kw">&lt;/cd:unit&gt;</span>
                        <span class="kw">&lt;/rdf:Description&gt;</span>
                <span class="kw">&lt;/rdf:RDF&gt;</span>
        <span class="kw">&lt;/metadata&gt;</span>
<span class="kw">&lt;/additionalMetadata&gt;</span>
<span class="kw">&lt;additionalMetadata&gt;</span>
        <span class="kw">&lt;describes&gt;</span>http://some.namespace#myUniqueAttributeId2<span class="kw">&lt;/describes&gt;</span>
        <span class="kw">&lt;metadata&gt;</span>
                <span class="co">&lt;!-- RDF stuff here that annotates http://some.namespace#myUniqueAttributeId2 --&gt;</span>
        <span class="kw">&lt;/metadata&gt;</span>
<span class="kw">&lt;/additionalMetadata&gt;</span>
<span class="kw">&lt;/eml&gt;</span></code></pre>
<p>A complete implementation goes beyond translating the schema vocabulary to ontogogical terms where it makes sense to do so (e.g. units of measurement, taxonomy)</p>
<p>Where else does it make sense to translate schema vocabularly into semantics?</p>
<ul>
<li>Licenses? (dc:rights)</li>
<li>Titles, people, dates, rest of citation information (via <code>dc:title</code>, <code>dc:contributor</code>, <code>prism:</code> or <code>fabio:</code> etc)</li>
<li>attribute</li>
</ul>
<p>Would it be possible to create a XSLT stylesheet that one would apply to an EML file to produce corresponding RDF/XML file (that I could then amend to the original EML in an <code>additionalMetadata</code> element)? I see the value of having the semantic metadata embedded in a single file, but rather than merely implementing R code that says “when you write a <code>&lt;standardUnit&gt;Celsius</code> node, also write this <code>additionalMetadata</code> node, it seems logical to have a tool that could simply read EML and say,”hey, <code>&lt;standardUnit&gt;Celsius</code>, that’s an <code>&lt;RDF ...</code>&quot;. Perhaps similar mappings could be made for taxonomy (though might need to query ubio for the identifier etc and so be beyond the scope of XSLT)?</p>
<p>Such an XSLT might also map things like <code>&lt;creator&gt;</code>, <code>&lt;title&gt;</code>, <code>&lt;intellectionalRights&gt;</code> etc to their ontological meanings too. Of course I don’t know anything about XSLT, and before we can programmatically combine spreadsheets, etc we really need to be adding custom semantics for all attributes, not just making RDFs of the things that are already defined…</p>
<p>Whoops, we can get lost down the rabbit hole quickly here. e.g. Why not write all data to triples (e.g. see tools for <a href="http://www.w3.org/wiki/ConverterToRdf#CSV_.28Comma-Separated_Values.29">RDF from CSV</a>… (Longevity of RDF? Of any given XML schema?)</p>
<ul>
<li>RDF as schema-less XML?</li>
</ul>
<p><strong>Reading Semantics</strong> (and actually doing something with it): the really hard problem motivating this exercise: <a href="https://github.com/ropensci/reml/issues/8">#8</a>. We’ll get to this one eventually.</p>
<h3 id="schema-oriented-design">Schema-oriented design</h3>
<p>We can leverage the <code>.xsd</code> schema files much more directly in creating our parser using a strategy built around S4 classes, as discussed in issue <a href="https://github.com/ropensci/reml/issues/38">#38</a>. This leaves us with a lot of coercion to do. On the up side, we have reasonble S4 classes representing the EML contents, so we can consider it “parsed”. Defining how we write into these S4 objects instead of how we write into XML directly also makes our implementation more robust to changes in the schema. Coercion with S4 can leverage a variety of heurstics to make this pretty easy (e.g. simple matching of slot names, guessing parsing etc). Also allows us to have things like <code>data.frames</code> that carry their metadata around with them.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/25/formal-semantics-with-reml.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

    <div class="row">
      
        <div class="span3">
          <header><h4><a href="/2013/07/22/XML-parsing-strategies.html">Xml Parsing Strategies</a></h4></header>
<p><span>parsing NeXML</span></p>
<p style="font-style:italic"> 22 Jul 2013</p>

<p style="font-style:italic"> pageviews: 14 </p>

<article>
<div class="excerpt">
<p>Writing out some advice for our GSOC student on XML parsing. (Now filed as <a href="https://github.com/shumelchyk/RNeXML/issues/11">RNeXML/#11</a></p>
<hr />
<p>Here is some quick background on different ways we might go about extracting NeXML into an R object we want to work with. We can use S4 classes, R’s native <code>data.frame</code> and <code>list</code> types, or extract specific terms of interest with xpath. I illustrate each of these below using the example “trees.xml” in your repository.</p>
<p>It would be good to decide on which strategy you want. As you seem to be leaning to the S4 already, note how I have made the slot labels correspond exactly to the attribute and node labels of the NeXML. This lets us use the <code>xmlToS4</code> method:</p>
<h2 id="using-s4-classes">Using S4 classes</h2>
<p>We can define S4 class structures to correspond to the elements we want to extract. Elements omitted from our definition will be ignored.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;node&quot;</span>,
    <span class="kw">representation</span>(<span class="dt">id    =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">label =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">otu   =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">about =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">meta =</span> <span class="st">&quot;meta&quot;</span>))</code></pre>
<p>Note that class “meta” is not yet defined, we will define it below. Note also that we could omit any slots we don’t care about, and they would simply be ignored. (Try deleting the <code>meta</code> line, or the <code>about</code> line.)</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;edge&quot;</span>,
    <span class="kw">representation</span>(<span class="dt">source =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">target =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">id     =</span> <span class="st">&quot;character&quot;</span>,
                   <span class="dt">length =</span> <span class="st">&quot;numeric&quot;</span>))</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setClass</span>(<span class="st">&quot;meta&quot;</span>,
         <span class="kw">representation</span>(<span class="dt">id =</span><span class="st">&quot;character&quot;</span>,
                        <span class="dt">property=</span><span class="st">&quot;character&quot;</span>,
                        <span class="dt">content=</span><span class="st">&quot;logical&quot;</span>,
                        <span class="st">&#39;xsi:type&#39;</span>=<span class="st">&quot;character&quot;</span>,
                        <span class="dt">datatype=</span><span class="st">&quot;character&quot;</span>))

<span class="kw">setAs</span>(<span class="st">&quot;XMLInternalElementNode&quot;</span>, <span class="st">&quot;meta&quot;</span>, function(from) <span class="kw">xmlToS4</span>(from))</code></pre>
<p>In order for <code>xmlToS4</code> to work on nodes, I need the Coercion method for <code>meta</code> too, defined above. We are now ready to rock and roll:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(XML)
doc &lt;- <span class="kw">xmlParse</span>(<span class="st">&quot;tests/examples/trees.xml&quot;</span>)

edges &lt;- <span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>)
edges &lt;- <span class="kw">lapply</span>(edges, xmlToS4)

nodes &lt;- <span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>)
nodes &lt;- <span class="kw">lapply</span>(nodes, xmlToS4)</code></pre>
<p>My opinion is that we should not bother writing validator methods etc for these classes we have just defined at this time. We can already validate the XML produced against the schema, and as the class definitions follow those from the schema, we should be reasonably robust. Happy to hear counter-arguments.</p>
<h2 id="extracting-attributes-or-values-directly-by-xpath-query">Extracting Attributes or values directly by xpath query</h2>
<p>A different route does not define class types at all, but just extracts the attributes we want. This is a bit more fast and loose.</p>
<pre class="sourceCode r"><code class="sourceCode r">edges = <span class="kw">xpathSApply</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces =</span> <span class="st">&quot;x&quot;</span>, xmlAttrs)
nodes = <span class="kw">xpathSApply</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces =</span> <span class="st">&quot;x&quot;</span>, 
  function(x) 
    <span class="kw">c</span>(<span class="kw">xmlGetAttr</span>(x, <span class="st">&quot;id&quot;</span>, <span class="ot">NA</span>, as.character),  
      <span class="kw">xmlGetAttr</span>(x, <span class="st">&quot;otu&quot;</span>, <span class="ot">NA</span>, as.character)))</code></pre>
<h2 id="coercing-into-standard-r-forms">Coercing into standard R forms</h2>
<p>This approach can be very quick and powerful if the data conforms to this structure.</p>
<pre class="sourceCode r"><code class="sourceCode r">XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:node&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>))
XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">getNodeSet</span>(doc, <span class="st">&quot;//x:tree[@id = &#39;tree1&#39;]/x:edge&quot;</span>, <span class="dt">namespaces=</span><span class="st">&quot;x&quot;</span>))</code></pre>
<h2 id="without-xpath">without XPath</h2>
<p>We can skip over XPath based expressions by using xmlToList immediately:</p>
<pre class="sourceCode r"><code class="sourceCode r">nex &lt;- <span class="kw">xmlToList</span>(doc)</code></pre>
<p>Unfortunately, this will do some possibly unexpected things: for instance, attributes are converted to list elements (<code>node$id</code>, <code>node$otu</code>, etc) on simple nodes (e.g. <code>&lt;nodes&gt;</code> without a <code>&lt;meta&gt;</code> node), while if they contain another node, attributes are one sub-list and the containing nodes another (e.g. <code>node$meta</code> and <code>node$.attr</code>, etc)</p>
<p>Also note that instead of xpath queries as in the <code>xpathSApply</code> and <code>getNodeSet</code> examples above, we can index nodes the way would lists:</p>
<p>e.g. <code>xmlRoot(doc)[[&quot;trees&quot;]][[&quot;tree&quot;]]</code> instead of <code>getNodeSet(doc, &quot;//trees/tree&quot;)[[1]]</code> (note that the former returns the first <code>&lt;tree&gt;</code> node in the <code>&lt;trees&gt;</code> node, while the latter returns all <code>&lt;tree&gt;</code> nodes unless we just ask for the <code>[[1]]</code></p>
<p>So this gives us a table of all node and edge elements:</p>
<pre class="sourceCode r"><code class="sourceCode r">XML:::<span class="kw">xmlAttrsToDataFrame</span>(<span class="kw">xmlRoot</span>(doc)[[<span class="st">&quot;trees&quot;</span>]][[<span class="st">&quot;tree&quot;</span>]])</code></pre>
<p>Which gives us:</p>
<pre class="sourceCode r"><code class="sourceCode r">   id label root  otu about source target  length
<span class="dv">1</span>  n1    n1 true &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">2</span>  n2    n2 &lt;<span class="ot">NA</span>&gt;   t1  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">3</span>  n3    n3 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">4</span>  n4    n4 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;   <span class="co">#n4   &lt;NA&gt;   &lt;NA&gt;    &lt;NA&gt;</span>
<span class="dv">5</span>  n5    n5 &lt;<span class="ot">NA</span>&gt;   t3  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">6</span>  n6    n6 &lt;<span class="ot">NA</span>&gt;   t2  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">7</span>  n7    n7 &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">8</span>  n8    n8 &lt;<span class="ot">NA</span>&gt;   t5  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">9</span>  n9    n9 &lt;<span class="ot">NA</span>&gt;   t4  &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;   &lt;<span class="ot">NA</span>&gt;    &lt;<span class="ot">NA</span>&gt;
<span class="dv">10</span> e1  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n1     n3 <span class="fl">0.34534</span>
<span class="dv">11</span> e2  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n1     n2  <span class="fl">0.4353</span>
<span class="dv">12</span> e3  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n3     n4   <span class="fl">0.324</span>
<span class="dv">13</span> e4  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n3     n7  <span class="fl">0.3247</span>
<span class="dv">14</span> e5  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n4     n5   <span class="fl">0.234</span>
<span class="dv">15</span> e6  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n4     n6  <span class="fl">0.3243</span>
<span class="dv">16</span> e7  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n7     n8 <span class="fl">0.32443</span>
<span class="dv">17</span> e8  &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt; &lt;<span class="ot">NA</span>&gt;  &lt;<span class="ot">NA</span>&gt;     n7     n9  <span class="fl">0.2342</span></code></pre>
<p>Sometimes this notation is cleaner than the xpath, sort of up to you.</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/22/XML-parsing-strategies.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/19/notes.html">Notes from chat with Duncan</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 19 Jul 2013</p>

<p style="font-style:italic"> pageviews: 11 </p>

<article>
<div class="excerpt">
<ul>
<li>Keep metadata as part of the object</li>
<li>Recursion problem coming up.</li>
</ul>
<h2 id="validating">Validating:</h2>
<ul>
<li>https://github.com/ropensci/reml/issues/7</li>
<li>Duncan has code using Matt’s java API</li>
<li>If the EML isn’t valid, so what? Important for developers, users don’t care.</li>
</ul>
<h2 id="api-design">API design</h2>
<ul>
<li>what’s with the <code>do.call</code>? This forces evaluation, can mess up lazyLoading, making for inefficient function handling.<br /></li>
<li>Schema to create S4 objects</li>
<li>Create R objects first, and then toXML</li>
</ul>
<h2 id="xmlschema">XMLSchema</h2>
<ul>
<li>will improve as we go</li>
</ul>
<h2 id="paper-outline">Paper outline</h2>
<ul>
<li>EML exists – metadata vs vertical</li>
<li>metadata manual in, manual out</li>
<li>data synthesis?</li>
</ul>
<h3 id="r-object-to-xml">R object to XML</h3>
<ul>
<li>Accept / identify losely defined fields (lists)</li>
<li>Manually create a coercision method between R’s objects, e.g. “person” (attempt match of named elements to list slots automatically).<br /></li>
<li>Separate coersion methods for read and write.</li>
</ul>
<h3 id="xml-to-r-object">XML to R object</h3>
<ul>
<li>ParseEML - because it saves us a lot to know this is EML.<br /></li>
<li>Coerce into generic S4 R classes (class definitions from schema)</li>
<li>Coerce automatically if element names are the same. Otherwise, manually coerce.</li>
</ul>
<h3 id="further-possibilities">Further possibilities</h3>
<ul>
<li>Translations to other languages (e.g. python) can be automated.<br /></li>
<li>Then add higher level mapping manually.</li>
</ul>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/19/notes.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
        <div class="span3">
          <header><h4><a href="/2013/07/17/updates-on-managing-references.html">Updates On Managing References</a></h4></header>
<p><span></span></p>
<p style="font-style:italic"> 17 Jul 2013</p>

<p style="font-style:italic"> pageviews: 26 </p>

<article>
<div class="excerpt">
<h3 id="mendeley-api-for-jekyll">Mendeley API for Jekyll</h3>
<p>Finally finished off my <a href="https://github.com/cboettig/labnotebook/blob/7f7137fbdd4ca2eed86765b61561c283b4cd5811/_plugins/jekyll-labnotebook-plugins/mendeley.rb">Mendeley API for Jekyll</a> (The top uses the Ruby API implementation from Michael Bus, copied here since it is not yet provided as a gem. scroll to the bottom for my plugin code). The Mendeley API implementation is remarkably clumsy. For instance, rather than allowing queries to the API that specify a user’s folder and a sort-by field, we are forced to first query the folder for a list of IDs and then query document details <em>on every document in the folder</em>, after which we can sort locally. Hence to show recent additions to a folder with 100s of documents requires 100s of API calls, instead of one single call. Straight-forward to program but incredibly slow and frustrating. Apparently this might be addressed in an eventual update of the API.</p>
<p>Meanwhile, at least I know have a reasonably powerful tool to add reading lists by subject area directly from my library rather than pulling from my rather more aggregated group RSS feeds. Due to the ridiculous number of API calls, the plugin writes out the data in JSON, which provides a nice local database of the publication metadata we can read in from instead. The plugin will use local JSON file for the data if it exists – so one must delete the JSON file to trigger an update.</p>
<p>Currently the plugin lists the <code>n</code> most recent papers in a given folder. Further adaptation might be nice to allow for richer filtering, e.g. by tag,author, etc.</p>
<h3 id="direct-access-to-the-local-sql-database">Direct access to the local SQL database</h3>
<p>Accessing personal Mendeley data from the API is actually quite silly when the data is already available in a local SQLite database.</p>
<p>Using R to explore and query my local Mendeley sqlite database directly:</p>
<p>Make a copy of the database first</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cp</span> ~/.local/share/data/Mendeley Ltd./Mendeley Desktop/cboettig@gmail.com@www.mendeley.com.sqlite tmp.sqlite</code></pre>
<p>Then in R, read in the database and show available tables (see <a href="http://sandymuspratt.blogspot.com/2012/11/r-and-sqlite-part-1.html">example tutorial</a>):</p>
<p><code>{r} library('sqldf') db &lt;- dbConnect(SQLite(), dbname=&quot;tmp.sqlite&quot;) dbListTables(db)</code></p>
<p>Most metadata for all documents in the library appears in table <code>Documents</code>. <code>dbListFields</code> shows the column headers, though alternatively we can just read this information into an R <code>data.frame</code> (This makes sense for reasonable size databases; for large databases we will want to query the data via SQL commands directly rather than attempting to load it all into memory).</p>
<p><code>{r} dbListFields(db, &quot;Documents&quot;) documents &lt;- dbReadTable(db, &quot;Documents&quot;)</code></p>
<p>Several useful fields are not included in this data frame. In particular, Author, Tag, Folder and Keywords. Some extra leg-work is needed to add this data. For instance <code>DocumentFolders</code> gives us the map of each article <code>id</code> to a folder <code>id</code>, and then <code>Folders</code> gives the details of that Folder.</p>
<p>Using the <code>data.table</code> package, indexing by <code>id</code>, and using join methods would probably be the most efficient way to do this. (example may follow when I get the chance).</p>
 
<!-- not that raw_content depends on custom generator, 
     see _plugins/jekyll-labnotebook-plugins/raw_content --> 
</article> 
<p> <a href="/2013/07/17/updates-on-managing-references.html"> <em>Read more</em></a> </p>
<br />
<br />


        </div>
      
    </div>

  </div>
</div> <!--end row -->

<div class="row">
  <div class="span11 offset1">
    <div class="socialicons">
      <p> <a href="/archive.html"><i class="icon-calendar"></i> All entries by date</a></p> 
      <p> <a href="/categories.html"><i class="icon-list"></i> All entries by category</a> </p>
      <p> <a href="/tags.html"><i class="icon-tags"></i> All entries by tag</a> </p>
    </div>
  </div> <!--end span9 -->
</div> <!--end row -->




<footer class="footer">
  <!-- ************* Buttons to toggle theme CSS ********************** -->
    <div class="row">
      <div class="span12">
        <form style="font-size:10px" class="pull-right">
          <a onclick="switch_style('dark'); 
                      recordOutboundLink(this, 'Outbound Links', 'dark theme'); 
                      return false;" 
             name="theme" value="dark" id="dark" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to dark theme"><i class="icon-adjust"></i>
                                   </span></a>
          
          <a onclick="switch_style('light'); 
                      recordOutboundLink(this, 'Outbound Links', 'light theme'); 
                      return false;" 
             name="theme" value="light" id="light" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to light theme"><i class="icon-certificate"></i>
                                  </span></a>


           <a onclick="switch_style('white'); 
                      recordOutboundLink(this, 'Outbound Links', 'white theme'); 
                      return false;" 
             name="theme" value="white" id="white" 
             class="btn btn-mini"><span class="showtooltip" 
                                        title="switch to white theme"><i class="icon-circle-blank"></i>
                                 </span></a>
        </form>
      </div>
    </div>

<!--************** FOAF information to social networks ***************************** -->
  <div class="row">
    <div class="span3 socialicons" style="font-size:20px" typeof="foaf:Person" about="http://carlboettiger.info#me">
      <p>
          <script type="text/javascript" src="/assets/js/obfuscate-email-link.js" language="javascript"></script> 
          <a rel="foaf:account" alt="twitter" href="https://twitter.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Twitter'); 
             return false;"><span class="showtooltip" title="follow me on twitter (reading, discussing)"><i class="icon-twitter"></i></span></a> 
          <a rel="foaf:account" alt="github" href="https://github.com/cboettig" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Github'); 
             return false;"><span class="showtooltip" title="follow me on Github (code, research)"><i class="icon-github"></i></span></a>
      <!--
          <a rel="foaf:account" href="https://plus.google.com/" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'GPlus'); 
             return false;"><i class="icon-google-plus"></i></a>
          <a rel="foaf:account" href="http://www.mendeley.com/profiles/carl-boettiger" 
             onclick="recordOutboundLink(this, 'Outbound Links', 'Mendeley'); 
             return false;"><img src="/assets/img/icon-mendeley.png" alt="mendeley" /></a> 
           citations on google-scholar
           stackoverflow
      -->
      <a alt="rss" type="application/atom+xml" href="/blog.xml"  
         class="showtooltip" title="RSS feeds for my blog-style entries. See the feed on my lab notebook (/atom.xml) to follow all entries instead." 
         onclick="recordOutboundLink(this, 'Outbound Links', 'RSS'); 
         return false;"><i class="icon-rss"></i></a>
       </p>
    </div>
    <!--**************** End social links **************************** -->
    <div class="span1">
      <br />
    </div>
    <div class="span4">
      <p>
      <a onclick="recordOutboundLink(this, 'Outbound Links', 'ONS_claim'); return false;" href="http://onsclaims.wikispaces.com/"><img src="http://onsclaims.wikispaces.com/file/view/ons-aci2-icon.png" alt="ONS" class="showtooltip" title="An Open Notebook Science (ONS) project claim: Entry provides all content (AC) immediately (I) or without significant delay.  See link for details"/></a>

      <a title="This site uses linked data semantics. Click to extract as RDF XML." class="btn btn-mini showtooltip" style="font-size: .8em" 
       href="http://any23.org/?format=rdfxml&validate=validate&uri=http://carlboettiger.info/lab-notebook.html"><i 
         class="icon-cloud-download"  onclick="recordOutboundLink(this, 'Outbound Links', 'RDF'); return false;"></i> RDF</a>
      </p>
    </div>
    <div class="span1">
      <br />
    </div>
    <div class="span3">
      <p>
      <a rel="license" property="http://creativecommons.org/ns#license" href="http://creativecommons.org/publicdomain/zero/1.0/" onclick="recordOutboundLink(this, 'Outbound Links', 'CC0'); return false;"><img src="http://i.creativecommons.org/l/zero/1.0/88x31.png" alt="CC0"/></a> 
      </p>
    </div>

  </div>
  
<!-- COinS metadata (for citation managers like Zotero etc), goes in body text -->
  <span
      class="Z3988" 
      title="ctx_ver=Z39.88-2004
      &amp;rft_val_fmt=info%3Aofi%2Ffmt%3Akev%3Amtx%3Adc
      &amp;rfr_id=info%3Asid%2Focoins.info%3Agenerator
      &amp;rft.title=Lab Notebook
      &amp;rft.creator=Carl Boettiger
      &amp;rft.date=
      &amp;rft.language=EN
      &amp;rft.rights=CC0
      &amp;rft_id=http://carlboettiger.info/lab-notebook.html">
  </span>



</footer>




    <!-- Le javascript
    ================================================== -->

    <!-- Placed at the end of the document so the pages load faster -->
    <!-- JQuery, used on a few pages -->
    <script type="text/javascript" src="/assets/js/jquery.js"></script>
    <!-- Equations using MathJax -->
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config"> MathJax.Hub.Config({ TeX: { equationNumbers: {autoNumber: "all"} } });       </script>
    <!-- Twitter Bootstrap Javascript -->
    <script src="/assets/js/bootstrap.min.js"></script>
    <!-- Tooltip javascript -->
    <script type="text/javascript">
      $(document).ready(function (){
        $(".showtooltip").tooltip();
      });
    </script>

    <!-- Marran's Search Javascript -->
    <script type="text/javascript" src="/assets/js/porter-stemmer.js"></script>
    <script type="text/javascript" src="/assets/js/site-search.js"></script>

    <!-- Code collapse Javascript -->
    <script type="text/javascript">
    $(document).ready(function(){
      $("#toggle_code").click(function(){
        $(".highlight").toggle();
        $(".sourceCode").toggle();
      });
    });
    </script>


  <script type="text/javascript">
          var _gaq = _gaq || [];
          _gaq.push(['_setAccount', 'UA-18401403-1']);
          _gaq.push(['_trackPageview']);
          (function() {
            var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
            ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
            var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
          })();
  </script>



<script type="text/javascript">
function recordOutboundLink(link, category, action) {
  try {
    var pageTracker=_gat._getTracker("UA-18401403-1");
    pageTracker._trackEvent(category, action);
    setTimeout('document.location = "' + link.href + '"', 100)
  }catch(err){}
}
</script>




    


    </div>
  </body>
</html>

